<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Detection of post-transcriptional modifications in high throughput sequencing data • RNAmodR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Detection of post-transcriptional modifications in high throughput sequencing data">
<meta property="og:description" content="RNAmodR provides classes and workflows for loading/aggregation data
    from high througput sequencing aimed at detecting post-transcriptional 
    modifications through analysis of specific patterns. In addition, utilities
    are provided to validate and visualize the results. The RNAmodR package
    provides a core functionality from which specific analysis strategies can
    be easily implemented as a seperate package.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.7.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/FelixErnst/RNAmodR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="rnamodr-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#rnamodr-" class="anchor"></a>RNAmodR <img src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/master/RNAmodR/RNAmodR.png" height="300" align="right">
</h1></div>
<!-- badges: start -->

<p>Post-transcriptional modifications can be found abundantly in rRNA and tRNA and can be detected classically via several strategies. However, difficulties arise if the identity and the position of the modified nucleotides is to be determined at the same time. Classically, a primer extension, a form of reverse transcription (RT), would allow certain modifications to be accessed by blocks during the RT or changes in the cDNA sequences. Other modification would need to be selectively treated by chemical reactions to influence the outcome of the reverse transcription.</p>
<p>With the increased availability of high throughput sequencing, these classical methods were adapted to high throughput methods allowing more RNA molecules to be accessed at the same time. With these advances post-transcriptional modifications were also detected on mRNA. Among these high throughput techniques are for example Pseudo-Seq (<a href="#Literature">Carlile et al. 2014</a>), RiboMethSeq (<a href="#Literature">Birkedal et al. 2015</a>) and AlkAnilineSeq (<a href="#Literature">Marchand et al. 2018</a>) each able to detect a specific type of modification from footprints in RNA-Seq data prepared with the selected methods.</p>
<p>Since similar pattern can be observed from some of these techniques, overlaps of the bioinformatical pipeline already are and will become more frequent with new emerging sequencing techniques.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>The current version of the <code>RNAmodR</code> package is available from Bioconductor.</p>
<pre><code>if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("RNAmodR")
library(RNAmodR)</code></pre>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><code>RNAmodR</code> implements classes and a workflow to detect post-transcriptional RNA modifications in high throughput sequencing data. It can be easily adapted for new methods and can help during the phase of initial method development as well as more complex screenings.</p>
<p>Briefly, from the <code>SequenceData</code> and <code>SequenceDataFrame</code> classes, specific subclasses are derived for accessing specific aspects of aligned reads, e.g.  5’-end positions or pileup data. With these, a <code>Modifier</code> class can be used to detect specific patterns for individual types of modifications. The <code>SequenceData</code> classes can be shared by different <code>Modifier</code> classes enabling easy adaptation to new methods and benefiting from other efforts in method development.</p>
<p>Whereas, the <code>SequenceData</code> classes are used to hold the data, <code>Modifier</code> classes are used to detect certain features within high throughput sequencing data to assign the presence of specific modifications for an established pattern. The <code>Modifier</code> class is virtual and can be adapted for individual methods, whereas the type of nucleotide under investigation is specified by inheriting from the virtual <code>RNAModifier</code> and <code>DNAModifier</code> classes. To fix the data processing and detection strategy, for each type of sequencing method a <code>Modifier</code> class can be developed alongside to detect modifications.</p>
<p>For further details have a look at the vignette or the additional packages for implementation examples:</p>
<ul>
<li><a href="https://doi.org/doi:10.18129/B9.bioc.RNAmodR.RiboMethSeq">RNAmodR.RiboMethSeq</a></li>
<li><a href="https://doi.org/doi:10.18129/B9.bioc.RNAmodR.AlkAnilineSeq">RNAmodR.AlkAnilineSeq</a></li>
<li>
<a href="https://doi.org/doi:10.18129/B9.bioc.RNAmodR.ML">RNAmodR.ML</a> for classes fascilitating machine learning approaches</li>
</ul>
</div>
<div id="literature" class="section level1">
<h1 class="hasAnchor">
<a href="#literature" class="anchor"></a>Literature</h1>
<ul>
<li><p>Carlile, Thomas M., Maria F. Rojas-Duran, Boris Zinshteyn, Hakyung Shin, Kristen M. Bartoli, and Wendy V. Gilbert (2014): “Pseudouridine Profiling Reveals Regulated mRNA Pseudouridylation in Yeast and Human Cells.” Nature 515 (7525): 143–46.</p></li>
<li><p>Birkedal, Ulf, Mikkel Christensen-Dalsgaard, Nicolai Krogh, Radhakrishnan Sabarinathan, Jan Gorodkin, and Henrik Nielsen (2015): “Profiling of Ribose Methylations in Rna by High-Throughput Sequencing.” Angewandte Chemie (International Ed. In English) 54 (2): 451–55. <a href="https://doi.org/10.1002/anie.201408362" class="uri">https://doi.org/10.1002/anie.201408362</a>.</p></li>
<li><p>Marchand, Virginie, Lilia Ayadi, <strong>Felix G. M. Ernst</strong>, Jasmin Hertler, Valérie Bourguignon-Igel, Adeline Galvanin, Annika Kotter, Mark Helm, <strong>Denis L. J. Lafontaine</strong>, and Yuri Motorin (2018): “AlkAniline-Seq: Profiling of m7G and m3C Rna Modifications at Single Nucleotide Resolution.” Angewandte Chemie International Edition 57 (51): 16785–90. <a href="https://doi.org/10.1002/anie.201810946" class="uri">https://doi.org/10.1002/anie.201810946</a>.</p></li>
</ul>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/RNAmodR">https://​www.bioconductor.org/​packages/​RNAmodR</a>
</li>
<li>Browse source code at <br><a href="https://github.com/FelixErnst/RNAmodR/">https://​github.com/​FelixErnst/​RNAmodR/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/FelixErnst/RNAmodR/issues">https://​github.com/​FelixErnst/​RNAmodR/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0">Artistic-2.0</a></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing RNAmodR</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Felix G.M. Ernst <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-5064-0928" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Denis L.J. Lafontaine <br><small class="roles"> Contributor, funder </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/FelixErnst/RNAmodR/actions/"><img src="https://github.com/FelixErnst/RNAmodR/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="http://bioconductor.org/checkResults/devel/bioc-LATEST/RNAmodR/"><img src="https://bioconductor.org/shields/build/devel/bioc/RNAmodR.svg" alt="BioC Build"></a></li>
<li><a href="https://codecov.io/gh/FelixErnst/RNAmodR"><img src="https://codecov.io/gh/FelixErnst/RNAmodR/branch/master/graph/badge.svg" alt="codecov"></a></li>
<li><a href="https://doi.org/doi:10.18129/B9.bioc.RNAmodR"><img src="https://bioconductor.org/shields/years-in-bioc/RNAmodR.svg" alt="BioC Years"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
