---
title: "RNAmod: pre-mapping preparation of genome and annotation"
author: "Felix G.M. Ernst"
date: "`r Sys.Date()`"
package: tRNAscan2GRanges
abstract: >
  Example conversion of extracting of sequences and annotation to construct a
  special transcriptome to map reads against
output: 
  BiocStyle::html_document:
    toc: true
    toc_float: true
    df_print: paged
vignette: >
  %\VignetteIndexEntry{RNAmod - pre-mapping preparation of genome and annotation}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown(css.files = c('custom.css'))
```

# Introduction

```{r tRNA,eval=FALSE}
library(RNAmodR)
matchFastaToGff <- function(gfffile, fafile){
  gff <- rtracklayer::import.gff3(gfffile)
  fsa <- Biostrings::readDNAStringSet(fafile)
  names(fsa) <- unique(as.character(GenomeInfoDb::seqnames(gff)))
  Biostrings::writeXStringSet(fsa, fafile)
}
```

# Converting rRNA, mRNA and tRNA genes to seperate chromosomes

## S. cerevisiae

```{r sacCer3,eval=FALSE}
folder <- "genomes/sacCer3/"
gfffile <- paste0(folder,"sacCer3_R64-2-1.gff3")
fafile <- paste0(folder,"sacCer3_R64-2-1.fasta")
tRNAscanfile <- paste0(folder,"sacCer3-tRNAs.ss.sort")
matchFastaToGff(gfffile,fafile)
# get specific rRNA genes and save as fasta/gff
gff <- rtracklayer::import.gff3(gfffile)
RDN25 <- S4Vectors::mcols(gff[grepl("RDN25",S4Vectors::mcols(gff)$ID),])$ID
RDN18 <- S4Vectors::mcols(gff[grepl("RDN18",S4Vectors::mcols(gff)$ID),])$ID
RDN58 <- S4Vectors::mcols(gff[grepl("RDN58",S4Vectors::mcols(gff)$ID),])$ID
RNDmt <- c(S4Vectors::mcols(gff[grepl("Q0020",S4Vectors::mcols(gff)$ID),])$ID,
           S4Vectors::mcols(gff[grepl("Q0158",S4Vectors::mcols(gff)$ID),])$ID)
rRNAFiles <- convertGeneToChrom(gfffile,
                                fafile,
                                c(RDN25,RDN18,RDN58,RNDmt),
                                "rRNA",
                                saveTrimmedFile = TRUE,
                                removeDuplicateSeqs = TRUE)
# Because annotation in yeast is also extended to the ETS sequences mask
# all rRNA_gene types
rRNA_depFiles <- convertGeneTypeToChrom(gfffile,
                                fafile,
                                "rRNA_gene",
                                "rRNA_depleted",
                                saveTrimmedFile = TRUE)
# Remove specific tRNA genes and append to fasta/gff
tRNAFiles <- converttRNAscanToChrom(rRNA_depFiles[["trim_gfffile"]],
                                    fafile,
                                    tRNAscanfile,
                                    "tRNAscan",
                                    saveTrimmedFile = TRUE)
# append all the files
out_fafile <- appendFastaFiles(fafile,
                               rRNAFiles[["convert_fafile"]],
                               tRNAFiles[["convert_fafile"]],
                               ident = "complete")
out_gfffile <- appendGFF(tRNAFiles[["trim_gfffile"]],
                         rRNAFiles[["convert_gfffile"]],
                         tRNAFiles[["convert_gfffile"]],
                         ident = "complete")

# generate bash script
script <- paste0("cat %s*_mask_* > %s\n",
               "dos2unix %s\n",
               "bedtools maskfasta -mc N -fi %s -bed %s -fo %s\n",
               "cp %s %s\n",
               "cp %s %s")
script <- gettextf(script,
                  folder,
       paste0(folder,"sacCer3_R64-2-1_mask.gff3"),
       out_fafile,
       out_fafile,
       paste0(folder,"sacCer3_R64-2-1_mask.gff3"),
       paste0(folder,"sacCer3_R64-2-1_masked.fasta"),
       out_gfffile,
       "sacCer3_R64-2-1_masked.gff3",
       paste0(folder,"sacCer3_R64-2-1_masked.fasta"),
       "sacCer3_R64-2-1_masked.fasta")
message("# Run this in command line:\n\n",
        script)
```

## E. coli

```{r eco,eval=FALSE}
folder <- "genomes/Eco/"
gfffile <- paste0(folder,"eco_MG1655.gff3")
fafile <- paste0(folder,"eco_MG1655.fasta")
tRNAscanfile <- paste0(folder,"eschColi_K_12_MG1655-tRNAs.ss.sort")
matchFastaToGff(gfffile,fafile)
# Remove rRNA genes and append to fasta/gff
rRNAFiles <- convertGeneTypeToChrom(gfffile,
                                    fafile,
                                    "rRNA",
                                    "rRNA",
                                    saveTrimmedFile = TRUE,
                                    removeDuplicateSeqs = TRUE)
# Remove specific tRNA genes and append to fasta/gff
tRNAFiles <- converttRNAscanToChrom(rRNAFiles[["trim_gfffile"]],
                                    fafile,
                                    tRNAscanfile,
                                    "tRNAscan",
                                    saveTrimmedFile = TRUE)
# append all the files
out_fafile <- appendFastaFiles(fafile,
                               rRNAFiles[["convert_fafile"]],
                               tRNAFiles[["convert_fafile"]],
                               ident = "complete")
out_gfffile <- appendGFF(tRNAFiles[["trim_gfffile"]],
                         rRNAFiles[["convert_gfffile"]],
                         tRNAFiles[["convert_gfffile"]],
                         ident = "complete")

# generate bash script
script <- paste0("cat %s*_mask_* > %s\n",
               "dos2unix %s\n",
               "bedtools maskfasta -mc N -fi %s -bed %s -fo %s\n",
               "cp %s %s\n",
               "cp %s %s")
script <- gettextf(script,
                  folder,
       paste0(folder,"eco_MG1655_mask.gff3"),
       out_fafile,
       out_fafile,
       paste0(folder,"eco_MG1655_mask.gff3"),
       paste0(folder,"eco_MG1655_masked.fasta"),
       out_gfffile,
       "eco_MG1655_masked.gff3",
       paste0(folder,"eco_MG1655_masked.fasta"),
       "eco_MG1655_masked.fasta")
message("# Run this in command line:\n\n",
        script)
```

## H. sapiens

```{r GRCh38,eval=FALSE}
folder <- "genomes/GRCh38/"
gfffile <- paste0(folder,"GRCh38_latest_genomic.gff")
gfffileS <- paste0(folder,"GRCh38_latest_genomic_s.gff")
fafile <- paste0(folder,"GRCh38_latest_genomic.fasta")
tRNAscanfile <- paste0(folder,"hg38-tRNAs.ss.sort")
matchFastaToGff(gfffile,fafile)
# get rRNA genes and save as fasta/gff
rRNAFiles <- convertGeneTypeToChrom(gfffile,
                                    fafile,
                                    "rRNA",
                                    "rRNA",
                                    saveTrimmedFile = TRUE,
                                    removeDuplicateSeqs = TRUE)
# Remove specific tRNA genes and append to fasta/gff
tRNAFiles <- converttRNAscanToChrom(rRNAFiles[["trim_gfffile"]],
                                    fafile,
                                    tRNAscanfile,
                                    "tRNAscan",
                                    saveTrimmedFile = TRUE)
# append all the files
out_fafile <- appendFastaFiles(fafile,
                 rRNAFiles[["convert_fafile"]],
                 tRNAFiles[["convert_fafile"]],
                 ident = "complete")
out_gfffile <- appendGFF(tRNAFiles[["trim_gfffile"]],
          rRNAFiles[["convert_gfffile"]],
          tRNAFiles[["convert_gfffile"]],
                 ident = "complete")

# generate bash script
script <- paste0("cat %s*_mask_* > %s\n",
               "dos2unix %s\n",
               "bedtools maskfasta -mc N -fi %s -bed %s -fo %s\n",
               "cp %s %s\n",
               "cp %s %s")
script <- gettextf(script,
                  folder,
       paste0(folder,"GRCh38_latest_genomic_mask.gff3"),
       out_fafile,
       out_fafile,
       paste0(folder,"GRCh38_latest_genomic_mask.gff3"),
       paste0(folder,"GRCh38_latest_genomic_masked.fasta"),
       out_gfffile,
       "GRCh38_latest_genomic_masked.gff3",
       paste0(folder,"GRCh38_latest_genomic_masked.fasta"),
       "GRCh38_latest_genomic_masked.fasta")
message("# Run this in command line:\n\n",
        script)
```

# Session info

```{r sessionInfo}
sessionInfo()
```

# References