---
title: "RNAmod: pre-mapping preparation of genome and annotation"
author: "Felix G.M. Ernst"
date: "`r Sys.Date()`"
output: 
    rmarkdown::html_document:
        toc: true
        toc_float: true
        df_print: paged
vignette: >
  %\VignetteIndexEntry{RNAmod - pre-mapping preparation of genome and annotation}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

```{r tRNA}
library(RNAmod)

matchFastaToGff <- function(gfffile, fafile){
  gff <- rtracklayer::import.gff3(gfffile)
  fsa <- Biostrings::readDNAStringSet(fafile)
  names(fsa) <- unique(as.character(rtracklayer::chrom(gff)))
  Biostrings::writeXStringSet(fsa, fafile)
}
```

# Converting tRNA and rRNA genes to seperate chromosomes

## sacCer3

```{r sacCer3}
folder <- "genomes/sacCer3/"
gfffile <- paste0(folder,"sacCer3_R64-2-1.gff3")
fafile <- paste0(folder,"sacCer3_R64-2-1.fasta")
tRNAscanfile <- paste0(folder,"sacCer3-tRNAs.ss.sort")
matchFastaToGff(gfffile,fafile)

# get specific rRNA genes and save as fasta/gff
gff <- rtracklayer::import.gff3(gfffile)
RDN25 <- S4Vectors::mcols(gff[grepl("RDN25",S4Vectors::mcols(gff)$ID),])$ID
RDN18 <- S4Vectors::mcols(gff[grepl("RDN18",S4Vectors::mcols(gff)$ID),])$ID
RDN58 <- S4Vectors::mcols(gff[grepl("RDN58",S4Vectors::mcols(gff)$ID),])$ID
RNDmt <- c(S4Vectors::mcols(gff[grepl("Q0020",S4Vectors::mcols(gff)$ID),])$ID,
           S4Vectors::mcols(gff[grepl("Q0158",S4Vectors::mcols(gff)$ID),])$ID)
rRNAFiles <- convertGeneToChrom(gfffile,
                                fafile,
                                c(RDN25,RDN18,RDN58,RNDmt),
                                "rRNA")
# get specific mRNA genes and other transcripts and save as fasta/gff
miscFiles <- convertGeneTypeToChrom(gfffile,
                                    fafile,
                                    c("mRNA",
                                      "ncRNA_gene",
                                      "snoRNA_gene",
                                      "snRNA_gene"),
                                    "misc")
# Remove specific tRNA genes and append to fasta/gff
tRNAFiles <- converttRNAscanToChrom(gfffile,
                                    fafile,
                                    tRNAscanfile,
                                    "tRNAscan")

# append all the files
appendFastaFiles(rRNAFiles[["convert_fafile"]],
                 miscFiles[["convert_fafile"]],
                 tRNAFiles[["convert_fafile"]],
                 ident = "complete")

appendGFF(rRNAFiles[["convert_gfffile"]],
          miscFiles[["convert_gfffile"]],
          tRNAFiles[["convert_gfffile"]],
                 ident = "complete")
```

## E. coli

```{r eco}
folder <- "genomes/Eco/"
gfffile <- paste0(folder,"eco_MG1655.gff3")
fafile <- paste0(folder,"eco_MG1655.fasta")
tRNAscanfile <- paste0(folder,"eschColi_K_12_MG1655-tRNAs.ss.sort")
matchFastaToGff(gfffile,fafile)

# Remove rRNA genes and append to fasta/gff
rRNAFiles <- convertGeneTypeToChrom(gfffile,
                                    fafile,
                                    "rRNA",
                                    "rRNA",
                                    appendToOriginal = TRUE)

# Manually remove any trace of rRNA genes from gff file
gff <- rtracklayer::import.gff3(rRNAFiles[["in_gfffile"]])
gff_rRNA <- gff[S4Vectors::mcols(gff)$type == "rRNA",]
gff_rRNA <- IRanges::subsetByOverlaps(gff, 
                                      gff_rRNA,
                                      minoverlap = 0,
                                      type = "equal",
                                      invert = TRUE)
rtracklayer::export.gff3(gff_rRNA,rRNAFiles[["out_gfffile"]])
rRNAFiles[["out_gfffile"]] <- appendGFF(rRNAFiles[["out_gfffile"]],
                                        rRNAFiles[["convert_gfffile"]],
                                        ident = "clean",
                                        msg = "Saved appended gff file: ")

# Remove specific tRNA genes and append to fasta/gff
tRNAFiles <- converttRNAscanToChrom(rRNAFiles[["out_gfffile"]],
                                    rRNAFiles[["out_fafile"]],
                                    tRNAscanfile,
                                    "tRNAscan",
                                    appendToOriginal = TRUE)

# generate script lines for masking nucleotides and copying files to wd folder
script <- paste0("cat %s/*_mask_* > %s\n",
               "dos2unix %s\n",
               "bedtools maskfasta -mc N -fi %s -bed %s -fo %s\n",
               "cp %s %s\n",
               "cp %s %s")
script <- gettextf(script,
                   dirname(tRNAFiles[["out_fafile"]]),
       paste0(dirname(tRNAFiles[["out_fafile"]]),"/eco-MG1655_mask.gff3"),
       tRNAFiles[["out_fafile"]],
       tRNAFiles[["out_fafile"]],
       paste0(dirname(tRNAFiles[["out_fafile"]]),"/eco-MG1655_mask.gff3"),
       paste0(dirname(tRNAFiles[["out_fafile"]]),"/eco-MG1655_masked.fasta"),
       tRNAFiles[["out_gfffile"]],
       "eco-MG1655_masked.gff3",
       paste0(dirname(tRNAFiles[["out_fafile"]]),"/eco-MG1655_masked.fasta"),
       "eco-MG1655_masked.fasta")
message("# Run this in command line:\n\n",
        script)
```



```{r GRCh38}
folder <- "genomes/GRCh38/"
gfffile <- paste0(folder,"GRCh38_latest_genomic.gff")
gfffileS <- paste0(folder,"GRCh38_latest_genomic_s.gff")
fafile <- paste0(folder,"GRCh38_latest_genomic.fasta")
tRNAscanfile <- paste0(folder,"hg38-tRNAs.ss.sort")
matchFastaToGff(gfffile,fafile)

# get rRNA genes and save as fasta/gff
rRNAFiles <- convertGeneTypeToChrom(gfffile,
                                fafile,
                                "rRNA",
                                "rRNA")

# get specific mRNA genes and other transcripts and save as fasta/gff
miscFiles <- convertGeneTypeToChrom(gfffile,
                                    fafile,
                                    c("mRNA",
                                      "ncRNA",
                                      "snoRNA",
                                      "snRNA",
                                      "RNase_P_RNA"),
                                    "misc")
# Remove specific tRNA genes and append to fasta/gff
tRNAFiles <- converttRNAscanToChrom(gfffile,
                                    fafile,
                                    tRNAscanfile,
                                    "tRNAscan")

# append all the files
appendFastaFiles(rRNAFiles[["convert_fafile"]],
                 miscFiles[["convert_fafile"]],
                 tRNAFiles[["convert_fafile"]],
                 ident = "complete")
appendGFF(rRNAFiles[["convert_gfffile"]],
          miscFiles[["convert_gfffile"]],
          tRNAFiles[["convert_gfffile"]],
                 ident = "complete")
```

# Session info

```{r sessionInfo}
sessionInfo()
```