---
title: "Vignette Title"
author: "Vignette Author"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```






create RNAmod
```{r RNAmodR, eval=FALSE}
library(RNAmodR)
library(SummarizedExperiment)
mod <- RNAmodR("test",
              "experiment_layout.csv",
              "sacCer3_R64-2-1_masked.gff3",
              "sacCer3_R64-2-1_masked.fasta")
modifications <- c("m7G","D","m3C")
no <- unique(getExperimentData(mod)$ExperimentNo)
for(i in seq_along(no)){
  parseForModifications(mod,no[i],modifications)
  se <- getSummarizedExperiment(mod,no[i],modifications)
  transcripts <- unique(c(names(assays(se)$m7G[assays(se)$m7G > 0,]),
                          names(assays(se)$m3C[assays(se)$m3C > 0,]),
                          names(assays(se)$D[assays(se)$D > 0,])))
  saveModPlot(mod,se,modifications,transcripts)
}
gc()

param <- BiocParallel::bpparam()
param$workers <- 3
BiocParallel::register(param)

no <- unique(getExperimentData(mod_hsa)$ExperimentNo)
for(i in seq_along(no)){
  parseForModifications(mod_hsa,no[i],modifications)
  se <- getSummarizedExperiment(mod_hsa,no[i],modifications)
  transcripts <- unique(c(names(assays(se)$m7G[assays(se)$m7G > 0,]),
                          names(assays(se)$m3C[assays(se)$m3C > 0,]),
                          names(assays(se)$D[assays(se)$D > 0,])))
  saveModPlot(mod_hsa,se,modifications,transcripts)
}
param$workers <- 6
BiocParallel::register(param)

param <- BiocParallel::bpparam()
param$workers <- 3
BiocParallel::register(param)

mod_hsa <- RNAmodR("test",
              "experiment_layout_hsa.csv",
              "GRCh38_latest_genomic_masked.gff3",
              "GRCh38_latest_genomic_masked.fasta")
no <- unique(getExperimentData(mod_hsa)$ExperimentNo)
lapply(no,function(x){
  parseForModifications(mod_hsa,x,modifications)
  se <- getSummarizedExperiment(mod_hsa,x,modifications)
  transcripts <- unique(c(names(assays(se)$m7G[assays(se)$m7G > 0,]),
                          names(assays(se)$m3C[assays(se)$m3C > 0,]),
                          names(assays(se)$D[assays(se)$D > 0,])))
  saveModPlot(mod_hsa,se,modifications,transcripts)
})

se1 <- getSummarizedExperiment(mod,1,modifications)
se2 <- getSummarizedExperiment(mod,2,modifications)
se3 <- getSummarizedExperiment(mod,3,modifications)

tRNA <- rownames(se1)[grepl("t", rownames(se1), fixed = TRUE)]
rRNA <- c("RDN18-1","RDN25-1")
m7Gtranscripts <- names(assays(se)$m7G[assays(se)$m7G > 0,])

saveModPlot(mod,se1,modifications,c(rRNA))
saveModPlot(mod,se2,modifications,c(rRNA))
saveModPlot(mod,se3,modifications,c(rRNA))
saveModPlot(mod,se1,modifications,c(tRNA),filetype = "png")
saveModPlot(mod,se2,modifications,c(tRNA),filetype = "png")
saveModPlot(mod,se3,modifications,c(tRNA),filetype = "png")



```

# plotting

```{r RNAmodR, eval=FALSE}
grl <- GenomicRanges::GRangesList(wt = getGffResult(mod,1,modifications),
                                  Bud23 = getGffResult(mod,2,modifications),
                                  Trm8 = getGffResult(mod,3,modifications))
genes <- as.character(grl[[1]]$Parent)
genes <- genes[grepl("^t", genes)]

ses <- list(wt = getSummarizedExperiment(mod,1,modifications),
                                  Bud23 = getSummarizedExperiment(mod,2,modifications),
                                  Trm8 = getSummarizedExperiment(mod,3,modifications))

transcripts <- unique(unlist(lapply(ses, function(se){
    c(names(SummarizedExperiment::assays(se)$m7G[SummarizedExperiment::assays(se)$m7G > 0,]),
    names(SummarizedExperiment::assays(se)$m3C[SummarizedExperiment::assays(se)$m3C > 0,]),
    names(SummarizedExperiment::assays(se)$D[SummarizedExperiment::assays(se)$D > 0,]))
})))
tRNAgenes <- transcripts[grepl("^t", transcripts)]


```


# Session info

```{r sessionInfo}
sessionInfo()
```

# References