
context("End5SequenceData")
test_that("End5SequenceData:",{
  # SequenceData using CoverageSequenceData as test case
  annotation <- system.file("extdata","example1.gff3",package = "RNAmodR.Data")
  sequences <- system.file("extdata","example1.fasta",package = "RNAmodR.Data")
  files <- c(control = system.file("extdata","example_wt_1.bam",
                                   package = "RNAmodR.Data"),
             treated = system.file("extdata","example_wt_2.bam",
                                   package = "RNAmodR.Data"))
  #
  e5sd <- End5SequenceData(files, annotation = annotation,
                              sequences = sequences)
  expect_s4_class(e5sd,"End5SequenceData")
  expect_named(e5sd,c("1","2"))
  expect_s4_class(colnames(e5sd),"CharacterList")
  expect_length(colnames(e5sd),2)
  expect_equal(lengths(colnames(e5sd)),c(2,2))
  expect_equal(colnames(e5sd)[[1]],colnames(e5sd)[[2]])
  expect_equal(colnames(e5sd)[[1]],c("end5.control.1","end5.treated.1"))
  actual <- aggregate(e5sd)
  expect_s4_class(actual,"CompressedSplitDataFrameList")
  expect_s4_class(actual,"SplitDataFrameList")
  expect_equal(length(actual),2)
  expect_length(colnames(actual),2)
  expect_equal(lengths(colnames(actual)),c(4,4))
  expect_equal(colnames(actual)[[1]],colnames(actual)[[2]])
  expect_equal(colnames(actual)[[1]],c("means.control","means.treated",
                                       "sds.control","sds.treated"))
  expect_s4_class(seqinfo(e5sd),"Seqinfo")
  expect_equal(length(seqinfo(e5sd)),11)
  actual <- aggregate(e5sd, condition = "Control")
  expect_s4_class(actual,"CompressedSplitDataFrameList")
  expect_s4_class(actual,"SplitDataFrameList")
  expect_equal(length(actual),2)
  expect_length(colnames(actual),2)
  expect_equal(lengths(colnames(actual)),c(2,2))
  expect_equal(colnames(actual)[[1]],colnames(actual)[[2]])
  expect_equal(colnames(actual)[[1]],c("means.control","sds.control"))
  expect_s4_class(seqinfo(e5sd),"Seqinfo")
  expect_equal(length(seqinfo(e5sd)),11)
  actual <- aggregate(e5sd, condition = "Treated")
  expect_s4_class(actual,"CompressedSplitDataFrameList")
  expect_s4_class(actual,"SplitDataFrameList")
  expect_equal(length(actual),2)
  expect_length(colnames(actual),2)
  expect_equal(lengths(colnames(actual)),c(2,2))
  expect_equal(colnames(actual)[[1]],colnames(actual)[[2]])
  expect_equal(colnames(actual)[[1]],c("means.treated","sds.treated"))
  expect_s4_class(seqinfo(e5sd),"Seqinfo")
  expect_equal(length(seqinfo(e5sd)),11)
  #
  e3sd <- End3SequenceData(files, annotation = annotation,
                          sequences = sequences)
  expect_s4_class(e3sd,"End3SequenceData")
  expect_named(e3sd,c("1","2"))
  expect_s4_class(colnames(e3sd),"CharacterList")
  expect_length(colnames(e3sd),2)
  expect_equal(lengths(colnames(e3sd)),c(2,2))
  expect_equal(colnames(e3sd)[[1]],colnames(e3sd)[[2]])
  expect_equal(colnames(e3sd)[[1]],c("end3.control.1","end3.treated.1"))
  actual <- aggregate(e3sd)
  expect_s4_class(actual,"CompressedSplitDataFrameList")
  expect_s4_class(actual,"SplitDataFrameList")
  expect_equal(length(actual),2)
  expect_length(colnames(actual),2)
  expect_equal(lengths(colnames(actual)),c(4,4))
  expect_equal(colnames(actual)[[1]],colnames(actual)[[2]])
  expect_equal(colnames(actual)[[1]],c("means.control","means.treated",
                                       "sds.control","sds.treated"))
  expect_s4_class(seqinfo(e3sd),"Seqinfo")
  expect_equal(length(seqinfo(e3sd)),11)
  actual <- aggregate(e3sd, condition = "Control")
  expect_s4_class(actual,"CompressedSplitDataFrameList")
  expect_s4_class(actual,"SplitDataFrameList")
  expect_equal(length(actual),2)
  expect_length(colnames(actual),2)
  expect_equal(lengths(colnames(actual)),c(2,2))
  expect_equal(colnames(actual)[[1]],colnames(actual)[[2]])
  expect_equal(colnames(actual)[[1]],c("means.control","sds.control"))
  expect_s4_class(seqinfo(e3sd),"Seqinfo")
  expect_equal(length(seqinfo(e3sd)),11)
  actual <- aggregate(e3sd, condition = "Treated")
  expect_s4_class(actual,"CompressedSplitDataFrameList")
  expect_s4_class(actual,"SplitDataFrameList")
  expect_equal(length(actual),2)
  expect_length(colnames(actual),2)
  expect_equal(lengths(colnames(actual)),c(2,2))
  expect_equal(colnames(actual)[[1]],colnames(actual)[[2]])
  expect_equal(colnames(actual)[[1]],c("means.treated","sds.treated"))
  expect_s4_class(seqinfo(e3sd),"Seqinfo")
  expect_equal(length(seqinfo(e3sd)),11)
  #
  esd <- EndSequenceData(files, annotation = annotation,
                          sequences = sequences)
  expect_s4_class(esd,"EndSequenceData")
  expect_named(esd,c("1","2"))
  expect_s4_class(colnames(esd),"CharacterList")
  expect_length(colnames(esd),2)
  expect_equal(lengths(colnames(esd)),c(2,2))
  expect_equal(colnames(esd)[[1]],colnames(esd)[[2]])
  expect_equal(colnames(esd)[[1]],c("end.control.1","end.treated.1"))
  actual <- aggregate(esd)
  expect_s4_class(actual,"CompressedSplitDataFrameList")
  expect_s4_class(actual,"SplitDataFrameList")
  expect_equal(length(actual),2)
  expect_length(colnames(actual),2)
  expect_equal(lengths(colnames(actual)),c(4,4))
  expect_equal(colnames(actual)[[1]],colnames(actual)[[2]])
  expect_equal(colnames(actual)[[1]],c("means.control","means.treated",
                                       "sds.control","sds.treated"))
  expect_s4_class(seqinfo(esd),"Seqinfo")
  expect_equal(length(seqinfo(esd)),11)
  actual <- aggregate(esd, condition = "Control")
  expect_s4_class(actual,"CompressedSplitDataFrameList")
  expect_s4_class(actual,"SplitDataFrameList")
  expect_equal(length(actual),2)
  expect_length(colnames(actual),2)
  expect_equal(lengths(colnames(actual)),c(2,2))
  expect_equal(colnames(actual)[[1]],colnames(actual)[[2]])
  expect_equal(colnames(actual)[[1]],c("means.control","sds.control"))
  expect_s4_class(seqinfo(esd),"Seqinfo")
  expect_equal(length(seqinfo(esd)),11)
  actual <- aggregate(esd, condition = "Treated")
  expect_s4_class(actual,"CompressedSplitDataFrameList")
  expect_s4_class(actual,"SplitDataFrameList")
  expect_equal(length(actual),2)
  expect_length(colnames(actual),2)
  expect_equal(lengths(colnames(actual)),c(2,2))
  expect_equal(colnames(actual)[[1]],colnames(actual)[[2]])
  expect_equal(colnames(actual)[[1]],c("means.treated","sds.treated"))
  expect_s4_class(seqinfo(esd),"Seqinfo")
  expect_equal(length(seqinfo(esd)),11)
})
