<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Comparison of Samples — compare • RNAmodR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Comparison of Samples — compare" />
<meta property="og:description" content="To compare data of different samples, a
ModifierSet can be used. To select the data
alongside the transcripts and their positions a
GRanges or a
GRangesList needs to be provided.
In case of a GRanges object, the parent column must match the
transcript names as defined by the out put of ranges(x), whereas in
case of a GRangesList the element names must match the transcript
names." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.15</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/FelixErnst/RNAmodR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Comparison of Samples</h1>
    <small class="dont-index">Source: <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/ModifierSet-comparison.R'><code>R/ModifierSet-comparison.R</code></a></small>
    <div class="hidden name"><code>compareByCoord.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>To compare data of different samples, a
<code><a href='ModifierSet-class.html'>ModifierSet</a></code> can be used. To select the data
alongside the transcripts and their positions a
<code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></code> or a
<code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRangesList</a></code> needs to be provided.
In case of a <code>GRanges</code> object, the parent column must match the
transcript names as defined by the out put of <code>ranges(x)</code>, whereas in
case of a <code>GRangesList</code> the element names must match the transcript
names.</p>
    </div>

    <pre class="usage"><span class='fu'>compare</span>(<span class='no'>x</span>, <span class='no'>name</span>, <span class='kw'>pos</span> <span class='kw'>=</span> <span class='fl'>1L</span>, <span class='no'>...</span>)

<span class='fu'>compareByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='fu'>plotCompare</span>(<span class='no'>x</span>, <span class='no'>name</span>, <span class='kw'>pos</span> <span class='kw'>=</span> <span class='fl'>1L</span>, <span class='no'>normalize</span>, <span class='no'>...</span>)

<span class='fu'>plotCompareByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>normalize</span>, <span class='no'>...</span>)

<span class='co'># S4 method for ModifierSet</span>
<span class='fu'>compare</span>(<span class='no'>x</span>, <span class='no'>name</span>, <span class='kw'>pos</span> <span class='kw'>=</span> <span class='fl'>1L</span>, <span class='no'>normalize</span>, <span class='no'>...</span>)

<span class='co'># S4 method for ModifierSet,GRanges</span>
<span class='fu'>compareByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>normalize</span>, <span class='no'>...</span>)

<span class='co'># S4 method for ModifierSet,GRangesList</span>
<span class='fu'>compareByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>normalize</span>, <span class='no'>...</span>)

<span class='co'># S4 method for ModifierSet</span>
<span class='fu'>plotCompare</span>(<span class='no'>x</span>, <span class='no'>name</span>, <span class='kw'>pos</span> <span class='kw'>=</span> <span class='fl'>1L</span>, <span class='no'>normalize</span>, <span class='no'>...</span>)

<span class='co'># S4 method for ModifierSet,GRanges</span>
<span class='fu'>plotCompareByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>normalize</span>, <span class='no'>...</span>)

<span class='co'># S4 method for ModifierSet,GRangesList</span>
<span class='fu'>plotCompareByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>normalize</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>a <code>Modifier</code> or <code>ModifierSet</code> object.</p></td>
    </tr>
    <tr>
      <th>name</th>
      <td><p>Only for <code>compare</code>: the transcript name</p></td>
    </tr>
    <tr>
      <th>pos</th>
      <td><p>Only for <code>compare</code>: pos for comparison</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>optional parameters:</p><ul>
<li><p><code>alias</code> a data.frame with two columns, <code>tx_id</code> and 
<code>name</code>, to convert transcipt ids to another identifier</p></li>
<li><p><code>name</code> Limit results to one specific gene or transcript</p></li>
<li><p><code>sequenceData</code> TRUE or FALSE? Should the aggregate of 
sequenceData be used for the comparison instead of the aggregate data if each
<code>Modifier</code> element? (default: <code>sequenceData = FALSE</code>)</p></li>
<li><p><code>compareType</code> a valid score type to use for the comparison. If
<code>sequenceData = FALSE</code> this defaults to <code><a href='Modifier-functions.html'>mainScore(x)</a></code>, whereas
if <code>sequenceData = TRUE</code> all columns will be used by setting 
<code>allTypes = TRUE</code>.</p></li>
<li><p><code>allTypes</code> TRUE or FALSE? Should all available score be 
compared? (default: <code>allTypes = sequenceData</code>)</p></li>
<li><p>... passed on to <code><a href='subsetByCoord.html'>subsetByCoord</a></code></p></li>
</ul></td>
    </tr>
    <tr>
      <th>coord</th>
      <td><p>coordinates of position to subset to. Either a <code>GRanges</code> or
a <code>GRangesList</code> object. For both types the 'Parent' column is expected
to match the transcript name. The <code>GRangesList</code> object is
unlisted and only non duplicated entries are retained.</p></td>
    </tr>
    <tr>
      <th>normalize</th>
      <td><p>either a single logical or character value. If it is a 
character, it must match one of the names in the <code>ModifierSet</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>compareByCoord</code> returns a
  <code><a href='https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html'>DataFrame</a></code> and
  <code>plotCompareByCoord</code> returns a <code>ggplot</code> object, which can be
  modified further. The <code>DataFrame</code> contains columns per sample as well
  as the columns <code>names</code>, <code>positions</code> and <code>mod</code> incorporated
  from the <code>coord</code> input. If <code>coord</code> contains a column
  <code>Activity</code> this is included in the results as well.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>msi</span>,<span class='kw'>package</span><span class='kw'>=</span><span class='st'>"RNAmodR"</span>)
<span class='co'># constructing a GRanges obejct to mark positive positions</span>
<span class='no'>mod</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='modify.html'>modifications</a></span>(<span class='no'>msi</span>)
<span class='no'>coord</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/unlist.html'>unlist</a></span>(<span class='no'>mod</span>))
<span class='no'>coord</span>$<span class='no'>score</span> <span class='kw'>&lt;-</span> <span class='kw'>NULL</span>
<span class='no'>coord</span>$<span class='no'>sd</span> <span class='kw'>&lt;-</span> <span class='kw'>NULL</span>
<span class='co'># return a DataFrame</span>
<span class='fu'>compareByCoord</span>(<span class='no'>msi</span>,<span class='no'>coord</span>)</div><div class='output co'>#&gt; DataFrame with 6 rows and 6 columns
#&gt;   SampleSet1 SampleSet2 SampleSet3    names positions         mod
#&gt;    &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt; &lt;factor&gt;  &lt;factor&gt; &lt;character&gt;
#&gt; 1   0.900932   0.998134   0.953651       2         34           I
#&gt; 2   0.899622   0.856241   0.976928       4         35           I
#&gt; 3   0.984035   0.992012   0.993128       6         34           I
#&gt; 4   0.934553   0.942905   0.943773       7         67           I
#&gt; 5   0.709758   0.766484   0.681451       9         7            I
#&gt; 6   0.874027   0.971474   0.954782       11        35           I</div><div class='input'><span class='co'># plot the comparison as a heatmap</span>
<span class='fu'>plotCompareByCoord</span>(<span class='no'>msi</span>,<span class='no'>coord</span>)</div><div class='img'><img src='compareByCoord-1.png' alt='' width='700' height='433' /></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


