<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Comparison of Samples — compare • RNAmodR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Comparison of Samples — compare"><meta property="og:description" content="To compare data of different samples, a
ModifierSet can be used. To select the data
alongside the transcripts and their positions a
GRanges or a
GRangesList needs to be provided.
In case of a GRanges object, the parent column must match the
transcript names as defined by the out put of ranges(x), whereas in
case of a GRangesList the element names must match the transcript
names."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.17.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/FelixErnst/RNAmodR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Comparison of Samples</h1>
    <small class="dont-index">Source: <a href="https://github.com/FelixErnst/RNAmodR/blob/HEAD/R/AllGenerics.R" class="external-link"><code>R/AllGenerics.R</code></a>, <a href="https://github.com/FelixErnst/RNAmodR/blob/HEAD/R/ModifierSet-comparison.R" class="external-link"><code>R/ModifierSet-comparison.R</code></a></small>
    <div class="hidden name"><code>compareByCoord.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>To compare data of different samples, a
<code><a href="ModifierSet-class.html">ModifierSet</a></code> can be used. To select the data
alongside the transcripts and their positions a
<code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> or a
<code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRangesList</a></code> needs to be provided.
In case of a <code>GRanges</code> object, the parent column must match the
transcript names as defined by the out put of <code>ranges(x)</code>, whereas in
case of a <code>GRangesList</code> the element names must match the transcript
names.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">compare</span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, pos <span class="op">=</span> <span class="fl">1L</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">compareByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotCompare</span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, pos <span class="op">=</span> <span class="fl">1L</span>, <span class="va">normalize</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotCompareByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">normalize</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for ModifierSet</span></span>
<span><span class="fu">compare</span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, pos <span class="op">=</span> <span class="fl">1L</span>, <span class="va">normalize</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for ModifierSet,GRanges</span></span>
<span><span class="fu">compareByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">normalize</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for ModifierSet,GRangesList</span></span>
<span><span class="fu">compareByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">normalize</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for ModifierSet</span></span>
<span><span class="fu">plotCompare</span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, pos <span class="op">=</span> <span class="fl">1L</span>, <span class="va">normalize</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for ModifierSet,GRanges</span></span>
<span><span class="fu">plotCompareByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">normalize</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for ModifierSet,GRangesList</span></span>
<span><span class="fu">plotCompareByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">normalize</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p>a <code>Modifier</code> or <code>ModifierSet</code> object.</p></dd>


<dt>name</dt>
<dd><p>Only for <code>compare</code>: the transcript name</p></dd>


<dt>pos</dt>
<dd><p>Only for <code>compare</code>: pos for comparison</p></dd>


<dt>...</dt>
<dd><p>optional parameters:</p><ul><li><p><code>alias</code> a data.frame with two columns, <code>tx_id</code> and 
<code>name</code>, to convert transcipt ids to another identifier</p></li>
<li><p><code>name</code> Limit results to one specific gene or transcript</p></li>
<li><p><code>sequenceData</code> TRUE or FALSE? Should the aggregate of 
sequenceData be used for the comparison instead of the aggregate data if each
<code>Modifier</code> element? (default: <code>sequenceData = FALSE</code>)</p></li>
<li><p><code>compareType</code> a valid score type to use for the comparison. If
<code>sequenceData = FALSE</code> this defaults to <code>mainScore(x)</code>, whereas
if <code>sequenceData = TRUE</code> all columns will be used by setting 
<code>allTypes = TRUE</code>.</p></li>
<li><p><code>allTypes</code> TRUE or FALSE? Should all available score be 
compared? (default: <code>allTypes = sequenceData</code>)</p></li>
<li><p>... passed on to <code><a href="subsetByCoord.html">subsetByCoord</a></code></p></li>
</ul></dd>


<dt>coord</dt>
<dd><p>coordinates of position to subset to. Either a <code>GRanges</code> or
a <code>GRangesList</code> object. For both types the 'Parent' column is expected
to match the transcript name. The <code>GRangesList</code> object is
unlisted and only non duplicated entries are retained.</p></dd>


<dt>normalize</dt>
<dd><p>either a single logical or character value. If it is a 
character, it must match one of the names in the <code>ModifierSet</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p><code>compareByCoord</code> returns a</p>
<p></p>
<p><code><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></code> and</p>
<p></p>
<p><code>plotCompareByCoord</code> returns a <code>ggplot</code> object, which can be
  modified further. The <code>DataFrame</code> contains columns per sample as well
  as the columns <code>names</code>, <code>positions</code> and <code>mod</code> incorporated
  from the <code>coord</code> input. If <code>coord</code> contains a column</p>
<p></p>
<p><code>Activity</code> this is included in the results as well.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">msi</span>,package<span class="op">=</span><span class="st">"RNAmodR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># constructing a GRanges obejct to mark positive positions</span></span></span>
<span class="r-in"><span><span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="modify.html">modifications</a></span><span class="op">(</span><span class="va">msi</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">coord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">mod</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">coord</span><span class="op">$</span><span class="va">score</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></span>
<span class="r-in"><span><span class="va">coord</span><span class="op">$</span><span class="va">sd</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></span>
<span class="r-in"><span><span class="co"># return a DataFrame</span></span></span>
<span class="r-in"><span><span class="fu">compareByCoord</span><span class="op">(</span><span class="va">msi</span>,<span class="va">coord</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 6 rows and 6 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   SampleSet1 SampleSet2 SampleSet3    names positions         mod</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt; &lt;factor&gt;  &lt;factor&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   0.900932   0.998134   0.953651       2         34           I</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   0.899622   0.856241   0.976928       4         35           I</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   0.984035   0.992012   0.993128       6         34           I</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   0.934553   0.942905   0.943773       7         67           I</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   0.709758   0.766484   0.681451       9         7            I</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6   0.874027   0.971474   0.954782       11        35           I</span>
<span class="r-in"><span><span class="co"># plot the comparison as a heatmap</span></span></span>
<span class="r-in"><span><span class="fu">plotCompareByCoord</span><span class="op">(</span><span class="va">msi</span>,<span class="va">coord</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Direct call of 'as.data.frame.numeric()' is deprecated.  Use 'as.data.frame.vector()' or 'as.data.frame()' instead</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Direct call of 'as.data.frame.numeric()' is deprecated.  Use 'as.data.frame.vector()' or 'as.data.frame()' instead</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Direct call of 'as.data.frame.numeric()' is deprecated.  Use 'as.data.frame.vector()' or 'as.data.frame()' instead</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Direct call of 'as.data.frame.factor()' is deprecated.  Use 'as.data.frame.vector()' or 'as.data.frame()' instead</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Direct call of 'as.data.frame.factor()' is deprecated.  Use 'as.data.frame.vector()' or 'as.data.frame()' instead</span>
<span class="r-plt img"><img src="compareByCoord-1.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

