<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>End5SequenceData/End3SequenceData/EndSequenceData — EndSequenceData-class • RNAmodR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="End5SequenceData/End3SequenceData/EndSequenceData — EndSequenceData-class" />
<meta property="og:description" content="The End5SequenceData/End3SequenceData/EndSequenceData
classes aggregate the counts of read ends at each position along a
transcript. End5SequenceData/End3SequenceData classes aggregate
either the 5'-end or 3'-end, the EndSequenceData aggregates both.
All three classes contain one column per data file named using the following
naming convention (end5/end3/end).condition.replicate.
aggregate calculates the mean and sd for samples in the control
and treated condition separatly." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/FelixErnst/RNAmodR">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>End5SequenceData/End3SequenceData/EndSequenceData</h1>
    <small class="dont-index">Source: <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/SequenceData-end-pos.R'><code>R/SequenceData-end-pos.R</code></a></small>
    <div class="hidden name"><code>EndSequenceData-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>End5SequenceData</code>/<code>End3SequenceData</code>/<code>EndSequenceData</code>
classes aggregate the counts of read ends at each position along a
transcript. <code>End5SequenceData</code>/<code>End3SequenceData</code> classes aggregate
either the 5'-end or 3'-end, the <code>EndSequenceData</code> aggregates both.</p>
<p>All three classes contain one column per data file named using the following
naming convention <code>(end5/end3/end).condition.replicate</code>.</p>
<p><code>aggregate</code> calculates the mean and sd for samples in the <code>control</code>
and <code>treated</code> condition separatly.</p>
    </div>

    <pre class="usage"><span class='fu'>End5SequenceDataFrame</span>(
  <span class='no'>df</span>,
  <span class='no'>ranges</span>,
  <span class='no'>sequence</span>,
  <span class='no'>replicate</span>,
  <span class='no'>condition</span>,
  <span class='no'>bamfiles</span>,
  <span class='no'>seqinfo</span>
)

<span class='fu'>End3SequenceDataFrame</span>(
  <span class='no'>df</span>,
  <span class='no'>ranges</span>,
  <span class='no'>sequence</span>,
  <span class='no'>replicate</span>,
  <span class='no'>condition</span>,
  <span class='no'>bamfiles</span>,
  <span class='no'>seqinfo</span>
)

<span class='fu'>EndSequenceDataFrame</span>(
  <span class='no'>df</span>,
  <span class='no'>ranges</span>,
  <span class='no'>sequence</span>,
  <span class='no'>replicate</span>,
  <span class='no'>condition</span>,
  <span class='no'>bamfiles</span>,
  <span class='no'>seqinfo</span>
)

<span class='fu'>End5SequenceData</span>(<span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='fu'>End3SequenceData</span>(<span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='fu'>EndSequenceData</span>(<span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='co'># S4 method for End5SequenceData,BamFileList,GRangesList,XStringSet,ScanBamParam</span>
<span class='fu'><a href='RNAmodR-development.html'>getData</a></span>(<span class='no'>x</span>, <span class='no'>bamfiles</span>, <span class='no'>grl</span>, <span class='no'>sequences</span>, <span class='no'>param</span>, <span class='no'>args</span>)

<span class='co'># S4 method for End3SequenceData,BamFileList,GRangesList,XStringSet,ScanBamParam</span>
<span class='fu'><a href='RNAmodR-development.html'>getData</a></span>(<span class='no'>x</span>, <span class='no'>bamfiles</span>, <span class='no'>grl</span>, <span class='no'>sequences</span>, <span class='no'>param</span>, <span class='no'>args</span>)

<span class='co'># S4 method for EndSequenceData,BamFileList,GRangesList,XStringSet,ScanBamParam</span>
<span class='fu'><a href='RNAmodR-development.html'>getData</a></span>(<span class='no'>x</span>, <span class='no'>bamfiles</span>, <span class='no'>grl</span>, <span class='no'>sequences</span>, <span class='no'>param</span>, <span class='no'>args</span>)

<span class='co'># S4 method for End5SequenceData</span>
<span class='fu'><a href='aggregate.html'>aggregateData</a></span>(<span class='no'>x</span>, <span class='kw'>condition</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"Both"</span>, <span class='st'>"Treated"</span>, <span class='st'>"Control"</span>))

<span class='co'># S4 method for End3SequenceData</span>
<span class='fu'><a href='aggregate.html'>aggregateData</a></span>(<span class='no'>x</span>, <span class='kw'>condition</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"Both"</span>, <span class='st'>"Treated"</span>, <span class='st'>"Control"</span>))

<span class='co'># S4 method for EndSequenceData</span>
<span class='fu'><a href='aggregate.html'>aggregateData</a></span>(<span class='no'>x</span>, <span class='kw'>condition</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"Both"</span>, <span class='st'>"Treated"</span>, <span class='st'>"Control"</span>))

<span class='co'># S4 method for EndSequenceData</span>
<span class='fu'><a href='plotData.html'>getDataTrack</a></span>(<span class='no'>x</span>, <span class='no'>name</span>, <span class='no'>...</span>)

<span class='co'># S4 method for End5SequenceData</span>
<span class='fu'><a href='plotData.html'>getDataTrack</a></span>(<span class='no'>x</span>, <span class='no'>name</span>, <span class='no'>...</span>)

<span class='co'># S4 method for End3SequenceData</span>
<span class='fu'><a href='plotData.html'>getDataTrack</a></span>(<span class='no'>x</span>, <span class='no'>name</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>df, ranges, sequence, replicate</th>
      <td><p>inputs for creating a 
<code>SequenceDataFrame</code>. See 
<code><a href='SequenceDataFrame-class.html'>SequenceDataFrame</a></code>.</p></td>
    </tr>
    <tr>
      <th>condition</th>
      <td><p>For <code><a href='aggregate.html'>aggregate</a></code>: condition for which the data
should be aggregated.</p></td>
    </tr>
    <tr>
      <th>bamfiles, annotation, seqinfo, grl, sequences, param, args, ...</th>
      <td><p>See
<code><a href='SequenceData-class.html'>SequenceData</a></code> and
<code><a href='SequenceData-functions.html'>SequenceData-functions</a></code></p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>a <code>End5SequenceData</code>, <code>End3SequenceData</code> or
<code>EndSequenceData</code> object</p></td>
    </tr>
    <tr>
      <th>name</th>
      <td><p>For <code><a href='plotData.html'>getDataTrack</a></code>: a valid
transcript name. Must be a name of <code>ranges(x).</code></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a <code>End5SequenceData</code>, a <code>End3SequenceData</code> or a
<code>EndSequenceData</code> object</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Construction of a End5SequenceData object</span>
<span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='no'>RNAmodR.Data</span>)
<span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='no'>rtracklayer</span>)
<span class='no'>annotation</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html'>GFF3File</a></span>(<span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/example.man.fasta.html'>RNAmodR.Data.example.man.gff3</a></span>())</div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2020-02-26</span></div><div class='output co'>#&gt; <span class='message'>see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='input'><span class='no'>sequences</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/example.man.fasta.html'>RNAmodR.Data.example.man.fasta</a></span>()</div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2020-02-26</span></div><div class='output co'>#&gt; <span class='message'>see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span></div><div class='output co'>#&gt; <span class='message'>downloading 1 resources</span></div><div class='output co'>#&gt; <span class='message'>retrieving 1 resource</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='input'><span class='no'>files</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='kw'>treated</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.wt.1</a></span>())</div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2020-02-26</span></div><div class='output co'>#&gt; <span class='message'>see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='input'><span class='no'>e5sd</span> <span class='kw'>&lt;-</span> <span class='fu'>End5SequenceData</span>(<span class='no'>files</span>, <span class='kw'>annotation</span> <span class='kw'>=</span> <span class='no'>annotation</span>,
                        <span class='kw'>sequences</span> <span class='kw'>=</span> <span class='no'>sequences</span>)</div><div class='output co'>#&gt; <span class='message'>Import genomic features from the file as a GRanges object ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Prepare the 'metadata' data frame ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Make the TxDb object ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Loading 5'-end position data from BAM files ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


