<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>End5SequenceData/End3SequenceData/EndSequenceData — EndSequenceData-class • RNAmodR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="End5SequenceData/End3SequenceData/EndSequenceData — EndSequenceData-class"><meta property="og:description" content="The End5SequenceData/End3SequenceData/EndSequenceData
classes aggregate the counts of read ends at each position along a
transcript. End5SequenceData/End3SequenceData classes aggregate
either the 5'-end or 3'-end, the EndSequenceData aggregates both.
All three classes contain one column per data file named using the following
naming convention (end5/end3/end).condition.replicate.
aggregate calculates the mean and sd for samples in the control
and treated condition separatly."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.17.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/FelixErnst/RNAmodR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>End5SequenceData/End3SequenceData/EndSequenceData</h1>
    <small class="dont-index">Source: <a href="https://github.com/FelixErnst/RNAmodR/blob/HEAD/R/SequenceData-end-pos.R" class="external-link"><code>R/SequenceData-end-pos.R</code></a></small>
    <div class="hidden name"><code>EndSequenceData-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>End5SequenceData</code>/<code>End3SequenceData</code>/<code>EndSequenceData</code>
classes aggregate the counts of read ends at each position along a
transcript. <code>End5SequenceData</code>/<code>End3SequenceData</code> classes aggregate
either the 5'-end or 3'-end, the <code>EndSequenceData</code> aggregates both.</p>
<p>All three classes contain one column per data file named using the following
naming convention <code>(end5/end3/end).condition.replicate</code>.</p>
<p><code>aggregate</code> calculates the mean and sd for samples in the <code>control</code>
and <code>treated</code> condition separatly.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">End5SequenceDataFrame</span><span class="op">(</span></span>
<span>  <span class="va">df</span>,</span>
<span>  <span class="va">ranges</span>,</span>
<span>  <span class="va">sequence</span>,</span>
<span>  <span class="va">replicate</span>,</span>
<span>  <span class="va">condition</span>,</span>
<span>  <span class="va">bamfiles</span>,</span>
<span>  <span class="va">seqinfo</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">End3SequenceDataFrame</span><span class="op">(</span></span>
<span>  <span class="va">df</span>,</span>
<span>  <span class="va">ranges</span>,</span>
<span>  <span class="va">sequence</span>,</span>
<span>  <span class="va">replicate</span>,</span>
<span>  <span class="va">condition</span>,</span>
<span>  <span class="va">bamfiles</span>,</span>
<span>  <span class="va">seqinfo</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">EndSequenceDataFrame</span><span class="op">(</span></span>
<span>  <span class="va">df</span>,</span>
<span>  <span class="va">ranges</span>,</span>
<span>  <span class="va">sequence</span>,</span>
<span>  <span class="va">replicate</span>,</span>
<span>  <span class="va">condition</span>,</span>
<span>  <span class="va">bamfiles</span>,</span>
<span>  <span class="va">seqinfo</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">End5SequenceData</span><span class="op">(</span><span class="va">bamfiles</span>, <span class="va">annotation</span>, <span class="va">sequences</span>, <span class="va">seqinfo</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">End3SequenceData</span><span class="op">(</span><span class="va">bamfiles</span>, <span class="va">annotation</span>, <span class="va">sequences</span>, <span class="va">seqinfo</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">EndSequenceData</span><span class="op">(</span><span class="va">bamfiles</span>, <span class="va">annotation</span>, <span class="va">sequences</span>, <span class="va">seqinfo</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for End5SequenceData,BamFileList,GRangesList,XStringSet,ScanBamParam</span></span>
<span><span class="fu"><a href="RNAmodR-development.html">getData</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">bamfiles</span>, <span class="va">grl</span>, <span class="va">sequences</span>, <span class="va">param</span>, <span class="va">args</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for End3SequenceData,BamFileList,GRangesList,XStringSet,ScanBamParam</span></span>
<span><span class="fu"><a href="RNAmodR-development.html">getData</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">bamfiles</span>, <span class="va">grl</span>, <span class="va">sequences</span>, <span class="va">param</span>, <span class="va">args</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for EndSequenceData,BamFileList,GRangesList,XStringSet,ScanBamParam</span></span>
<span><span class="fu"><a href="RNAmodR-development.html">getData</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">bamfiles</span>, <span class="va">grl</span>, <span class="va">sequences</span>, <span class="va">param</span>, <span class="va">args</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for End5SequenceData</span></span>
<span><span class="fu"><a href="aggregate.html">aggregateData</a></span><span class="op">(</span><span class="va">x</span>, condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Both"</span>, <span class="st">"Treated"</span>, <span class="st">"Control"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for End3SequenceData</span></span>
<span><span class="fu"><a href="aggregate.html">aggregateData</a></span><span class="op">(</span><span class="va">x</span>, condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Both"</span>, <span class="st">"Treated"</span>, <span class="st">"Control"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for EndSequenceData</span></span>
<span><span class="fu"><a href="aggregate.html">aggregateData</a></span><span class="op">(</span><span class="va">x</span>, condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Both"</span>, <span class="st">"Treated"</span>, <span class="st">"Control"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for EndSequenceData</span></span>
<span><span class="fu"><a href="plotData.html">getDataTrack</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for End5SequenceData</span></span>
<span><span class="fu"><a href="plotData.html">getDataTrack</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for End3SequenceData</span></span>
<span><span class="fu"><a href="plotData.html">getDataTrack</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>df, ranges, sequence, replicate</dt>
<dd><p>inputs for creating a 
<code>SequenceDataFrame</code>. See 
<code><a href="SequenceDataFrame-class.html">SequenceDataFrame</a></code>.</p></dd>


<dt>condition</dt>
<dd><p>For <code><a href="aggregate.html">aggregate</a></code>: condition for which the data
should be aggregated.</p></dd>


<dt>bamfiles, annotation, seqinfo, grl, sequences, param, args, ...</dt>
<dd><p>See
<code><a href="SequenceData-class.html">SequenceData</a></code> and
<code><a href="SequenceData-functions.html">SequenceData-functions</a></code></p></dd>


<dt>x</dt>
<dd><p>a <code>End5SequenceData</code>, <code>End3SequenceData</code> or
<code>EndSequenceData</code> object</p></dd>


<dt>name</dt>
<dd><p>For <code><a href="plotData.html">getDataTrack</a></code>: a valid
transcript name. Must be a name of <code>ranges(x).</code></p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a <code>End5SequenceData</code>, a <code>End3SequenceData</code> or a
<code>EndSequenceData</code> object</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Construction of a End5SequenceData object</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/FelixErnst/RNAmodR.Data" class="external-link">RNAmodR.Data</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html" class="external-link">GFF3File</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/example.man.fasta.html" class="external-link">RNAmodR.Data.example.man.gff3</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-in"><span><span class="va">sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/example.man.fasta.html" class="external-link">RNAmodR.Data.example.man.fasta</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-in"><span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>treated <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html" class="external-link">RNAmodR.Data.example.wt.1</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-in"><span><span class="va">e5sd</span> <span class="op">&lt;-</span> <span class="fu">End5SequenceData</span><span class="op">(</span><span class="va">files</span>, annotation <span class="op">=</span> <span class="va">annotation</span>,</span></span>
<span class="r-in"><span>                        sequences <span class="op">=</span> <span class="va">sequences</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>makeTxDbFromGFF() has moved to the txdbmaker package. Please call</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   txdbmaker::makeTxDbFromGFF() to get rid of this warning.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Import genomic features from the file as a GRanges object ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Prepare the 'metadata' data frame ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Make the TxDb object ... </span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Direct call of 'as.data.frame.factor()' is deprecated.  Use 'as.data.frame.vector()' or 'as.data.frame()' instead</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Direct call of 'as.data.frame.factor()' is deprecated.  Use 'as.data.frame.vector()' or 'as.data.frame()' instead</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Direct call of 'as.data.frame.factor()' is deprecated.  Use 'as.data.frame.vector()' or 'as.data.frame()' instead</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Direct call of 'as.data.frame.factor()' is deprecated.  Use 'as.data.frame.vector()' or 'as.data.frame()' instead</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Direct call of 'as.data.frame.factor()' is deprecated.  Use 'as.data.frame.vector()' or 'as.data.frame()' instead</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Direct call of 'as.data.frame.factor()' is deprecated.  Use 'as.data.frame.vector()' or 'as.data.frame()' instead</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Direct call of 'as.data.frame.factor()' is deprecated.  Use 'as.data.frame.vector()' or 'as.data.frame()' instead</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Direct call of 'as.data.frame.factor()' is deprecated.  Use 'as.data.frame.vector()' or 'as.data.frame()' instead</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading 5'-end position data from BAM files ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

