<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>The ModifierSet class — ModifierSet-class • RNAmodR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="The ModifierSet class — ModifierSet-class" />
<meta property="og:description" content="The ModifierSet class allows multiple
Modifier objects to be created from the same
annotation and sequence data varying only the bam input files.
In addition the comparison of samples is also done via calling functions on 
the ModifierSet objects.
The ModifierSet is a virtual class, which derives from the 
SimpleList class with the slot elementType = &quot;Modifier&quot;. The
ModifierSet class has to be implemented for each specific analysis.#'" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.15</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/FelixErnst/RNAmodR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>The ModifierSet class</h1>
    <small class="dont-index">Source: <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/ModifierSet-class.R'><code>R/ModifierSet-class.R</code></a></small>
    <div class="hidden name"><code>ModifierSet-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>ModifierSet</code> class allows multiple
<code><a href='Modifier-class.html'>Modifier</a></code> objects to be created from the same
annotation and sequence data varying only the bam input files.</p>
<p>In addition the comparison of samples is also done via calling functions on 
the <code>ModifierSet</code> objects.</p>
<p>The <code>ModifierSet</code> is a virtual class, which derives from the 
<code>SimpleList</code> class with the slot <code>elementType = "Modifier"</code>. The
<code>ModifierSet</code> class has to be implemented for each specific analysis.#'</p>
    </div>

    <pre class="usage"><span class='fu'>ModifierSet</span>(<span class='no'>className</span>, <span class='no'>x</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='co'># S4 method for list</span>
<span class='fu'>ModifierSet</span>(
  <span class='no'>className</span>,
  <span class='no'>x</span>,
  <span class='kw'>annotation</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>sequences</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>seqinfo</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)

<span class='co'># S4 method for character</span>
<span class='fu'>ModifierSet</span>(
  <span class='no'>className</span>,
  <span class='no'>x</span>,
  <span class='kw'>annotation</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>sequences</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>seqinfo</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)

<span class='co'># S4 method for BamFileList</span>
<span class='fu'>ModifierSet</span>(
  <span class='no'>className</span>,
  <span class='no'>x</span>,
  <span class='kw'>annotation</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>sequences</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>seqinfo</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)

<span class='co'># S4 method for Modifier</span>
<span class='fu'>ModifierSet</span>(<span class='no'>className</span>, <span class='no'>x</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>className</th>
      <td><p>The name of the class which should be constructed.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>the input which can be of the following types</p><ul>
<li><p><code>Modifier</code>: a single <code>Modifier</code> or a list containg only 
<code>Modifier</code> objects. The input will just be used as elements of the
<code>ModifierSet</code></p></li>
<li><p><code>BamFileList</code>: a named <code>BamFileList</code> or a list of 
named <code>BamFileList</code></p></li>
<li><p><code>list</code>: a list of one or more types of elements: 
<code>BamFileList</code>, a named <code>list</code> or named <code>character</code> vector. All
elements must be or be coercible to a named <code>BamFileList</code> referencing 
existing bam files. Valid names are <code>control</code> and <code>treated</code></p></li>
</ul></td>
    </tr>
    <tr>
      <th>annotation</th>
      <td><p>annotation data, which must match the information contained
in the BAM files. This is parameter is only required, if <code>x</code> is not a 
<code>Modifier</code> object.</p></td>
    </tr>
    <tr>
      <th>sequences</th>
      <td><p>sequences matching the target sequences the reads were 
mapped onto. This must match the information contained in the BAM files. This
is parameter is only required, if <code>x</code> is not a <code>Modifier</code> object.</p></td>
    </tr>
    <tr>
      <th>seqinfo</th>
      <td><p>An optional <code><a href='https://rdrr.io/pkg/GenomeInfoDb/man/Seqinfo-class.html'>Seqinfo</a></code> 
argument or character vector, which can be coerced to one, to subset the 
sequences to be analyzed on a per chromosome basis.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional otpional parameters:</p><ul>
<li><p>internalBP <code>TRUE</code> or <code>FALSE</code>: should parallelization used
internally during creation of each <code>Modifier</code> or should the creation of
the <code>Modifier</code> objects be parallalized? (default: <code>internalBP =
FALSE</code>). Setting <code>internalBP</code> only makes sense, if the
<code><a href='RNAmodR-development.html'>getData</a></code> function for <code><a href='SequenceData-class.html'>SequenceData</a></code> class, the
<code><a href='aggregate.html'>aggregateData</a></code> or the <code><a href='modify.html'>findMod</a></code>
function contains parallelized code.</p></li>
</ul><p>All other arguments will be passed onto the <code>Modifier</code> objects.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a <code>ModifierSet</code> object of type <code>className</code></p>
    <h2 class="hasAnchor" id="creation"><a class="anchor" href="#creation"></a>Creation</h2>

    

<p>The input files have to be provided as a <code>list</code> of elements. Each
element in itself must be valid for the creation of <code><a href='Modifier-class.html'>Modifier</a></code>
object (Have a look at the man page for more details) and must be named.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


