<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>ModInosine — ModInosine • RNAmodR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="ModInosine — ModInosine" />
<meta property="og:description" content="Inosine can be detected in RNA-Seq data by the conversion of A positions to 
G. This conversion is detected by ModInosine and used to search for 
Inosine positions. dataType is &quot;PileupSequenceData&quot;.
Only samples labeled with the condition treated are used for this
analysis, since the A to G conversion is common feature among the reverse
transcriptases usually emploied. Let us know, if that is not the case, and
the class needs to be modified.
Further information on Functions of 
ModInosine." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/FelixErnst/RNAmodR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>ModInosine</h1>
    <small class="dont-index">Source: <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/Modifier-Inosine-class.R'><code>R/Modifier-Inosine-class.R</code></a></small>
    <div class="hidden name"><code>ModInosine.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Inosine can be detected in RNA-Seq data by the conversion of A positions to 
G. This conversion is detected by <code>ModInosine</code> and used to search for 
Inosine positions. <code>dataType</code> is <code>"PileupSequenceData"</code>.</p>
<p>Only samples labeled with the condition <code>treated</code> are used for this
analysis, since the A to G conversion is common feature among the reverse
transcriptases usually emploied. Let us know, if that is not the case, and
the class needs to be modified.</p>
<p>Further information on <code><a href='ModInosine-functions.html'>Functions</a></code> of 
<code>ModInosine</code>.</p>
    </div>

    <pre class="usage"><span class='fu'>ModInosine</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>annotation</span>, <span class='va'>sequences</span>, <span class='va'>seqinfo</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='fu'>ModSetInosine</span><span class='op'>(</span><span class='va'>x</span>, annotation <span class='op'>=</span> <span class='cn'>NA</span>, sequences <span class='op'>=</span> <span class='cn'>NA</span>, seqinfo <span class='op'>=</span> <span class='cn'>NA</span>, <span class='va'>...</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>the input which can be of the different types depending on whether
a <code>ModRiboMethSeq</code> or a <code>ModSetRiboMethSeq</code> object is to be 
constructed. For more information have a look at the documentation of
the <code><a href='Modifier-class.html'>Modifier</a></code> and 
<code><a href='ModifierSet-class.html'>ModifierSet</a></code> classes.</p></td>
    </tr>
    <tr>
      <th>annotation</th>
      <td><p>annotation data, which must match the information contained
in the BAM files. This is parameter is only required, if <code>x</code> is not a 
<code>Modifier</code> object.</p></td>
    </tr>
    <tr>
      <th>sequences</th>
      <td><p>sequences matching the target sequences the reads were 
mapped onto. This must match the information contained in the BAM files. This
is parameter is only required, if <code>x</code> is not a <code>Modifier</code> object.</p></td>
    </tr>
    <tr>
      <th>seqinfo</th>
      <td><p>An optional <code><a href='https://rdrr.io/pkg/GenomeInfoDb/man/Seqinfo-class.html'>Seqinfo</a></code> 
argument or character vector, which can be coerced to one, to subset the 
sequences to be analyzed on a per chromosome basis.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Optional arguments overwriting default values, which are</p><ul>
<li><p>minCoverage: The minimal coverage at the position as integer value 
(default: <code>minCoverage = 10L</code>).</p></li>
<li><p>minReplicate: minimum number of replicates needed for the analysis 
(default: <code>minReplicate = 1L</code>).</p></li>
<li><p>minScore: minimum score to identify Inosine positions de novo 
(default: <code>minScore = 0.4</code>).</p></li>
</ul></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a <code>ModInosine</code> or <code>ModSetInosine</code> object</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><code>ModInosine</code> score: the scores for reported Inosine positions are 
between 0 and 1. They are calculated as the relative amount of called G bases 
(<code>(G / N)</code>) and only saved for genomic A positions.</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Felix G.M. Ernst [aut]</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># construction of ModInosine object</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/FelixErnst/RNAmodR.Data'>RNAmodR.Data</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>rtracklayer</span><span class='op'>)</span>
<span class='va'>annotation</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html'>GFF3File</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/example.man.fasta.html'>RNAmodR.Data.example.man.gff3</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2021-01-12</span></div><div class='output co'>#&gt; <span class='message'>see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='input'><span class='va'>sequences</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/example.man.fasta.html'>RNAmodR.Data.example.man.fasta</a></span><span class='op'>(</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2021-01-12</span></div><div class='output co'>#&gt; <span class='message'>see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='input'><span class='va'>files</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>treated <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.wt.1</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2021-01-12</span></div><div class='output co'>#&gt; <span class='message'>see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='input'><span class='va'>mi</span> <span class='op'>&lt;-</span> <span class='fu'>ModInosine</span><span class='op'>(</span><span class='va'>files</span>,annotation <span class='op'>=</span> <span class='va'>annotation</span> ,sequences <span class='op'>=</span> <span class='va'>sequences</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Import genomic features from the file as a GRanges object ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Prepare the 'metadata' data frame ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Make the TxDb object ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Loading Pileup data from BAM files ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Aggregating data and calculating scores ... </span></div><div class='output co'>#&gt; <span class='message'>Starting to search for 'Inosine' ... </span></div><div class='output co'>#&gt; <span class='message'>done.</span></div><div class='input'><span class='co'># construction of ModSetInosine object</span>
<span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='va'>files</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='st'>"SampleSet1"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>treated <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.wt.1</a></span><span class='op'>(</span><span class='op'>)</span>,
                               treated <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.wt.2</a></span><span class='op'>(</span><span class='op'>)</span>,
                               treated <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.wt.3</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>,
              <span class='st'>"SampleSet2"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>treated <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.bud23.1</a></span><span class='op'>(</span><span class='op'>)</span>,
                               treated <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.bud23.2</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>,
              <span class='st'>"SampleSet3"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>treated <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.trm8.1</a></span><span class='op'>(</span><span class='op'>)</span>,
                               treated <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.trm8.2</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>msi</span> <span class='op'>&lt;-</span> <span class='fu'>ModSetInosine</span><span class='op'>(</span><span class='va'>files</span>, annotation <span class='op'>=</span> <span class='va'>annotation</span>, sequences <span class='op'>=</span> <span class='va'>sequences</span><span class='op'>)</span>
<span class='op'>}</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


