<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Modifier/ModifierSet functions — Modifier-functions • RNAmodR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Modifier/ModifierSet functions — Modifier-functions" />
<meta property="og:description" content="For the Modifier and  ModifierSet classes a number of functions
are implemented to access the data stored by the object.
The validAggregate and validModification functions check if
settings have been modified, after the data was 
loaded. This potentially invalidates them. To update the data, run the
aggregate or the modify function." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/FelixErnst/RNAmodR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Modifier/ModifierSet functions</h1>
    <small class="dont-index">Source: <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/Modifier-class.R'><code>R/Modifier-class.R</code></a>, <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/ModifierSet-class.R'><code>R/ModifierSet-class.R</code></a></small>
    <div class="hidden name"><code>Modifier-functions.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>For the <code>Modifier</code> and  <code>ModifierSet</code> classes a number of functions
are implemented to access the data stored by the object.</p>
<p>The <code>validAggregate</code> and <code>validModification</code> functions check if
<code><a href='settings.html'>settings</a></code> have been modified, after the data was 
loaded. This potentially invalidates them. To update the data, run the
<code>aggregate</code> or the <code>modify</code> function.</p>
    </div>

    <pre class="usage"><span class='fu'>bamfiles</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='fu'>mainScore</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='fu'>modifierType</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='fu'>modType</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='fu'>dataType</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='fu'>sequenceData</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='fu'>sequences</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='fu'>validAggregate</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='fu'>validModification</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier</span>
<span class='fu'>show</span><span class='op'>(</span><span class='va'>object</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier</span>
<span class='fu'>bamfiles</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier</span>
<span class='fu'><a href='https://rdrr.io/r/base/conditions.html'>conditions</a></span><span class='op'>(</span><span class='va'>object</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier</span>
<span class='fu'>mainScore</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier</span>
<span class='fu'>modifierType</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier</span>
<span class='fu'>modType</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier</span>
<span class='fu'>dataType</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier</span>
<span class='fu'>ranges</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier</span>
<span class='fu'><a href='SequenceData-functions.html'>replicates</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier</span>
<span class='fu'>seqinfo</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier</span>
<span class='fu'>seqtype</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier</span>
<span class='fu'>sequenceData</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier</span>
<span class='fu'>sequences</span><span class='op'>(</span><span class='va'>x</span>, modified <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier</span>
<span class='fu'>validAggregate</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier</span>
<span class='fu'>validModification</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for ModifierSet</span>
<span class='fu'>show</span><span class='op'>(</span><span class='va'>object</span><span class='op'>)</span>

<span class='co'># S4 method for ModifierSet</span>
<span class='fu'>bamfiles</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for ModifierSet</span>
<span class='fu'><a href='https://rdrr.io/r/base/conditions.html'>conditions</a></span><span class='op'>(</span><span class='va'>object</span><span class='op'>)</span>

<span class='co'># S4 method for ModifierSet</span>
<span class='fu'>mainScore</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for ModifierSet</span>
<span class='fu'><a href='modify.html'>modifications</a></span><span class='op'>(</span><span class='va'>x</span>, perTranscript <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>

<span class='co'># S4 method for ModifierSet</span>
<span class='fu'>modifierType</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for ModifierSet</span>
<span class='fu'>modType</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for ModifierSet</span>
<span class='fu'>dataType</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for ModifierSet</span>
<span class='fu'>ranges</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for ModifierSet</span>
<span class='fu'><a href='SequenceData-functions.html'>replicates</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for ModifierSet</span>
<span class='fu'>seqinfo</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for ModifierSet</span>
<span class='fu'>seqtype</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for ModifierSet</span>
<span class='fu'>sequences</span><span class='op'>(</span><span class='va'>x</span>, modified <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x, object</th>
      <td><p>a <code>Modifier</code> or <code>ModifierSet</code> class</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>modified</th>
      <td><p>For <code>sequences</code>: <code>TRUE</code> or <code>FALSE</code>: Should
the sequences be returned as a <code>ModRNAString</code>/<code>ModDNAString</code> with
the found modifications added on top of the <code>RNAString</code>/
<code>DNAString</code>? See 
<code><a href='https://rdrr.io/pkg/Modstrings/man/separate.html'>combineIntoModstrings</a></code>.</p></td>
    </tr>
    <tr>
      <th>perTranscript</th>
      <td><p><code>TRUE</code> or <code>FALSE</code>: Should the positions shown
per transcript? (default: <code>perTranscript = FALSE</code>)</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    
<ul>
<li><p><code>modifierType</code>: a character vector with the appropriate class
Name of a <code><a href='Modifier-class.html'>Modifier</a></code>.</p></li>
<li><p><code>modType</code>: a character vector with the modifications detected by
the <code>Modifier</code> class.</p></li>
<li><p><code>seqtype</code>: a single character value defining if either
"RNA" or "DNA" modifications are detected by the <code>Modifier</code> class.</p></li>
<li><p><code>mainScore</code>: a character vector.</p></li>
<li><p><code>sequenceData</code>: a <code>SequenceData</code> object.</p></li>
<li><p><code>modifications</code>: a <code>GRanges</code> or <code>GRangesList</code> object
describing the found modifications.</p></li>
<li><p><code>seqinfo</code>: a <code>Seqinfo</code> object.</p></li>
<li><p><code>sequences</code>: a <code>RNAStingSet</code> object.</p></li>
<li><p><code>ranges</code>: a <code>GRangesList</code> object with each element per
transcript.</p></li>
<li><p><code>bamfiles</code>: a <code>BamFileList</code> object.</p></li>
<li><p><code>validAggregate</code>: <code>TRUE</code> or <code>FALSE</code>. Checks if current
settings are the same for which the data was aggregate</p></li>
<li><p><code>validModification</code>: <code>TRUE</code> or <code>FALSE</code>. Checks if 
current settings are the same for which modification were found</p></li>
</ul>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='settings.html'>settings</a></code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>msi</span>,package<span class='op'>=</span><span class='st'>"RNAmodR"</span><span class='op'>)</span>
<span class='va'>mi</span> <span class='op'>&lt;-</span> <span class='va'>msi</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>
<span class='fu'>modifierType</span><span class='op'>(</span><span class='va'>mi</span><span class='op'>)</span> <span class='co'># The class name of the Modifier object</span>
</div><div class='output co'>#&gt; [1] "ModInosine"</div><div class='input'><span class='fu'>modifierType</span><span class='op'>(</span><span class='va'>msi</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] "ModInosine"</div><div class='input'><span class='fu'>seqtype</span><span class='op'>(</span><span class='va'>mi</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] "RNA"</div><div class='input'><span class='fu'>modType</span><span class='op'>(</span><span class='va'>mi</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] "I"</div><div class='input'><span class='fu'>mainScore</span><span class='op'>(</span><span class='va'>mi</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] "score"</div><div class='input'><span class='fu'>sequenceData</span><span class='op'>(</span><span class='va'>mi</span><span class='op'>)</span>
</div><div class='output co'>#&gt; PileupSequenceData with 11 elements containing 15 data columns and 3 metadata columns
#&gt; - Data columns:
#&gt;  pileup.treated.1.- pileup.treated.1.G pileup.treated.1.A pileup.treated.1.T
#&gt;           &lt;integer&gt;          &lt;integer&gt;          &lt;integer&gt;          &lt;integer&gt;
#&gt;  pileup.treated.1.C pileup.treated.2.- pileup.treated.2.G pileup.treated.2.A
#&gt;           &lt;integer&gt;          &lt;integer&gt;          &lt;integer&gt;          &lt;integer&gt;
#&gt;  pileup.treated.2.T pileup.treated.2.C pileup.treated.3.- pileup.treated.3.G
#&gt;           &lt;integer&gt;          &lt;integer&gt;          &lt;integer&gt;          &lt;integer&gt;
#&gt;  pileup.treated.3.A pileup.treated.3.T pileup.treated.3.C
#&gt;           &lt;integer&gt;          &lt;integer&gt;          &lt;integer&gt;
#&gt; -  Seqinfo object with 11 sequences from an unspecified genome; no seqlengths:</div><div class='input'><span class='fu'><a href='modify.html'>modifications</a></span><span class='op'>(</span><span class='va'>mi</span><span class='op'>)</span>
</div><div class='output co'>#&gt; GRanges object with 6 ranges and 5 metadata columns:
#&gt;       seqnames    ranges strand |         mod      source        type     score
#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;numeric&gt;
#&gt;   [1]     chr2        34      + |           I     RNAmodR      RNAMOD  0.900932
#&gt;   [2]     chr4        35      + |           I     RNAmodR      RNAMOD  0.899622
#&gt;   [3]     chr6        34      + |           I     RNAmodR      RNAMOD  0.984035
#&gt;   [4]     chr7        67      + |           I     RNAmodR      RNAMOD  0.934553
#&gt;   [5]     chr9         7      + |           I     RNAmodR      RNAMOD  0.709758
#&gt;   [6]    chr11        35      + |           I     RNAmodR      RNAMOD  0.874027
#&gt;            Parent
#&gt;       &lt;character&gt;
#&gt;   [1]           2
#&gt;   [2]           4
#&gt;   [3]           6
#&gt;   [4]           7
#&gt;   [5]           9
#&gt;   [6]          11
#&gt;   -------
#&gt;   seqinfo: 11 sequences from an unspecified genome; no seqlengths</div><div class='input'><span class='co'># general accessors</span>
<span class='fu'>seqinfo</span><span class='op'>(</span><span class='va'>mi</span><span class='op'>)</span>
</div><div class='output co'>#&gt; Seqinfo object with 11 sequences from an unspecified genome; no seqlengths:
#&gt;   seqnames seqlengths isCircular genome
#&gt;   chr1             NA         NA   &lt;NA&gt;
#&gt;   chr2             NA         NA   &lt;NA&gt;
#&gt;   chr3             NA         NA   &lt;NA&gt;
#&gt;   chr4             NA         NA   &lt;NA&gt;
#&gt;   chr5             NA         NA   &lt;NA&gt;
#&gt;   chr6             NA         NA   &lt;NA&gt;
#&gt;   chr7             NA         NA   &lt;NA&gt;
#&gt;   chr8             NA         NA   &lt;NA&gt;
#&gt;   chr9             NA         NA   &lt;NA&gt;
#&gt;   chr10            NA         NA   &lt;NA&gt;
#&gt;   chr11            NA         NA   &lt;NA&gt;</div><div class='input'><span class='fu'>sequences</span><span class='op'>(</span><span class='va'>mi</span><span class='op'>)</span>
</div><div class='output co'>#&gt; RNAStringSet object of length 11:
#&gt;      width seq                                              names               
#&gt;  [1]  1800 UAUCUGGUUGAUCCUGCCAGUAG...UGAACCUGCGGAAGGAUCAUUA 1
#&gt;  [2]    85 GGCAACUUGGCCGAGUGGUUAAG...UCGAGUCCUGCAGUUGUCGCCA 2
#&gt;  [3]    76 GCGGAUUUAGCUCAGUUGGGAGA...UCGAUCCACAGAAUUCGCACCA 3
#&gt;  [4]    77 GGUCUCUUGGCCCAGUUGGUUAA...UCGAUCCCGCUAGAGACCACCA 4
#&gt;  [5]    74 GCGCAAGUGGUUUAGUGGUAAAA...UCGAUUCCGGGCUUGCGCACCA 5
#&gt;  ...   ... ...
#&gt;  [7]    75 GGUUUUAUAGUGUAGUGGUUAUC...UCGAAUCCGGGUAAGACCUCCA 7
#&gt;  [8]    75 GCUCGUAUGGCGCAGUGGUAGCG...UCGAUCCUGAGUGCGAGCUCCA 8
#&gt;  [9]    75 UCCGAUAUAGUGUAACGGCUAUC...UCGACUCCCCGUAUCGGAGCCA 9
#&gt; [10]    85 GGCACUAUGGCCGAGUGGUUAAG...UCAAAUCCUGCUGGUGUCGCCA 10
#&gt; [11]    77 GGUUUCGUGGUCUAGUCGGUUAU...UCGAUCCUGGGCGAAAUCACCA 11</div><div class='input'><span class='fu'>ranges</span><span class='op'>(</span><span class='va'>mi</span><span class='op'>)</span>
</div><div class='output co'>#&gt; GRangesList object of length 11:
#&gt; $`1`
#&gt; GRanges object with 1 range and 3 metadata columns:
#&gt;       seqnames    ranges strand |   exon_id   exon_name exon_rank
#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt; &lt;integer&gt;
#&gt;   [1]     chr1    1-1800      + |         1     RDN18-1         1
#&gt;   -------
#&gt;   seqinfo: 11 sequences from an unspecified genome; no seqlengths
#&gt; 
#&gt; $`2`
#&gt; GRanges object with 1 range and 3 metadata columns:
#&gt;       seqnames    ranges strand |   exon_id   exon_name exon_rank
#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt; &lt;integer&gt;
#&gt;   [1]     chr2      1-85      + |         2    tS(AGA)A         1
#&gt;   -------
#&gt;   seqinfo: 11 sequences from an unspecified genome; no seqlengths
#&gt; 
#&gt; $`3`
#&gt; GRanges object with 1 range and 3 metadata columns:
#&gt;       seqnames    ranges strand |   exon_id   exon_name exon_rank
#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt; &lt;integer&gt;
#&gt;   [1]     chr3      1-76      + |         3    tF(GAA)B         1
#&gt;   -------
#&gt;   seqinfo: 11 sequences from an unspecified genome; no seqlengths
#&gt; 
#&gt; ...
#&gt; &lt;8 more elements&gt;</div><div class='input'><span class='fu'>bamfiles</span><span class='op'>(</span><span class='va'>mi</span><span class='op'>)</span>
</div><div class='output co'>#&gt; BamFileList of length 3
#&gt; names(3): treated treated treated</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


