<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>The Modifier class — Modifier-class • RNAmodR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="The Modifier class — Modifier-class" />
<meta property="og:description" content="The Modifier class is a virtual class, which provides the central
functionality to search for post-transcriptional RNA modification patterns in
high throughput sequencing data.
Each subclass has to implement the following functions:

Slot nucleotide:  Either &quot;RNA&quot; or &quot;DNA&quot;. For conveniance the
subclasses RNAModifier and DNAModifier are already available
and can be inherited from.
Function aggregateData: used for specific data 
aggregation
Function findMod: used for specific search for 
modifications


Optionally the function settings&amp;lt;- can be
implemented to store additional arguments, which the base class does not
recognize.
Modifier objects are constructed centrally by calling
Modifier() with a className matching the specific class to be
constructed. This will trigger the immediate analysis, if find.mod is
not set to FALSE." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/FelixErnst/RNAmodR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>The Modifier class</h1>
    <small class="dont-index">Source: <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/Modifier-class.R'><code>R/Modifier-class.R</code></a></small>
    <div class="hidden name"><code>Modifier-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>Modifier</code> class is a virtual class, which provides the central
functionality to search for post-transcriptional RNA modification patterns in
high throughput sequencing data.</p>
<p>Each subclass has to implement the following functions:</p>
<ul>
<li><p>Slot <code>nucleotide</code>:  Either "RNA" or "DNA". For conveniance the
subclasses <code>RNAModifier</code> and <code>DNAModifier</code> are already available
and can be inherited from.</p></li>
<li><p>Function <code><a href='aggregate.html'>aggregateData</a></code>: used for specific data 
aggregation</p></li>
<li><p>Function <code><a href='modify.html'>findMod</a></code>: used for specific search for 
modifications</p></li>
</ul>

<p>Optionally the function <code><a href='Modifier-functions.html'>settings&lt;-</a></code> can be
implemented to store additional arguments, which the base class does not
recognize.</p>
<p><code>Modifier</code> objects are constructed centrally by calling
<code>Modifier()</code> with a <code>className</code> matching the specific class to be
constructed. This will trigger the immediate analysis, if <code>find.mod</code> is
not set to <code>FALSE</code>.</p>
    </div>

    <pre class="usage"><span class='fu'>Modifier</span>(<span class='no'>className</span>, <span class='no'>x</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SequenceData</span>
<span class='fu'>Modifier</span>(
  <span class='no'>className</span>,
  <span class='no'>x</span>,
  <span class='kw'>annotation</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>sequences</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>seqinfo</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)

<span class='co'># S4 method for SequenceDataSet</span>
<span class='fu'>Modifier</span>(
  <span class='no'>className</span>,
  <span class='no'>x</span>,
  <span class='kw'>annotation</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>sequences</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>seqinfo</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)

<span class='co'># S4 method for SequenceDataList</span>
<span class='fu'>Modifier</span>(
  <span class='no'>className</span>,
  <span class='no'>x</span>,
  <span class='kw'>annotation</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>sequences</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>seqinfo</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)

<span class='co'># S4 method for character</span>
<span class='fu'>Modifier</span>(
  <span class='no'>className</span>,
  <span class='no'>x</span>,
  <span class='kw'>annotation</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>sequences</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>seqinfo</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)

<span class='co'># S4 method for list</span>
<span class='fu'>Modifier</span>(
  <span class='no'>className</span>,
  <span class='no'>x</span>,
  <span class='kw'>annotation</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>sequences</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>seqinfo</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)

<span class='co'># S4 method for BamFileList</span>
<span class='fu'>Modifier</span>(
  <span class='no'>className</span>,
  <span class='no'>x</span>,
  <span class='kw'>annotation</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>sequences</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>seqinfo</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>className</th>
      <td><p>The name of the class which should be constructed.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>the input which can be of the following types</p><ul>
<li><p><code>SequenceData</code>: a single <code>SequenceData</code> or a list
containing only <code>SequenceData</code> objects. The input will just be used to
file the <code>data</code> slot of the <code>Modifier</code> and must match the
requirements of specific <code>Modifier</code> class.</p></li>
<li><p><code>BamFileList</code>: a named <code>BamFileList</code></p></li>
<li><p><code>character</code>: a <code>character</code> vector, which must be coercible
to a named <code>BamFileList</code> referencing existing bam files. Valid names are
<code>control</code> and <code>treated</code> to define conditions and replicates</p></li>
</ul></td>
    </tr>
    <tr>
      <th>annotation</th>
      <td><p>annotation data, which must match the information contained
in the BAM files. This parameter is only required if <code>x</code> is not a
<code>SequenceData</code> object or a list of <code>SequenceData</code> objects.</p></td>
    </tr>
    <tr>
      <th>sequences</th>
      <td><p>sequences matching the target sequences the reads were
mapped onto. This must match the information contained in the BAM files.
TThis parameter is only required if <code>x</code> is not a <code>SequenceData</code>
object or a list of <code>SequenceData</code> objects.</p></td>
    </tr>
    <tr>
      <th>seqinfo</th>
      <td><p>An optional <code><a href='https://rdrr.io/pkg/GenomeInfoDb/man/Seqinfo-class.html'>Seqinfo</a></code>
argument or character vector, which can be coerced to one, to subset the
sequences to be analyzed on a per chromosome basis.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional otpional parameters:</p><ul>
<li><p><code>find.mod</code>: <code>TRUE</code> or <code>FALSE</code>: should the search for
for modifications be triggered upon construction? If not the search can be
started by calling the <code><a href='modify.html'>modify()</a></code> function.</p></li>
<li><p>additional parameters depending on the specific <code>Modifier</code> class</p></li>
</ul><p>All additional options must be named and will be passed to the
<code><a href='settings.html'>settings</a></code> function and onto the <code>SequenceData</code>
objects, if <code>x</code> is not a <code>SequenceData</code> object or a list of
<code>SequenceData</code> objects.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a <code>Modifier</code> object of type <code>className</code></p>
    <h2 class="hasAnchor" id="slots"><a class="anchor" href="#slots"></a>Slots</h2>

    

<dl'>
<dt><code>nucleotide</code></dt><dd><p>a <code>character</code> value, which needs to contain "RNA" or 
"DNA"</p></dd>

<dt><code>mod</code></dt><dd><p>a <code>character</code> value, which needs to contain one or more
elements from the alphabet of a
<code><a href='https://rdrr.io/pkg/Modstrings/man/ModRNAString.html'>ModRNAString</a></code> or 
<code><a href='https://rdrr.io/pkg/Modstrings/man/ModDNAString.html'>ModDNAString</a></code> class.</p></dd>

<dt><code>score</code></dt><dd><p>the main score identifier used for visualizations</p></dd>

<dt><code>dataType</code></dt><dd><p>the class name(s) of the <code>SequenceData</code> class used</p></dd>

<dt><code>bamfiles</code></dt><dd><p>the input bam files as <code>BamFileList</code></p></dd>

<dt><code>condition</code></dt><dd><p>conditions along the <code>BamFileList</code>: Either
<code>control</code> or <code>treated</code></p></dd>

<dt><code>replicate</code></dt><dd><p>replicate number along the <code>BamFileList</code> for each of the
condition types.</p></dd>

<dt><code>data</code></dt><dd><p>The sequence data object: Either a <code>SequenceData</code>,
<code>SequenceDataSet</code> or a <code>SequenceDataList</code> object, if more than one
<code>dataType</code> is used.</p></dd>

<dt><code>aggregate</code></dt><dd><p>the aggregated data as a <code>SplitDataFrameList</code></p></dd>

<dt><code>modifications</code></dt><dd><p>the found modifications as a <code>GRanges</code> object</p></dd>

<dt><code>settings</code></dt><dd><p>arguments used for the analysis as a <code>list</code></p></dd>

<dt><code>aggregateValidForCurrentArguments</code></dt><dd><p><code>TRUE</code> or <code>FALSE</code> whether
the aggregate data was constructed with the current arguments</p></dd>

<dt><code>modificationsValidForCurrentArguments</code></dt><dd><p><code>TRUE</code> or <code>FALSE</code>
whether the modifications were found with the current arguments</p></dd>

</dl>
    <h2 class="hasAnchor" id="creation"><a class="anchor" href="#creation"></a>Creation</h2>

    

<p><code>Modifier</code> objects can be created in two ways, either by providing a
list of bamfiles or
<code>SequenceData</code>/<code>SequenceDataSet</code>/<code>SequenceDataList</code> objects,
which match the structure in <code><a href='Modifier-functions.html'>dataType()</a></code>.</p>
<p><code><a href='Modifier-functions.html'>dataType()</a></code> can be a <code>character</code> vector or a <code>list</code> of
<code>character</code> vectors and depending on this the input files have to
follow this structure:</p>
<ul>
<li><p>a single <code>character</code>: a <code>SequenceData</code> is
constructed/expected.</p></li>
<li><p>a <code>character</code> vector: a <code>SequenceDataSet</code> is
constructed/expected.</p></li>
<li><p>a <code>list</code> of <code>character</code> vectors: a <code>SequenceDataList</code>
is constructed/expected.</p></li>
</ul>

<p>The cases for a <code>SequenceData</code> or <code>SequenceDataSet</code> are straight
forward, since the input remains the same. The last case is special, since it
is a hypothetical option, in which bam files from two or more different
methods have to be combined to reliably detect a single modification (The
elements of a <code>SequenceDataList</code> don't have to be created from the
bamfiles, whereas from a <code>SequenceDataSet</code> they have to be).</p>
<p>For this example a <code>list</code> of <code>character</code> vectors is expected.
Each element must be named according to the names of <code><a href='Modifier-functions.html'>dataType()</a></code> and
contain a <code>character</code> vector for creating a <code>SequenceData</code> object.</p>
<p>All additional options must be named and will be passed to the
<code><a href='settings.html'>settings</a></code> function and onto the <code>SequenceData</code>
objects, if <code>x</code> is not a <code>SequenceData</code> object or a list of
<code>SequenceData</code> objects.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


