<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>NormEnd5SequenceData/NormEnd3SequenceData — NormEndSequenceData-class • RNAmodR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="NormEnd5SequenceData/NormEnd3SequenceData — NormEndSequenceData-class" />
<meta property="og:description" content="The NormEnd5SequenceData/NormEnd3SequenceData
aggregate the counts of read ends (Either 5' or 3') at each position along a
transcript. In addition, the number of counts are then normalized to the
length of the transcript and to the overlapping reads.
Both classes contain three columns per data file named using the
following naming convention (normend5/normend3).condition.replicate.
The three columns are distinguished by additional identifiers ends,
norm.tx and norm.ol.
aggregate calculates the mean and sd for samples in the control
and treated condition separatly. Similar to the stored results for 
each of the two conditions six columns are returned (three for mean and sd 
each) ending in ends, tx and ol." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/FelixErnst/RNAmodR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>NormEnd5SequenceData/NormEnd3SequenceData</h1>
    <small class="dont-index">Source: <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/SequenceData-normalized-end-pos.R'><code>R/SequenceData-normalized-end-pos.R</code></a></small>
    <div class="hidden name"><code>NormEndSequenceData-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>NormEnd5SequenceData</code>/<code>NormEnd3SequenceData</code>
aggregate the counts of read ends (Either 5' or 3') at each position along a
transcript. In addition, the number of counts are then normalized to the
length of the transcript and to the overlapping reads.</p>
<p>Both classes contain three columns per data file named using the
following naming convention <code>(normend5/normend3).condition.replicate</code>.
The three columns are distinguished by additional identifiers <code>ends</code>,
<code>norm.tx</code> and <code>norm.ol</code>.</p>
<p><code>aggregate</code> calculates the mean and sd for samples in the <code>control</code>
and <code>treated</code> condition separatly. Similar to the stored results for 
each of the two conditions six columns are returned (three for mean and sd 
each) ending in <code>ends</code>, <code>tx</code> and <code>ol</code>.</p>
    </div>

    <pre class="usage"><span class='fu'>NormEnd5SequenceDataFrame</span>(
  <span class='no'>df</span>,
  <span class='no'>ranges</span>,
  <span class='no'>sequence</span>,
  <span class='no'>replicate</span>,
  <span class='no'>condition</span>,
  <span class='no'>bamfiles</span>,
  <span class='no'>seqinfo</span>
)

<span class='fu'>NormEnd3SequenceDataFrame</span>(
  <span class='no'>df</span>,
  <span class='no'>ranges</span>,
  <span class='no'>sequence</span>,
  <span class='no'>replicate</span>,
  <span class='no'>condition</span>,
  <span class='no'>bamfiles</span>,
  <span class='no'>seqinfo</span>
)

<span class='fu'>NormEnd5SequenceData</span>(<span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='fu'>NormEnd3SequenceData</span>(<span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='co'># S4 method for NormEnd5SequenceData,BamFileList,GRangesList,XStringSet,ScanBamParam</span>
<span class='fu'><a href='RNAmodR-development.html'>getData</a></span>(<span class='no'>x</span>, <span class='no'>bamfiles</span>, <span class='no'>grl</span>, <span class='no'>sequences</span>, <span class='no'>param</span>, <span class='no'>args</span>)

<span class='co'># S4 method for NormEnd3SequenceData,BamFileList,GRangesList,XStringSet,ScanBamParam</span>
<span class='fu'><a href='RNAmodR-development.html'>getData</a></span>(<span class='no'>x</span>, <span class='no'>bamfiles</span>, <span class='no'>grl</span>, <span class='no'>sequences</span>, <span class='no'>param</span>, <span class='no'>args</span>)

<span class='co'># S4 method for NormEnd5SequenceData</span>
<span class='fu'><a href='aggregate.html'>aggregateData</a></span>(<span class='no'>x</span>, <span class='kw'>condition</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"Both"</span>, <span class='st'>"Treated"</span>, <span class='st'>"Control"</span>))

<span class='co'># S4 method for NormEnd3SequenceData</span>
<span class='fu'><a href='aggregate.html'>aggregateData</a></span>(<span class='no'>x</span>, <span class='kw'>condition</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"Both"</span>, <span class='st'>"Treated"</span>, <span class='st'>"Control"</span>))

<span class='co'># S4 method for NormEnd5SequenceData</span>
<span class='fu'><a href='plotData.html'>getDataTrack</a></span>(<span class='no'>x</span>, <span class='no'>name</span>, <span class='no'>...</span>)

<span class='co'># S4 method for NormEnd3SequenceData</span>
<span class='fu'><a href='plotData.html'>getDataTrack</a></span>(<span class='no'>x</span>, <span class='no'>name</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>df, ranges, sequence, replicate</th>
      <td><p>inputs for creating a 
<code>SequenceDataFrame</code>. See 
<code><a href='SequenceDataFrame-class.html'>SequenceDataFrame</a></code>.</p></td>
    </tr>
    <tr>
      <th>condition</th>
      <td><p>For <code><a href='aggregate.html'>aggregate</a></code>: condition for which the data 
should be aggregated.</p></td>
    </tr>
    <tr>
      <th>bamfiles, annotation, seqinfo, grl, sequences, param, args, ...</th>
      <td><p>See 
<code><a href='SequenceData-class.html'>SequenceData</a></code> and
<code><a href='SequenceData-functions.html'>SequenceData-functions</a></code></p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>a <code>CoverageSequenceData</code></p></td>
    </tr>
    <tr>
      <th>name</th>
      <td><p>For <code><a href='plotData.html'>getDataTrack</a></code>: a valid 
transcript name. Must be a name of <code>ranges(x)</code></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a <code>NormEnd5SequenceData</code> or <code>NormEnd3SequenceData</code> object</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Construction of a NormEnd5SequenceData object</span>
<span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='no'>RNAmodR.Data</span>)
<span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='no'>rtracklayer</span>)
<span class='no'>annotation</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html'>GFF3File</a></span>(<span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/example.man.fasta.html'>RNAmodR.Data.example.man.gff3</a></span>())
<span class='no'>sequences</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/example.man.fasta.html'>RNAmodR.Data.example.man.fasta</a></span>()
<span class='no'>files</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='kw'>treated</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.wt.1</a></span>())
<span class='no'>ne5sd</span> <span class='kw'>&lt;-</span> <span class='fu'>NormEnd5SequenceData</span>(<span class='no'>files</span>, <span class='kw'>annotation</span> <span class='kw'>=</span> <span class='no'>annotation</span>,
                              <span class='kw'>sequences</span> <span class='kw'>=</span> <span class='no'>sequences</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


