<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Visualizing data data from a SequenceData, 
SequenceDataSet, SequenceDataList, Modifier or 
ModifierSet object. — plotData • RNAmodR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Visualizing data data from a SequenceData, 
SequenceDataSet, SequenceDataList, Modifier or 
ModifierSet object. — plotData"><meta property="og:description" content="With the plotData and plotDataByCoord functions data
from a SequenceData, SequenceDataSet, SequenceDataList,
Modifier or ModifierSet object can be visualized.
Internally the functionality of the Gviz package is used. For each
SequenceData and Modifier class the getDataTrack is
implemented returning a DataTrack object
from the Gviz package.
Positions to be visualized are selected by defining a genomic coordinate,
for which x has to contain data."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.17.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/FelixErnst/RNAmodR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Visualizing data data from a <code>SequenceData</code>, 
<code>SequenceDataSet</code>, <code>SequenceDataList</code>, <code>Modifier</code> or 
<code>ModifierSet</code> object.</h1>
    <small class="dont-index">Source: <a href="https://github.com/FelixErnst/RNAmodR/blob/HEAD/R/AllGenerics.R" class="external-link"><code>R/AllGenerics.R</code></a>, <a href="https://github.com/FelixErnst/RNAmodR/blob/HEAD/R/Modifier-viz.R" class="external-link"><code>R/Modifier-viz.R</code></a>, <a href="https://github.com/FelixErnst/RNAmodR/blob/HEAD/R/ModifierSet-viz.R" class="external-link"><code>R/ModifierSet-viz.R</code></a>, and 3 more</small>
    <div class="hidden name"><code>plotData.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>With the <code>plotData</code> and <code>plotDataByCoord</code> functions data
from a <code>SequenceData</code>, <code>SequenceDataSet</code>, <code>SequenceDataList</code>,
<code>Modifier</code> or <code>ModifierSet</code> object can be visualized.</p>
<p>Internally the functionality of the <code>Gviz</code> package is used. For each
<code>SequenceData</code> and <code>Modifier</code> class the <code>getDataTrack</code> is
implemented returning a <code><a href="https://rdrr.io/pkg/Gviz/man/DataTrack-class.html" class="external-link">DataTrack</a></code> object
from the <code>Gviz</code> package.</p>
<p>Positions to be visualized are selected by defining a genomic coordinate,
for which <code>x</code> has to contain data.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotData</span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, from <span class="op">=</span> <span class="fl">1L</span>, to <span class="op">=</span> <span class="fl">30L</span>, <span class="va">type</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotDataByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">type</span>, window.size <span class="op">=</span> <span class="fl">15L</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">getDataTrack</span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for Modifier,GRanges</span></span>
<span><span class="fu">plotDataByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, type <span class="op">=</span> <span class="cn">NA</span>, window.size <span class="op">=</span> <span class="fl">15L</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for Modifier</span></span>
<span><span class="fu">plotData</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">name</span>,</span>
<span>  <span class="va">from</span>,</span>
<span>  <span class="va">to</span>,</span>
<span>  type <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  showSequenceData <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  showSequence <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  showAnnotation <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for Modifier</span></span>
<span><span class="fu">getDataTrack</span><span class="op">(</span><span class="va">x</span>, name <span class="op">=</span> <span class="va">name</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for ModifierSet,GRanges</span></span>
<span><span class="fu">plotDataByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, type <span class="op">=</span> <span class="cn">NA</span>, window.size <span class="op">=</span> <span class="fl">15L</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for ModifierSet</span></span>
<span><span class="fu">plotData</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">name</span>,</span>
<span>  <span class="va">from</span>,</span>
<span>  <span class="va">to</span>,</span>
<span>  type <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  showSequenceData <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  showSequence <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  showAnnotation <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceData,GRanges</span></span>
<span><span class="fu">plotDataByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, type <span class="op">=</span> <span class="cn">NA</span>, window.size <span class="op">=</span> <span class="fl">15L</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceData</span></span>
<span><span class="fu">plotData</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">name</span>,</span>
<span>  <span class="va">from</span>,</span>
<span>  <span class="va">to</span>,</span>
<span>  perTranscript <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  showSequence <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  showAnnotation <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceData</span></span>
<span><span class="fu">getDataTrack</span><span class="op">(</span><span class="va">x</span>, name <span class="op">=</span> <span class="va">name</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceDataList</span></span>
<span><span class="fu">getDataTrack</span><span class="op">(</span><span class="va">x</span>, name <span class="op">=</span> <span class="va">name</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceDataList,GRanges</span></span>
<span><span class="fu">plotDataByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, type <span class="op">=</span> <span class="cn">NA</span>, window.size <span class="op">=</span> <span class="fl">15L</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceDataList</span></span>
<span><span class="fu">plotData</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">name</span>,</span>
<span>  <span class="va">from</span>,</span>
<span>  <span class="va">to</span>,</span>
<span>  perTranscript <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  showSequence <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  showAnnotation <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceDataSet</span></span>
<span><span class="fu">getDataTrack</span><span class="op">(</span><span class="va">x</span>, name <span class="op">=</span> <span class="va">name</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceDataSet,GRanges</span></span>
<span><span class="fu">plotDataByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, type <span class="op">=</span> <span class="cn">NA</span>, window.size <span class="op">=</span> <span class="fl">15L</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceDataSet</span></span>
<span><span class="fu">plotData</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">name</span>,</span>
<span>  <span class="va">from</span>,</span>
<span>  <span class="va">to</span>,</span>
<span>  perTranscript <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  showSequence <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  showAnnotation <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p>a <code>SequenceData</code>, <code>SequenceDataSet</code>,
<code>SequenceDataList</code>, <code>Modifier</code> or <code>ModifierSet</code> object.</p></dd>


<dt>name</dt>
<dd><p>Only for <code>plotData</code>: the transcript name</p></dd>


<dt>from</dt>
<dd><p>Only for <code>plotData</code>: start position</p></dd>


<dt>to</dt>
<dd><p>Only for <code>plotData</code>: end position</p></dd>


<dt>type</dt>
<dd><p>the data type of data show as data tracks.</p></dd>


<dt>...</dt>
<dd><p>optional parameters:</p><ul><li><p><code>modified.seq</code> <code>TRUE</code> or <code>FALSE</code>. Should the sequence 
shown with modified nucleotide positions? (default: 
<code>modified.seq = FALSE</code>)</p></li>
<li><p><code>additional.mod</code> other modifications, which should be shown
in the annotation and sequence track. The must be a <code>GRanges</code> compatible
with <code><a href="https://rdrr.io/pkg/Modstrings/man/separate.html" class="external-link">combineIntoModstrings</a></code>.</p></li>
<li><p><code>annotation.track.pars</code> Parameters passed onto the 
<code><a href="https://rdrr.io/pkg/Gviz/man/AnnotationTrack-class.html" class="external-link">AnnotationTrack</a></code>.</p></li>
<li><p><code>sequence.track.pars</code> Parameters passed onto the 
<code><a href="https://rdrr.io/pkg/Gviz/man/SequenceTrack-class.html" class="external-link">SequenceTrack</a></code>.</p></li>
</ul></dd>


<dt>coord</dt>
<dd><p>coordinates of a positions to subset to as a 
<code>GRanges</code> object. The 'Parent' column is expected to match the 
transcript name.</p></dd>


<dt>window.size</dt>
<dd><p>integer value for the number of positions on the left and 
right site of the selected positions included in the plotting (default: 
<code>window.size = 15L</code>)</p></dd>


<dt>showSequenceData</dt>
<dd><p><code>TRUE</code> or <code>FALSE</code>: should the sequence data
be shown? (default: <code>seqdata = FALSE</code>)</p></dd>


<dt>showSequence</dt>
<dd><p><code>TRUE</code> or <code>FALSE</code>: should a sequence track be 
shown? (default: <code>seqdata = TRUE</code>)</p></dd>


<dt>showAnnotation</dt>
<dd><p><code>TRUE</code> or <code>FALSE</code>: should a annotation track 
be shown? (default: <code>seqdata = FALSE</code>)</p></dd>


<dt>perTranscript</dt>
<dd><p><code>TRUE</code> or <code>FALSE</code>: Should the positions shown
per transcript? (default: <code>perTranscript = FALSE</code>)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a plot send to the active graphic device</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">msi</span>,package<span class="op">=</span><span class="st">"RNAmodR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">plotData</span><span class="op">(</span><span class="va">msi</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">"2"</span>, from <span class="op">=</span> <span class="fl">10L</span>, to <span class="op">=</span> <span class="fl">45L</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotData-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="fu">plotData</span><span class="op">(</span><span class="va">msi</span>, <span class="st">"2"</span>, from <span class="op">=</span> <span class="fl">10L</span>, to <span class="op">=</span> <span class="fl">45L</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

