<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>RNAmodR developments functions — RNAmodR-development • RNAmodR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="RNAmodR developments functions — RNAmodR-development" />
<meta property="og:description" content="These functions are not intended for general use, but are used for 
additional package development.
getData is used to load data into a
SequenceData object and must be
implented for all SequenceData classes. The results must match the
requirements outlined in the value section.
In addition the following functions should be implemented for complete
functionality:
aggregateData for each SequenceData and Modifier class.
See also aggregateData
findMod for each Modifier class. See also 
findMod.
plotData/plotDataByCoord for each Modifier 
and ModifierSet class. See also 
plotData.
The following helper function can be called from within findMod to
construct a coordinate for each modification found:
constructModRanges constructs a GRanges object describing the
location, type and associated scores of a modification.
constructModRanges is typically called from the modify
function, which must be implemented for all
Modifier classes." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/FelixErnst/RNAmodR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>RNAmodR developments functions</h1>
    <small class="dont-index">Source: <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/RNAmodR.R'><code>R/RNAmodR.R</code></a>, <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/Modifier-utils.R'><code>R/Modifier-utils.R</code></a>, and 1 more</small>
    <div class="hidden name"><code>RNAmodR-development.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>These functions are not intended for general use, but are used for 
additional package development.</p>
<p><code>getData</code> is used to load data into a
<code><a href='SequenceData-class.html'>SequenceData</a></code> object and must be
implented for all <code>SequenceData</code> classes. The results must match the
requirements outlined in the value section.</p>
<p>In addition the following functions should be implemented for complete
functionality:</p>
<p><code>aggregateData</code> for each <code>SequenceData</code> and <code>Modifier</code> class.
See also <code><a href='aggregate.html'>aggregateData</a></code></p>
<p><code>findMod</code> for each <code>Modifier</code> class. See also 
<code><a href='modify.html'>findMod</a></code>.</p>
<p><code>plotData</code>/<code>plotDataByCoord</code> for each <code>Modifier</code> 
and <code>ModifierSet</code> class. See also 
<code><a href='plotData.html'>plotData</a></code>.</p>
<p>The following helper function can be called from within <code>findMod</code> to
construct a coordinate for each modification found:</p>
<p><code>constructModRanges</code> constructs a <code>GRanges</code> object describing the
location, type and associated scores of a modification.
<code>constructModRanges</code> is typically called from the <code>modify</code>
function, which must be implemented for all
<code><a href='Modifier-class.html'>Modifier</a></code> classes.</p>
    </div>

    <pre class="usage"><span class='fu'>constructModRanges</span><span class='op'>(</span><span class='va'>range</span>, <span class='va'>data</span>, <span class='va'>modType</span>, <span class='va'>scoreFun</span>, <span class='va'>source</span>, <span class='va'>type</span><span class='op'>)</span>

<span class='fu'>getData</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>bamfiles</span>, <span class='va'>grl</span>, <span class='va'>sequences</span>, <span class='va'>param</span>, <span class='va'>args</span><span class='op'>)</span>

<span class='co'># S4 method for GRanges,DataFrame</span>
<span class='fu'>constructModRanges</span><span class='op'>(</span><span class='va'>range</span>, <span class='va'>data</span>, <span class='va'>modType</span>, <span class='va'>scoreFun</span>, <span class='va'>source</span>, <span class='va'>type</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>range</th>
      <td><p>for <code>constructModRanges</code>: a <code>GRanges</code> object</p></td>
    </tr>
    <tr>
      <th>data</th>
      <td><p>for <code>constructModRanges</code>: a <code>DataFrame</code> object</p></td>
    </tr>
    <tr>
      <th>modType</th>
      <td><p>for <code>constructModRanges</code>: a valid shortName for the 
modification found. Must be present in <code>shortName(ModRNAString())</code>.</p></td>
    </tr>
    <tr>
      <th>scoreFun</th>
      <td><p>for <code>constructModRanges</code>: a custom function for 
extracting scores from <code>data</code>. The result must be a <code>list</code>.</p></td>
    </tr>
    <tr>
      <th>source</th>
      <td><p>for <code>constructModRanges</code>: a single character vector for
populating the source column of the result.</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p>for <code>constructModRanges</code>: a single character vector for
populating the source column of the result.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>for <code>getData</code>:a <code>SequenceData</code> object.</p></td>
    </tr>
    <tr>
      <th>bamfiles</th>
      <td><p>for <code>getData</code>:a <code>BamFileList</code> object.</p></td>
    </tr>
    <tr>
      <th>grl</th>
      <td><p>for <code>getData</code>:a <code>GRangesList</code> object.</p></td>
    </tr>
    <tr>
      <th>sequences</th>
      <td><p>for <code>getData</code>:a <code>XStringSet</code> object.</p></td>
    </tr>
    <tr>
      <th>param</th>
      <td><p>for <code>getData</code>:a <code>ScanBamParam</code> object.</p></td>
    </tr>
    <tr>
      <th>args</th>
      <td><p>for <code>getData</code>: a <code>list</code> with optional arguments.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    
<ul>
<li><p><code>getData</code>: returns a list with elements per BamFile in 
<code>bamfiles</code>. Elements can be 
<code><a href='https://rdrr.io/pkg/IRanges/man/AtomicList-class.html'>IntegerList</a></code>, 
<code><a href='https://rdrr.io/pkg/IRanges/man/AtomicList-class.html'>NumericList</a></code> or a
<code><a href='https://rdrr.io/pkg/IRanges/man/DataFrameList-class.html'>CompressedSplitDataFrameList</a></code>. The
data in the elements must be order by increasing positions numbers. However, 
names and rownames will be discarded.</p></li>
<li><p><code>constructModRanges</code>: returns a <code>GRanges</code> object with
genomic coordinates of modified nucleotides in the associated transcripts.</p></li>
</ul>


    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># new SequenceData class</span>
<span class='kw'>setClass</span><span class='op'>(</span>Class <span class='op'>=</span> <span class='st'>"ExampleSequenceData"</span>,
         contains <span class='op'>=</span> <span class='st'>"SequenceData"</span>,
         prototype <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>minQuality <span class='op'>=</span> <span class='fl'>5L</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>ExampleSequenceData</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>bamfiles</span>, <span class='va'>annotation</span>, <span class='va'>sequences</span>, <span class='va'>seqinfo</span>, <span class='va'>...</span><span class='op'>)</span><span class='op'>{</span>
  <span class='fu'>RNAmodR</span><span class='fu'>:::</span><span class='fu'><a href='SequenceData-class.html'>SequenceData</a></span><span class='op'>(</span><span class='st'>"Example"</span>, bamfiles <span class='op'>=</span> <span class='va'>bamfiles</span>, 
                         annotation <span class='op'>=</span> <span class='va'>annotation</span>, sequences <span class='op'>=</span> <span class='va'>sequences</span>,
                         seqinfo <span class='op'>=</span> <span class='va'>seqinfo</span>, <span class='va'>...</span><span class='op'>)</span>
<span class='op'>}</span>
<span class='kw'>setMethod</span><span class='op'>(</span><span class='st'>"getData"</span>,
          signature <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='st'>"ExampleSequenceData"</span>,
                        bamfiles <span class='op'>=</span> <span class='st'>"BamFileList"</span>,
                        grl <span class='op'>=</span> <span class='st'>"GRangesList"</span>,
                        sequences <span class='op'>=</span> <span class='st'>"XStringSet"</span>,
                        param <span class='op'>=</span> <span class='st'>"ScanBamParam"</span><span class='op'>)</span>,
          definition <span class='op'>=</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>bamfiles</span>, <span class='va'>grl</span>, <span class='va'>sequences</span>, <span class='va'>param</span>, <span class='va'>args</span><span class='op'>)</span><span class='op'>{</span>
            <span class='co'>###</span>
          <span class='op'>}</span>
<span class='op'>)</span>
<span class='kw'>setMethod</span><span class='op'>(</span><span class='st'>"aggregateData"</span>,
          signature <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='st'>"ExampleSequenceData"</span><span class='op'>)</span>,
          <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>condition</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Both"</span>,<span class='st'>"Treated"</span>,<span class='st'>"Control"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>{</span>
            <span class='co'>###</span>
          <span class='op'>}</span>
<span class='op'>)</span>
<span class='kw'>setMethod</span><span class='op'>(</span>
  f <span class='op'>=</span> <span class='st'>"getDataTrack"</span>,
  signature <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='st'>"ExampleSequenceData"</span><span class='op'>)</span>,
  definition <span class='op'>=</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>name</span>, <span class='va'>...</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='co'>###</span>
  <span class='op'>}</span>
<span class='op'>)</span>

<span class='co'># new Modifier class</span>
<span class='kw'>setClass</span><span class='op'>(</span><span class='st'>"ModExample"</span>,
         contains <span class='op'>=</span> <span class='st'>"Modifier"</span>,
         prototype <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>mod <span class='op'>=</span> <span class='st'>"X"</span>,
                          score <span class='op'>=</span> <span class='st'>"score"</span>,
                          dataType <span class='op'>=</span> <span class='st'>"ExampleSequenceData"</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>ModExample</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>annotation</span>, <span class='va'>sequences</span>, <span class='va'>seqinfo</span>, <span class='va'>...</span><span class='op'>)</span><span class='op'>{</span>
  <span class='fu'>RNAmodR</span><span class='fu'>:::</span><span class='fu'><a href='Modifier-class.html'>Modifier</a></span><span class='op'>(</span><span class='st'>"ModExample"</span>, x <span class='op'>=</span> <span class='va'>x</span>, annotation <span class='op'>=</span> <span class='va'>annotation</span>,
                     sequences <span class='op'>=</span> <span class='va'>sequences</span>, seqinfo <span class='op'>=</span> <span class='va'>seqinfo</span>, <span class='va'>...</span><span class='op'>)</span>
<span class='op'>}</span>

<span class='kw'>setMethod</span><span class='op'>(</span>f <span class='op'>=</span> <span class='st'>"aggregateData"</span>,
          signature <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='st'>"ModExample"</span><span class='op'>)</span>,
          definition <span class='op'>=</span>
            <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>force</span> <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span><span class='op'>{</span>
              <span class='co'># Some data with element per transcript</span>
            <span class='op'>}</span>
<span class='op'>)</span>

<span class='kw'>setMethod</span><span class='op'>(</span><span class='st'>"findMod"</span>,
          signature <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='st'>"ModExample"</span><span class='op'>)</span>,
          <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>{</span>
            <span class='co'># an element per modification found.</span>
          <span class='op'>}</span>
<span class='op'>)</span>
<span class='kw'>setMethod</span><span class='op'>(</span>
  f <span class='op'>=</span> <span class='st'>"getDataTrack"</span>,
  signature <span class='op'>=</span> <span class='fu'>signature</span><span class='op'>(</span>x <span class='op'>=</span> <span class='st'>"ModExample"</span><span class='op'>)</span>,
  definition <span class='op'>=</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>name</span>, <span class='va'>type</span>, <span class='va'>...</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='op'>}</span>
<span class='op'>)</span>
<span class='kw'>setMethod</span><span class='op'>(</span>
  f <span class='op'>=</span> <span class='st'>"plotDataByCoord"</span>,
  signature <span class='op'>=</span> <span class='fu'>signature</span><span class='op'>(</span>x <span class='op'>=</span> <span class='st'>"ModExample"</span>, coord <span class='op'>=</span> <span class='st'>"GRanges"</span><span class='op'>)</span>,
  definition <span class='op'>=</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>coord</span>, <span class='va'>type</span> <span class='op'>=</span> <span class='st'>"score"</span>, <span class='va'>window.size</span> <span class='op'>=</span> <span class='fl'>15L</span>, <span class='va'>...</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='op'>}</span>
<span class='op'>)</span>
<span class='kw'>setMethod</span><span class='op'>(</span>
  f <span class='op'>=</span> <span class='st'>"plotData"</span>,
  signature <span class='op'>=</span> <span class='fu'>signature</span><span class='op'>(</span>x <span class='op'>=</span> <span class='st'>"ModExample"</span><span class='op'>)</span>,
  definition <span class='op'>=</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>name</span>, <span class='va'>from</span>, <span class='va'>to</span>, <span class='va'>type</span> <span class='op'>=</span> <span class='st'>"score"</span>, <span class='va'>...</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='op'>}</span>
<span class='op'>)</span>
 
<span class='co'># new ModifierSet class</span>
<span class='kw'>setClass</span><span class='op'>(</span><span class='st'>"ModSetExample"</span>,
         contains <span class='op'>=</span> <span class='st'>"ModifierSet"</span>,
         prototype <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>elementType <span class='op'>=</span> <span class='st'>"ModExample"</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>ModSetExample</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>annotation</span>, <span class='va'>sequences</span>, <span class='va'>seqinfo</span>, <span class='va'>...</span><span class='op'>)</span><span class='op'>{</span>
  <span class='fu'>RNAmodR</span><span class='fu'>:::</span><span class='fu'><a href='ModifierSet-class.html'>ModifierSet</a></span><span class='op'>(</span><span class='st'>"ModExample"</span>, x <span class='op'>=</span> <span class='va'>x</span>, annotation <span class='op'>=</span> <span class='va'>annotation</span>,
                        sequences <span class='op'>=</span> <span class='va'>sequences</span>, seqinfo <span class='op'>=</span> <span class='va'>seqinfo</span>, <span class='va'>...</span><span class='op'>)</span>
<span class='op'>}</span>

<span class='kw'>setMethod</span><span class='op'>(</span>
  f <span class='op'>=</span> <span class='st'>"plotDataByCoord"</span>,
  signature <span class='op'>=</span> <span class='fu'>signature</span><span class='op'>(</span>x <span class='op'>=</span> <span class='st'>"ModSetExample"</span>, coord <span class='op'>=</span> <span class='st'>"GRanges"</span><span class='op'>)</span>,
  definition <span class='op'>=</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>coord</span>, <span class='va'>type</span> <span class='op'>=</span> <span class='st'>"score"</span>, <span class='va'>window.size</span> <span class='op'>=</span> <span class='fl'>15L</span>, <span class='va'>...</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='op'>}</span>
<span class='op'>)</span>
<span class='kw'>setMethod</span><span class='op'>(</span>
  f <span class='op'>=</span> <span class='st'>"plotData"</span>,
  signature <span class='op'>=</span> <span class='fu'>signature</span><span class='op'>(</span>x <span class='op'>=</span> <span class='st'>"ModSetExample"</span><span class='op'>)</span>,
  definition <span class='op'>=</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>name</span>, <span class='va'>from</span>, <span class='va'>to</span>, <span class='va'>type</span> <span class='op'>=</span> <span class='st'>"score"</span>, <span class='va'>...</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='op'>}</span>
<span class='op'>)</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


