<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>The SequenceDataFrame class — SequenceDataFrame-class • RNAmodR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="The SequenceDataFrame class — SequenceDataFrame-class"><meta property="og:description" content="The SequenceDataFrame class is a virtual class and  contains data for
positions along a single transcript. In addition to being used for returning
elements from a SequenceData object, the SequenceDataFrame class is
used to store the unlisted data within a
SequenceData object. Therefore, a matching
SequenceData and SequenceDataFrame class must be implemented.
The SequenceDataFrame class is derived from the
DataFrame class. To follow the 
functionallity in the S4Vectors package, SequenceDataFrame 
implements the concept, whereas SequenceDFrame is the implementation
for in-memory data representation from which some specific 
*SequenceDataFrame class derive from, e.g. 
CoverageSequenceData.
Subsetting of a SequenceDataFrame returns a SequenceDataFrame or 
DataFrame, if it is subset by a column or row, respectively. The 
drop argument is ignored for column subsetting."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.17.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/FelixErnst/RNAmodR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>The SequenceDataFrame class</h1>
    <small class="dont-index">Source: <a href="https://github.com/FelixErnst/RNAmodR/blob/HEAD/R/SequenceDataFrame-class.R" class="external-link"><code>R/SequenceDataFrame-class.R</code></a></small>
    <div class="hidden name"><code>SequenceDataFrame-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>SequenceDataFrame</code> class is a virtual class and  contains data for
positions along a single transcript. In addition to being used for returning
elements from a <code>SequenceData</code> object, the SequenceDataFrame class is
used to store the unlisted data within a
<code><a href="SequenceData-class.html">SequenceData</a></code> object. Therefore, a matching
<code>SequenceData</code> and <code>SequenceDataFrame</code> class must be implemented.</p>
<p>The <code>SequenceDataFrame</code> class is derived from the
<code><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></code> class. To follow the 
functionallity in the <code>S4Vectors</code> package, <code>SequenceDataFrame</code> 
implements the concept, whereas <code>SequenceDFrame</code> is the implementation
for in-memory data representation from which some specific 
<code>*SequenceDataFrame</code> class derive from, e.g. 
<code><a href="CoverageSequenceData-class.html">CoverageSequenceData</a></code>.</p>
<p>Subsetting of a <code>SequenceDataFrame</code> returns a <code>SequenceDataFrame</code> or 
<code>DataFrame</code>, if it is subset by a column or row, respectively. The 
<code>drop</code> argument is ignored for column subsetting.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre><code># S4 method for SequenceDataFrame
cbind(..., deparse.level = 1)

# S4 method for SequenceDataFrame,ANY,ANY,ANY
[(x, i, j, ..., drop = TRUE)</code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x, i, j, ..., drop, deparse.level</dt>
<dd><p>arguments used for 
<code><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">subsetting</a></code> or 
<code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">base::cbind</a></code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A <code>SequenceDataFrame</code> object or if subset to row a 
<code>DataFrame</code></p>


    </div>
    <div id="slots">
    <h2>Slots</h2>
    

<dl><dt><code>ranges</code></dt>
<dd><p>a <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> 
object each element describing a transcript including its element. The 
<code>GRanges</code> is constructed from the unlisted results of the
<code><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html" class="external-link">exonsBy(x, by="tx")</a></code> function.
If during construction a <code>GRangesList</code> is provided instead of a 
character value pointing to a gff3 file or a <code>TxDb</code> object, it must have
a comparable structure.</p></dd>


<dt><code>sequence</code></dt>
<dd><p>a <code><a href="https://rdrr.io/pkg/Biostrings/man/XString-class.html" class="external-link">XString</a></code> of 
type <code>sequencesType</code> from the parent 
<code><a href="SequenceData-class.html">SequenceData</a></code> object.</p></dd>


<dt><code>condition</code></dt>
<dd><p>conditions along the 
<code><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFileList</a></code>: Either <code>control</code>
or <code>treated</code></p></dd>


<dt><code>replicate</code></dt>
<dd><p>replicate number along the <code>BamFileList</code> for each of the
condition types.</p></dd>


<dt><code>bamfiles</code></dt>
<dd><p>the input bam files as 
<code><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFileList</a></code></p></dd>


<dt><code>seqinfo</code></dt>
<dd><p>a <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/Seqinfo-class.html" class="external-link">Seqinfo</a></code> describing
the avialable/used chromosomes.</p></dd>


</dl></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>for an example see
<code><a href="ProtectedEndSequenceData-class.html">ProtectedEndSequenceData</a></code>
and for more information see <code><a href="SequenceData-class.html">SequenceData</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">e5sd</span>,package<span class="op">=</span><span class="st">"RNAmodR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># A SequenceDataFrame can is usually constructed by subsetting from </span></span></span>
<span class="r-in"><span><span class="co"># a SequenceData object</span></span></span>
<span class="r-in"><span><span class="va">sdf</span> <span class="op">&lt;-</span> <span class="va">e5sd</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="co"># Its also used to store the unlisted data in a SequenceData object</span></span></span>
<span class="r-in"><span><span class="va">sdf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">e5sd</span><span class="op">)</span> <span class="co"># should probably only used internally</span></span></span>
<span class="r-in"><span><span class="va">e5sd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/relist.html" class="external-link">relist</a></span><span class="op">(</span><span class="va">sdf</span>,<span class="va">e5sd</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

