<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>The SequenceData class — SequenceData-class • RNAmodR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="The SequenceData class — SequenceData-class" />
<meta property="og:description" content="The SequenceData class is implemented to contain data on each position
along transcripts and holds the corresponding annotation data and
nucleotide sequence of these transcripts. To access this data several
functions are available. The
SequenceData class is a virtual class, from which specific classes can
be extended. Currently the following classes are implemented:

CoverageSequenceData
End5SequenceData,
End3SequenceData,
EndSequenceData
NormEnd5SequenceData,
NormEnd5SequenceData
PileupSequenceData
ProtectedEndSequenceData


The annotation and sequence data can be accessed through the functions
ranges and sequences, respectively. Beaware, that the data is
always provided according to genomic positions with increasing
rownames, but the sequence is given as the actual sequence of the
transcript. Therefore, it is necessary to treat the minus strand accordingly.
The SequenceData class is derived from the
CompressedSplitDataFrameList class
with additional slots for annotation and sequence data. Some functionality is
not inherited and might not available to full extend, e.g.relist.
SequenceDataFrame
The SequenceDataFrame class is a virtual class and  contains data for
positions along a single transcript. In addition to being used for returning
elements from a SequenceData object, the SequenceDataFrame class is
used to store the unlisted data within a
SequenceData object. Therefore, a matching
SequenceData and SequenceDataFrame class must be implemented.
The SequenceDataFrame class is derived from the
DataFrame class.
Subsetting of a SequenceDataFrame returns a SequenceDataFrame or
DataFrame, if it is subset by a column or row, respectively. The
drop argument is ignored for column subsetting." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/FelixErnst/RNAmodR">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>The SequenceData class</h1>
    <small class="dont-index">Source: <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/SequenceData-class.R'><code>R/SequenceData-class.R</code></a></small>
    <div class="hidden name"><code>SequenceData-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>SequenceData</code> class is implemented to contain data on each position
along transcripts and holds the corresponding annotation data and
nucleotide sequence of these transcripts. To access this data several
<code><a href='SequenceData-functions.html'>functions</a></code> are available. The
<code>SequenceData</code> class is a virtual class, from which specific classes can
be extended. Currently the following classes are implemented:</p>
<ul>
<li><p><code><a href='CoverageSequenceData-class.html'>CoverageSequenceData</a></code></p></li>
<li><p><code><a href='EndSequenceData-class.html'>End5SequenceData</a></code>,
<code><a href='EndSequenceData-class.html'>End3SequenceData</a></code>,
<code><a href='EndSequenceData-class.html'>EndSequenceData</a></code></p></li>
<li><p><code><a href='NormEndSequenceData-class.html'>NormEnd5SequenceData</a></code>,
<code><a href='NormEndSequenceData-class.html'>NormEnd5SequenceData</a></code></p></li>
<li><p><code><a href='PileupSequenceData-class.html'>PileupSequenceData</a></code></p></li>
<li><p><code><a href='ProtectedEndSequenceData-class.html'>ProtectedEndSequenceData</a></code></p></li>
</ul>

<p>The annotation and sequence data can be accessed through the functions
<code>ranges</code> and <code>sequences</code>, respectively. Beaware, that the data is
always provided according to genomic positions with increasing
<code>rownames</code>, but the sequence is given as the actual sequence of the
transcript. Therefore, it is necessary to treat the minus strand accordingly.</p>
<p>The <code>SequenceData</code> class is derived from the
<code><a href='https://rdrr.io/pkg/IRanges/man/DataFrameList-class.html'>CompressedSplitDataFrameList</a></code> class
with additional slots for annotation and sequence data. Some functionality is
not inherited and might not available to full extend, e.g.<code>relist</code>.</p>
<p><strong>SequenceDataFrame</strong></p>
<p>The <code>SequenceDataFrame</code> class is a virtual class and  contains data for
positions along a single transcript. In addition to being used for returning
elements from a <code>SequenceData</code> object, the SequenceDataFrame class is
used to store the unlisted data within a
<code>SequenceData</code> object. Therefore, a matching
<code>SequenceData</code> and <code>SequenceDataFrame</code> class must be implemented.</p>
<p>The <code>SequenceDataFrame</code> class is derived from the
<code><a href='https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html'>DataFrame</a></code> class.</p>
<p>Subsetting of a <code>SequenceDataFrame</code> returns a <code>SequenceDataFrame</code> or
<code>DataFrame</code>, if it is subset by a column or row, respectively. The
<code>drop</code> argument is ignored for column subsetting.</p>
    </div>

    <pre class="usage"><span class='co'># S4 method for SequenceData</span>
<span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span>(<span class='no'>...</span>, <span class='kw'>deparse.level</span> <span class='kw'>=</span> <span class='fl'>1</span>)

<span class='co'># S4 method for SequenceData</span>
<span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>rbind</a></span>(<span class='no'>...</span>, <span class='kw'>deparse.level</span> <span class='kw'>=</span> <span class='fl'>1</span>)

<span class='fu'>SequenceData</span>(<span class='no'>dataType</span>, <span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='co'># S4 method for character,character</span>
<span class='fu'>SequenceData</span>(<span class='no'>dataType</span>, <span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='co'># S4 method for character,BSgenome</span>
<span class='fu'>SequenceData</span>(<span class='no'>dataType</span>, <span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='co'># S4 method for TxDb,character</span>
<span class='fu'>SequenceData</span>(<span class='no'>dataType</span>, <span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='co'># S4 method for TxDb,BSgenome</span>
<span class='fu'>SequenceData</span>(<span class='no'>dataType</span>, <span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='co'># S4 method for GRangesList,character</span>
<span class='fu'>SequenceData</span>(<span class='no'>dataType</span>, <span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='co'># S4 method for GRangesList,BSgenome</span>
<span class='fu'>SequenceData</span>(<span class='no'>dataType</span>, <span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='co'># S4 method for GFF3File,BSgenome</span>
<span class='fu'>SequenceData</span>(<span class='no'>dataType</span>, <span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='co'># S4 method for GFF3File,character</span>
<span class='fu'>SequenceData</span>(<span class='no'>dataType</span>, <span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='co'># S4 method for character,FaFile</span>
<span class='fu'>SequenceData</span>(<span class='no'>dataType</span>, <span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='co'># S4 method for GFF3File,FaFile</span>
<span class='fu'>SequenceData</span>(<span class='no'>dataType</span>, <span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='co'># S4 method for TxDb,FaFile</span>
<span class='fu'>SequenceData</span>(<span class='no'>dataType</span>, <span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)

<span class='co'># S4 method for GRangesList,FaFile</span>
<span class='fu'>SequenceData</span>(<span class='no'>dataType</span>, <span class='no'>bamfiles</span>, <span class='no'>annotation</span>, <span class='no'>sequences</span>, <span class='no'>seqinfo</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>...</th>
      <td><p>Optional arguments overwriting default values. Not all
<code>SequenceData</code> classes use all arguments. The arguments are:</p><ul>
<li><p><code>minLength</code> single integer value setting a threshold for minimum
read length. Shorther reads are discarded (default: <code>minLength = NA</code>).</p></li>
<li><p><code>maxLength</code> single integer value setting a threshold for maximum
read length. Longer reads are discarded (default: <code>maxLength = NA</code>).</p></li>
<li><p><code>minQuality</code> single integer value setting a threshold for maximum
read quality. Reads with a lower quality are discarded (default:
<code>minQuality = 5L</code>, but this is class dependent).</p></li>
<li><p><code>max_depth</code> maximum depth for pileup loading (default:
<code>max_depth = 10000L</code>).</p></li>
</ul></td>
    </tr>
    <tr>
      <th>deparse.level</th>
      <td><p>See <code><a href='https://rdrr.io/r/base/cbind.html'>base::cbind</a></code> for a
description of this argument.</p></td>
    </tr>
    <tr>
      <th>dataType</th>
      <td><p>The prefix for construction the class name of the
<code>SequenceData</code> subclass to be constructed.</p></td>
    </tr>
    <tr>
      <th>bamfiles</th>
      <td><p>the input which can be of the following types</p><ul>
<li><p><code>BamFileList</code>: a named <code>BamFileList</code></p></li>
<li><p><code>character</code>: a <code>character</code> vector, which must be coercible
to a named <code>BamFileList</code> referencing existing bam files. Valid names are
<code>control</code> and <code>treated</code> to define conditions and replicates</p></li>
</ul></td>
    </tr>
    <tr>
      <th>annotation</th>
      <td><p>annotation data, which must match the information contained
in the BAM files.</p></td>
    </tr>
    <tr>
      <th>sequences</th>
      <td><p>sequences matching the target sequences the reads were
mapped onto. This must match the information contained in the BAM files.</p></td>
    </tr>
    <tr>
      <th>seqinfo</th>
      <td><p>optional <code><a href='https://rdrr.io/pkg/GenomeInfoDb/man/Seqinfo-class.html'>Seqinfo</a></code> to
subset the transcripts analyzed on a chromosome basis.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A SequenceData object</p>
    <h2 class="hasAnchor" id="slots"><a class="anchor" href="#slots"></a>Slots</h2>

    

<dl class='dl-horizontal'>
<dt><code>sequencesType</code></dt><dd><p>a <code>character</code> value for the class name of
<code>sequences</code>. Either <code>RNAStringSet</code>, <code>ModRNAStringSet</code>,
<code>DNAStringSet</code> or <code>ModDNAStringSet</code>.</p></dd>

<dt><code>minQuality</code></dt><dd><p>a <code>integer</code> value describing a threshold of the minimum
quality of reads to be used.</p></dd>

</dl>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#slots">Slots</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


