<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Retrieving information about used reads in RNAmodR — stats • RNAmodR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Retrieving information about used reads in RNAmodR — stats" />
<meta property="og:description" content="stats returns information about reads used in the RNAmodR analysis.
Three modes are available depending on which type of object is provided. If a
SequenceData object is provided, a
BamFile or
BamFileList must be provided as well. If a
Modifier object is used, the bam files
returned from the bamfiles function are used. This is also the case,
if a ModifierSet object is used." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/FelixErnst/RNAmodR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Retrieving information about used reads in RNAmodR</h1>
    <small class="dont-index">Source: <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/SequenceData-stats.R'><code>R/SequenceData-stats.R</code></a></small>
    <div class="hidden name"><code>stats.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>stats</code> returns information about reads used in the RNAmodR analysis.
Three modes are available depending on which type of object is provided. If a
<code><a href='SequenceData-class.html'>SequenceData</a></code> object is provided, a
<code><a href='https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html'>BamFile</a></code> or
<code><a href='https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html'>BamFileList</a></code> must be provided as well. If a
<code><a href='Modifier-class.html'>Modifier</a></code> object is used, the bam files
returned from the <code>bamfiles</code> function are used. This is also the case,
if a <code><a href='ModifierSet-class.html'>ModifierSet</a></code> object is used.</p>
    </div>

    <pre class="usage"><span class='fu'>stats</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>file</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S4 method for SequenceData,BamFile</span>
<span class='fu'>stats</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>file</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S4 method for SequenceData,BamFileList</span>
<span class='fu'>stats</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>file</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S4 method for Modifier,missing</span>
<span class='fu'>stats</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for ModifierSet,missing</span>
<span class='fu'>stats</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>a <code><a href='SequenceData-class.html'>SequenceData</a></code>, 
<code><a href='Modifier-class.html'>Modifier</a></code> or 
<code><a href='ModifierSet-class.html'>ModifierSet</a></code> object</p></td>
    </tr>
    <tr>
      <th>file</th>
      <td><p>a <code><a href='https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html'>BamFile</a></code> or 
<code><a href='https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html'>BamFileList</a></code>, if <code>x</code> is a 
<code><a href='SequenceData-class.html'>SequenceData</a></code> object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>optional parameters used as stated 
<code><a href='SequenceData-class.html'>here</a></code> (except <code>minQuality</code>),
if <code>x</code> is a <code><a href='SequenceData-class.html'>SequenceData</a></code> object.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a <code>DataFrame</code>, <code>DataFrameList</code> or <code>SimpleList</code> with 
  the results in aggregated form</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/FelixErnst/RNAmodR.Data'>RNAmodR.Data</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>rtracklayer</span><span class='op'>)</span>
<span class='va'>sequences</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.AAS.fasta</a></span><span class='op'>(</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2021-01-11</span></div><div class='output co'>#&gt; <span class='message'>see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span></div><div class='output co'>#&gt; <span class='message'>downloading 2 resources</span></div><div class='output co'>#&gt; <span class='message'>retrieving 2 resources</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='input'><span class='va'>annotation</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html'>GFF3File</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.AAS.gff3</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2021-01-11</span></div><div class='output co'>#&gt; <span class='message'>see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span></div><div class='output co'>#&gt; <span class='message'>downloading 1 resources</span></div><div class='output co'>#&gt; <span class='message'>retrieving 1 resource</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='input'><span class='va'>files</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='st'>"SampleSet1"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>treated <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.wt.1</a></span><span class='op'>(</span><span class='op'>)</span>,
                               treated <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.wt.2</a></span><span class='op'>(</span><span class='op'>)</span>,
                               treated <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.wt.3</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>,
              <span class='st'>"SampleSet2"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>treated <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.bud23.1</a></span><span class='op'>(</span><span class='op'>)</span>,
                               treated <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.bud23.2</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>,
              <span class='st'>"SampleSet3"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>treated <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.trm8.1</a></span><span class='op'>(</span><span class='op'>)</span>,
                               treated <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html'>RNAmodR.Data.example.trm8.2</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2021-01-11</span></div><div class='output co'>#&gt; <span class='message'>see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2021-01-11</span></div><div class='output co'>#&gt; <span class='message'>see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span></div><div class='output co'>#&gt; <span class='message'>downloading 2 resources</span></div><div class='output co'>#&gt; <span class='message'>retrieving 2 resources</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2021-01-11</span></div><div class='output co'>#&gt; <span class='message'>see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span></div><div class='output co'>#&gt; <span class='message'>downloading 2 resources</span></div><div class='output co'>#&gt; <span class='message'>retrieving 2 resources</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2021-01-11</span></div><div class='output co'>#&gt; <span class='message'>see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span></div><div class='output co'>#&gt; <span class='message'>downloading 2 resources</span></div><div class='output co'>#&gt; <span class='message'>retrieving 2 resources</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2021-01-11</span></div><div class='output co'>#&gt; <span class='message'>see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span></div><div class='output co'>#&gt; <span class='message'>downloading 2 resources</span></div><div class='output co'>#&gt; <span class='message'>retrieving 2 resources</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2021-01-11</span></div><div class='output co'>#&gt; <span class='message'>see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span></div><div class='output co'>#&gt; <span class='message'>downloading 2 resources</span></div><div class='output co'>#&gt; <span class='message'>retrieving 2 resources</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2021-01-11</span></div><div class='output co'>#&gt; <span class='message'>see ?RNAmodR.Data and browseVignettes('RNAmodR.Data') for documentation</span></div><div class='output co'>#&gt; <span class='message'>downloading 2 resources</span></div><div class='output co'>#&gt; <span class='message'>retrieving 2 resources</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='input'><span class='va'>msi</span> <span class='op'>&lt;-</span> <span class='fu'><a href='ModInosine.html'>ModSetInosine</a></span><span class='op'>(</span><span class='va'>files</span>, annotation <span class='op'>=</span> <span class='va'>annotation</span>, sequences <span class='op'>=</span> <span class='va'>sequences</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Import genomic features from the file as a GRanges object ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Prepare the 'metadata' data frame ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Make the TxDb object ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>1. ModInosine analysis 'SampleSet1':</span></div><div class='output co'>#&gt; <span class='message'>Loading Pileup data from BAM files ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Aggregating data and calculating scores ... </span></div><div class='output co'>#&gt; <span class='message'>Starting to search for 'Inosine' ... </span></div><div class='output co'>#&gt; <span class='message'>done.</span></div><div class='output co'>#&gt; <span class='message'>2. ModInosine analysis 'SampleSet2':</span></div><div class='output co'>#&gt; <span class='message'>Loading Pileup data from BAM files ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Aggregating data and calculating scores ... </span></div><div class='output co'>#&gt; <span class='message'>Starting to search for 'Inosine' ... </span></div><div class='output co'>#&gt; <span class='message'>done.</span></div><div class='output co'>#&gt; <span class='message'>3. ModInosine analysis 'SampleSet3':</span></div><div class='output co'>#&gt; <span class='message'>Loading Pileup data from BAM files ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Aggregating data and calculating scores ... </span></div><div class='output co'>#&gt; <span class='message'>Starting to search for 'Inosine' ... </span></div><div class='output co'>#&gt; <span class='message'>done.</span></div><div class='input'><span class='co'># smallest chunk of information</span>
<span class='fu'>stats</span><span class='op'>(</span><span class='fu'><a href='Modifier-functions.html'>sequenceData</a></span><span class='op'>(</span><span class='va'>msi</span><span class='op'>[[</span><span class='fl'>1L</span><span class='op'>]</span><span class='op'>]</span><span class='op'>)</span>,<span class='fu'><a href='Modifier-functions.html'>bamfiles</a></span><span class='op'>(</span><span class='va'>msi</span><span class='op'>[[</span><span class='fl'>1L</span><span class='op'>]</span><span class='op'>]</span><span class='op'>)</span><span class='op'>[[</span><span class='fl'>1L</span><span class='op'>]</span><span class='op'>]</span><span class='op'>)</span>
</div><div class='output co'>#&gt; DataFrame with 11 rows and 6 columns
#&gt;    seqnames seqlength    mapped  unmapped          used     used_distro
#&gt;    &lt;factor&gt; &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;IntegerList&gt;          &lt;List&gt;
#&gt; 1     chr1       1800    197050         0        159782 83,1252,860,...
#&gt; 2     chr2         85      5863         0          2459     2,16,16,...
#&gt; 3     chr3         76     76905         0         63497 35,478,4106,...
#&gt; 4     chr4         77      8299         0          6554     6,27,36,...
#&gt; 5     chr5         74     11758         0          8818  520,105,93,...
#&gt; 6     chr6         76      3504         0          2427     1,38,45,...
#&gt; 7     chr7         75     14556         0         12243  1532,46,12,...
#&gt; 8     chr8         75    144293         0        143068    14,44,48,...
#&gt; 9     chr9         75     13790         0          9753     1,49,43,...
#&gt; 10    chr10        85     19861         0         17729    35,21,10,...
#&gt; 11    chr11        77     10532         0          9086   53,92,185,...</div><div class='input'><span class='co'># partial information</span>
<span class='fu'>stats</span><span class='op'>(</span><span class='fu'><a href='Modifier-functions.html'>sequenceData</a></span><span class='op'>(</span><span class='va'>msi</span><span class='op'>[[</span><span class='fl'>1L</span><span class='op'>]</span><span class='op'>]</span><span class='op'>)</span>,<span class='fu'><a href='Modifier-functions.html'>bamfiles</a></span><span class='op'>(</span><span class='va'>msi</span><span class='op'>[[</span><span class='fl'>1L</span><span class='op'>]</span><span class='op'>]</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; DataFrameList of length 3
#&gt; names(3): treated treated treated</div><div class='input'><span class='co'># the whole stats</span>
<span class='fu'>stats</span><span class='op'>(</span><span class='va'>msi</span><span class='op'>)</span>
</div><div class='output co'>#&gt; List of length 3
#&gt; names(3): SampleSet1 SampleSet2 SampleSet3</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


