<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Subsetting data from a SequenceData, SequenceDataSet,
SequenceDataList, Modifier or ModifierSet object. — subsetByCoord • RNAmodR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Subsetting data from a SequenceData, SequenceDataSet,
SequenceDataList, Modifier or ModifierSet object. — subsetByCoord"><meta property="og:description" content="With the subsetByCoord function data from a SequenceData,
SequenceDataSet, SequenceDataList, Modifier or
ModifierSet object can be subset to positions as defined in
coord.
If coord contains a column mod and x is a
Modifier object, it will be filtered to identifiers matching the
modType of x. To disable this
behaviour remove the column mod from coord or set type =
NA
labelByCoord functions similarly. It will return a
SplitDataFrameList, which matches the dimensions of the aggregated
data plus the labels column, which contains logical values to indicate
selected positions."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.17.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/FelixErnst/RNAmodR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Subsetting data from a <code>SequenceData</code>, <code>SequenceDataSet</code>,
<code>SequenceDataList</code>, <code>Modifier</code> or <code>ModifierSet</code> object.</h1>
    <small class="dont-index">Source: <a href="https://github.com/FelixErnst/RNAmodR/blob/HEAD/R/AllGenerics.R" class="external-link"><code>R/AllGenerics.R</code></a>, <a href="https://github.com/FelixErnst/RNAmodR/blob/HEAD/R/Modifier-subset.R" class="external-link"><code>R/Modifier-subset.R</code></a>, <a href="https://github.com/FelixErnst/RNAmodR/blob/HEAD/R/SequenceData-subset.R" class="external-link"><code>R/SequenceData-subset.R</code></a></small>
    <div class="hidden name"><code>subsetByCoord.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>With the <code>subsetByCoord</code> function data from a <code>SequenceData</code>,
<code>SequenceDataSet</code>, <code>SequenceDataList</code>, <code>Modifier</code> or
<code>ModifierSet</code> object can be subset to positions as defined in
<code>coord</code>.</p>
<p>If <code>coord</code> contains a column <code>mod</code> and <code>x</code> is a
<code>Modifier</code> object, it will be filtered to identifiers matching the
<code><a href="Modifier-functions.html">modType</a></code> of <code>x</code>. To disable this
behaviour remove the column <code>mod</code> from <code>coord</code> or set <code>type =
NA</code></p>
<p><code>labelByCoord</code> functions similarly. It will return a
<code>SplitDataFrameList</code>, which matches the dimensions of the aggregated
data plus the <code>labels</code> column, which contains logical values to indicate
selected positions.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">subsetByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">labelByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for Modifier,GRanges</span></span>
<span><span class="fu">subsetByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for Modifier,GRangesList</span></span>
<span><span class="fu">subsetByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for ModifierSet</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, pos <span class="op">=</span> <span class="fl">1L</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for ModifierSet,GRanges</span></span>
<span><span class="fu">subsetByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for ModifierSet,GRangesList</span></span>
<span><span class="fu">subsetByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for Modifier,GRanges</span></span>
<span><span class="fu">labelByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for Modifier,GRangesList</span></span>
<span><span class="fu">labelByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for ModifierSet,GRanges</span></span>
<span><span class="fu">labelByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for ModifierSet,GRangesList</span></span>
<span><span class="fu">labelByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SplitDataFrameList,GRanges</span></span>
<span><span class="fu">subsetByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceData</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, pos <span class="op">=</span> <span class="fl">1L</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceData,GRanges</span></span>
<span><span class="fu">subsetByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceData,GRangesList</span></span>
<span><span class="fu">subsetByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceDataSet</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, pos <span class="op">=</span> <span class="fl">1L</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceDataSet,GRanges</span></span>
<span><span class="fu">subsetByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceDataSet,GRangesList</span></span>
<span><span class="fu">subsetByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceDataList</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, pos <span class="op">=</span> <span class="fl">1L</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceDataList,GRanges</span></span>
<span><span class="fu">subsetByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceDataList,GRangesList</span></span>
<span><span class="fu">subsetByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceData,GRanges</span></span>
<span><span class="fu">labelByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceData,GRangesList</span></span>
<span><span class="fu">labelByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceDataSet,GRanges</span></span>
<span><span class="fu">labelByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceDataSet,GRangesList</span></span>
<span><span class="fu">labelByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceDataList,GRanges</span></span>
<span><span class="fu">labelByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SequenceDataList,GRangesList</span></span>
<span><span class="fu">labelByCoord</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p>a <code>SequenceData</code>, <code>SequenceDataSet</code>,
<code>SequenceDataList</code>, <code>Modifier</code> or <code>ModifierSet</code> object.</p></dd>


<dt>coord</dt>
<dd><p>coordinates of position to subset to. Either a <code>GRanges</code> or
a <code>GRangesList</code> object. For both types the 'Parent' column is expected to
match the transcript name.</p></dd>


<dt>...</dt>
<dd><p>Optional parameters:</p><ul><li><p><code>type</code>: the modification type used for subsetting. By default 
this is derived from the <code>modType(x)</code>, but it can be overwritten using 
<code>type</code>. It must be a valid shortName for a modification according to
<code>shortName(ModRNAString())</code> or <code>shortName(ModDNAString())</code> 
(depending on the type of Modifier class) and of course be present in 
metadata column <code>mod</code> of <code>coord</code>. To disable subsetting based on 
type, set <code>type = NA</code>.</p></li>
<li><p><code>flanking</code>: a single integer value to select how many flanking
position should be included in the subset (default: <code>flanking = 0L</code>).</p></li>
<li><p><code>merge</code>: <code>TRUE</code> or <code>FALSE</code>: Should the 
overlapping selections be merged? This is particular important, if flanking
value <code>!= 0L</code> are set. (default: <code>merge = TRUE</code>).</p></li>
<li><p><code>perTranscript</code>: <code>TRUE</code> or <code>FALSE</code>: Should the 
positions labeled per transcript and not per chromosome?
(default: <code>perTranscript = FALSE</code>).</p></li>
</ul></dd>


<dt>name</dt>
<dd><p>Optional: Limit results to one specific transcript.</p></dd>


<dt>pos</dt>
<dd><p>Optional: Limit results to a specific position.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>If 'x' is a</p><ul><li><p><code><a href="SequenceData-class.html">SequenceData</a></code> or 
<code><a href="Modifier-class.html">Modifier</a></code>: a <code>SplitDataFrameList</code>
with elments per transcript.</p></li>
<li><p><code><a href="SequenceDataSet-class.html">SequenceDataSet</a></code>,
<code><a href="SequenceDataList-class.html">SequenceDataList</a></code> or
<code><a href="ModifierSet-class.html">ModifierSet</a></code>: a <code>SimpleList</code> of
<code>SplitDataFrameList</code> with elments per transcript.</p></li>
</ul></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">msi</span>,package<span class="op">=</span><span class="st">"RNAmodR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="modify.html">modifications</a></span><span class="op">(</span><span class="va">msi</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">coord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">mod</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">coord</span><span class="op">$</span><span class="va">score</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></span>
<span class="r-in"><span><span class="va">coord</span><span class="op">$</span><span class="va">sd</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></span>
<span class="r-in"><span><span class="fu">subsetByCoord</span><span class="op">(</span><span class="va">msi</span>,<span class="va">coord</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $SampleSet1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SplitDataFrameList of length 6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`2`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 34  0.900932</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`4`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 35  0.899622</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`6`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 34  0.984035</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`7`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 67  0.934553</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`9`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7  0.709758</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`11`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 35  0.874027</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $SampleSet2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SplitDataFrameList of length 6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`2`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 34  0.998134</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`4`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 35  0.856241</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`6`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 34  0.992012</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`7`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 67  0.942905</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`9`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7  0.766484</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`11`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 35  0.971474</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $SampleSet3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SplitDataFrameList of length 6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`2`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 34  0.953651</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`4`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 35  0.976928</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`6`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 34  0.993128</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`7`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 67  0.943773</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`9`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7  0.681451</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`11`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 1 column</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 35  0.954782</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

