<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Subsetting data from a <code>SequenceData</code>, <code>SequenceDataSet</code>,
<code>SequenceDataList</code>, <code>Modifier</code> or <code>ModifierSet</code> object. — subsetByCoord • RNAmodR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Subsetting data from a <code>SequenceData</code>, <code>SequenceDataSet</code>,
<code>SequenceDataList</code>, <code>Modifier</code> or <code>ModifierSet</code> object. — subsetByCoord" />
<meta property="og:description" content="With the subsetByCoord function data from a SequenceData,
SequenceDataSet, SequenceDataList, Modifier or
ModifierSet object can be subset to positions as defined in
coord.
If coord contains a column mod and x is a
Modifier object, it will be filtered to identifiers matching the
modType of x. To disable this
behaviour remove the column mod from coord or set type =
NA
labelByCoord functions similarly. It will return a
SplitDataFrameList, which matches the dimensions of the aggregated
data plus the labels column, which contains logical values to indicate
selected positions." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/FelixErnst/RNAmodR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Subsetting data from a <code>SequenceData</code>, <code>SequenceDataSet</code>,
<code>SequenceDataList</code>, <code>Modifier</code> or <code>ModifierSet</code> object.</h1>
    <small class="dont-index">Source: <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/Modifier-subset.R'><code>R/Modifier-subset.R</code></a>, <a href='https://github.com/FelixErnst/RNAmodR/blob/master/R/SequenceData-subset.R'><code>R/SequenceData-subset.R</code></a></small>
    <div class="hidden name"><code>subsetByCoord.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>With the <code>subsetByCoord</code> function data from a <code>SequenceData</code>,
<code>SequenceDataSet</code>, <code>SequenceDataList</code>, <code>Modifier</code> or
<code>ModifierSet</code> object can be subset to positions as defined in
<code>coord</code>.</p>
<p>If <code>coord</code> contains a column <code>mod</code> and <code>x</code> is a
<code>Modifier</code> object, it will be filtered to identifiers matching the
<code><a href='Modifier-functions.html'>modType</a></code> of <code>x</code>. To disable this
behaviour remove the column <code>mod</code> from <code>coord</code> or set <code>type =
NA</code></p>
<p><code>labelByCoord</code> functions similarly. It will return a
<code>SplitDataFrameList</code>, which matches the dimensions of the aggregated
data plus the <code>labels</code> column, which contains logical values to indicate
selected positions.</p>
    </div>

    <pre class="usage"><span class='fu'>subsetByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='fu'>labelByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for Modifier,GRanges</span>
<span class='fu'>subsetByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for Modifier,GRangesList</span>
<span class='fu'>subsetByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for ModifierSet</span>
<span class='fu'><a href='https://rdrr.io/r/base/subset.html'>subset</a></span>(<span class='no'>x</span>, <span class='no'>name</span>, <span class='kw'>pos</span> <span class='kw'>=</span> <span class='fl'>1L</span>, <span class='no'>...</span>)

<span class='co'># S4 method for ModifierSet,GRanges</span>
<span class='fu'>subsetByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for ModifierSet,GRangesList</span>
<span class='fu'>subsetByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for Modifier,GRanges</span>
<span class='fu'>labelByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for Modifier,GRangesList</span>
<span class='fu'>labelByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for ModifierSet,GRanges</span>
<span class='fu'>labelByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for ModifierSet,GRangesList</span>
<span class='fu'>labelByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SplitDataFrameList,GRanges</span>
<span class='fu'>subsetByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SequenceData</span>
<span class='fu'><a href='https://rdrr.io/r/base/subset.html'>subset</a></span>(<span class='no'>x</span>, <span class='no'>name</span>, <span class='kw'>pos</span> <span class='kw'>=</span> <span class='fl'>1L</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SequenceData,GRanges</span>
<span class='fu'>subsetByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SequenceData,GRangesList</span>
<span class='fu'>subsetByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SequenceDataSet</span>
<span class='fu'><a href='https://rdrr.io/r/base/subset.html'>subset</a></span>(<span class='no'>x</span>, <span class='no'>name</span>, <span class='kw'>pos</span> <span class='kw'>=</span> <span class='fl'>1L</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SequenceDataSet,GRanges</span>
<span class='fu'>subsetByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SequenceDataSet,GRangesList</span>
<span class='fu'>subsetByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SequenceDataList</span>
<span class='fu'><a href='https://rdrr.io/r/base/subset.html'>subset</a></span>(<span class='no'>x</span>, <span class='no'>name</span>, <span class='kw'>pos</span> <span class='kw'>=</span> <span class='fl'>1L</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SequenceDataList,GRanges</span>
<span class='fu'>subsetByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SequenceDataList,GRangesList</span>
<span class='fu'>subsetByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SequenceData,GRanges</span>
<span class='fu'>labelByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SequenceData,GRangesList</span>
<span class='fu'>labelByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SequenceDataSet,GRanges</span>
<span class='fu'>labelByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SequenceDataSet,GRangesList</span>
<span class='fu'>labelByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SequenceDataList,GRanges</span>
<span class='fu'>labelByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SequenceDataList,GRangesList</span>
<span class='fu'>labelByCoord</span>(<span class='no'>x</span>, <span class='no'>coord</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>a <code>SequenceData</code>, <code>SequenceDataSet</code>,
<code>SequenceDataList</code>, <code>Modifier</code> or <code>ModifierSet</code> object.</p></td>
    </tr>
    <tr>
      <th>coord</th>
      <td><p>coordinates of position to subset to. Either a <code>GRanges</code> or
a <code>GRangesList</code> object. For both types the 'Parent' column is expected to
match the transcript name.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Optional parameters:</p><ul>
<li><p><code>type</code>: the modification type used for subsetting. By default 
this is derived from the <code><a href='Modifier-functions.html'>modType(x)</a></code>, but it can be overwritten using 
<code>type</code>. It must be a valid shortName for a modification according to
<code>shortName(ModRNAString())</code> or <code>shortName(ModDNAString())</code> 
(depending on the type of Modifier class) and of course be present in 
metadata column <code>mod</code> of <code>coord</code>. To disable subsetting based on 
type, set <code>type = NA</code>.</p></li>
<li><p><code>flanking</code>: a single integer value to select how many flanking
position should be included in the subset (default: <code>flanking = 0L</code>).</p></li>
<li><p><code>merge</code>: <code>TRUE</code> or <code>FALSE</code>: Should the 
overlapping selections be merged? This is particular important, if flanking
value <code>!= 0L</code> are set. (default: <code>merge = TRUE</code>).</p></li>
<li><p><code>perTranscript</code>: <code>TRUE</code> or <code>FALSE</code>: Should the 
positions labeled per transcript and not per chromosome?
(default: <code>perTranscript = FALSE</code>).</p></li>
</ul></td>
    </tr>
    <tr>
      <th>name</th>
      <td><p>Optional: Limit results to one specific transcript.</p></td>
    </tr>
    <tr>
      <th>pos</th>
      <td><p>Optional: Limit results to a specific position.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>If 'x' is a</p><ul>
<li><p><code><a href='SequenceData-class.html'>SequenceData</a></code> or 
<code><a href='Modifier-class.html'>Modifier</a></code>: a <code>SplitDataFrameList</code>
with elments per transcript.</p></li>
<li><p><code><a href='SequenceDataSet-class.html'>SequenceDataSet</a></code>,
<code><a href='SequenceDataList-class.html'>SequenceDataList</a></code> or
<code><a href='ModifierSet-class.html'>ModifierSet</a></code>: a <code>SimpleList</code> of
<code>SplitDataFrameList</code> with elments per transcript.</p></li>
</ul>


    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>msi</span>,<span class='kw'>package</span><span class='kw'>=</span><span class='st'>"RNAmodR"</span>)
<span class='no'>mod</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='modify.html'>modifications</a></span>(<span class='no'>msi</span>)
<span class='no'>coord</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/unlist.html'>unlist</a></span>(<span class='no'>mod</span>))
<span class='no'>coord</span>$<span class='no'>score</span> <span class='kw'>&lt;-</span> <span class='kw'>NULL</span>
<span class='no'>coord</span>$<span class='no'>sd</span> <span class='kw'>&lt;-</span> <span class='kw'>NULL</span>
<span class='fu'>subsetByCoord</span>(<span class='no'>msi</span>,<span class='no'>coord</span>)</div><div class='output co'>#&gt; $SampleSet1
#&gt; SplitDataFrameList of length 6
#&gt; $`2`
#&gt; DataFrame with 1 row and 1 column
#&gt;        score
#&gt;    &lt;numeric&gt;
#&gt; 34  0.900932
#&gt; 
#&gt; $`4`
#&gt; DataFrame with 1 row and 1 column
#&gt;        score
#&gt;    &lt;numeric&gt;
#&gt; 35  0.899622
#&gt; 
#&gt; $`6`
#&gt; DataFrame with 1 row and 1 column
#&gt;        score
#&gt;    &lt;numeric&gt;
#&gt; 34  0.984035
#&gt; 
#&gt; $`7`
#&gt; DataFrame with 1 row and 1 column
#&gt;        score
#&gt;    &lt;numeric&gt;
#&gt; 67  0.934553
#&gt; 
#&gt; $`9`
#&gt; DataFrame with 1 row and 1 column
#&gt;       score
#&gt;   &lt;numeric&gt;
#&gt; 7  0.709758
#&gt; 
#&gt; $`11`
#&gt; DataFrame with 1 row and 1 column
#&gt;        score
#&gt;    &lt;numeric&gt;
#&gt; 35  0.874027
#&gt; 
#&gt; 
#&gt; $SampleSet2
#&gt; SplitDataFrameList of length 6
#&gt; $`2`
#&gt; DataFrame with 1 row and 1 column
#&gt;        score
#&gt;    &lt;numeric&gt;
#&gt; 34  0.998134
#&gt; 
#&gt; $`4`
#&gt; DataFrame with 1 row and 1 column
#&gt;        score
#&gt;    &lt;numeric&gt;
#&gt; 35  0.856241
#&gt; 
#&gt; $`6`
#&gt; DataFrame with 1 row and 1 column
#&gt;        score
#&gt;    &lt;numeric&gt;
#&gt; 34  0.992012
#&gt; 
#&gt; $`7`
#&gt; DataFrame with 1 row and 1 column
#&gt;        score
#&gt;    &lt;numeric&gt;
#&gt; 67  0.942905
#&gt; 
#&gt; $`9`
#&gt; DataFrame with 1 row and 1 column
#&gt;       score
#&gt;   &lt;numeric&gt;
#&gt; 7  0.766484
#&gt; 
#&gt; $`11`
#&gt; DataFrame with 1 row and 1 column
#&gt;        score
#&gt;    &lt;numeric&gt;
#&gt; 35  0.971474
#&gt; 
#&gt; 
#&gt; $SampleSet3
#&gt; SplitDataFrameList of length 6
#&gt; $`2`
#&gt; DataFrame with 1 row and 1 column
#&gt;        score
#&gt;    &lt;numeric&gt;
#&gt; 34  0.953651
#&gt; 
#&gt; $`4`
#&gt; DataFrame with 1 row and 1 column
#&gt;        score
#&gt;    &lt;numeric&gt;
#&gt; 35  0.976928
#&gt; 
#&gt; $`6`
#&gt; DataFrame with 1 row and 1 column
#&gt;        score
#&gt;    &lt;numeric&gt;
#&gt; 34  0.993128
#&gt; 
#&gt; $`7`
#&gt; DataFrame with 1 row and 1 column
#&gt;        score
#&gt;    &lt;numeric&gt;
#&gt; 67  0.943773
#&gt; 
#&gt; $`9`
#&gt; DataFrame with 1 row and 1 column
#&gt;       score
#&gt;   &lt;numeric&gt;
#&gt; 7  0.681451
#&gt; 
#&gt; $`11`
#&gt; DataFrame with 1 row and 1 column
#&gt;        score
#&gt;    &lt;numeric&gt;
#&gt; 35  0.954782
#&gt; 
#&gt; </div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


