<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RNAmodR: analyzing high throughput sequencing data for post-transcriptional RNA modification footprints • RNAmodR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="RNAmodR: analyzing high throughput sequencing data for post-transcriptional RNA modification footprints">
<meta property="og:description" content="RNAmodR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.17.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/FelixErnst/RNAmodR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>RNAmodR: analyzing high throughput sequencing
data for post-transcriptional RNA modification footprints</h1>
                        <h4 data-toc-skip class="author">Felix G.M.
Ernst and Denis L.J. Lafontaine</h4>
            
            <h4 data-toc-skip class="date">2024-03-26</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/FelixErnst/RNAmodR/blob/HEAD/vignettes/RNAmodR.Rmd" class="external-link"><code>vignettes/RNAmodR.Rmd</code></a></small>
      <div class="hidden name"><code>RNAmodR.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Post-transcriptional modifications can be found abundantly in rRNA
and tRNA and can be detected classically via several strategies.
However, difficulties arise if the identity and the position of the
modified nucleotides is to be determined at the same time. Classically,
a primer extension, a form of reverse transcription (RT), would allow
certain modifications to be accessed by blocks during the RT or changes
in the cDNA sequences. Other modification would need to be selectively
treated by chemical reactions to influence the outcome of the reverse
transcription.</p>
<p>With the increased availability of high throughput sequencing, these
classical methods were adapted to high throughput methods allowing more
RNA molecules to be accessed at the same time. With these advances
post-transcriptional modifications were also detected on mRNA. Among
these high throughput techniques are for example Pseudo-Seq <a href="#References"><span class="citation">(Carlile et al.
2014)</span></a>, RiboMethSeq <a href="#References"><span class="citation">(Birkedal et al. 2015)</span></a> and AlkAnilineSeq <a href="#References"><span class="citation">(Marchand et al.
2018)</span></a> each able to detect a specific type of modification
from footprints in RNA-Seq data prepared with the selected methods.</p>
<p>Since similar pattern can be observed from some of these techniques,
overlaps of the bioinformatical pipeline already are and will become
more frequent with new emerging sequencing techniques.</p>
<p><code>RNAmodR</code> implements classes and a workflow to detect
post-transcriptional RNA modifications in high throughput sequencing
data. It is easily adaptable to new methods and can help during the
phase of initial method development as well as more complex
screenings.</p>
<p>Briefly, from the <code>SequenceData</code>, specific subclasses are
derived for accessing specific aspects of aligned reads, e.g. 5’-end
positions or pileup data. With this a <code>Modifier</code> class can be
used to detect specific patterns for individual types of modifications.
The <code>SequenceData</code> classes can be shared by different
<code>Modifier</code> classes allowing easy adaptation to new
methods.</p>
<pre><code><span><span class="co">## Warning: replacing previous import 'utils::findMatches' by</span></span>
<span><span class="co">## 'S4Vectors::findMatches' when loading 'ExperimentHubData'</span></span></code></pre>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">Rsamtools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicFeatures" class="external-link">GenomicFeatures</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/txdbmaker" class="external-link">txdbmaker</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/FelixErnst/RNAmodR.Data" class="external-link">RNAmodR.Data</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/FelixErnst/RNAmodR" class="external-link">RNAmodR</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="sequencedata">SequenceData<a class="anchor" aria-label="anchor" href="#sequencedata"></a>
</h3>
<p>Each <code>SequenceData</code> object is created with a named
character vector, which can be coerced to a <code>BamFileList</code>, or
named <code>BamFileList</code>. The names must be either “treated” or
“control” describing the condition the data file belongs to. Multiple
files can be given per condition and are used as replicates.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html" class="external-link">GFF3File</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.example.html" class="external-link">RNAmodR.Data.example.gff3</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.example.html" class="external-link">RNAmodR.Data.example.fasta</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Treated <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.example.html" class="external-link">RNAmodR.Data.example.bam.1</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>           Treated <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.example.html" class="external-link">RNAmodR.Data.example.bam.2</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>           Treated <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.example.html" class="external-link">RNAmodR.Data.example.bam.3</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>For <code>annotation</code> and <code>sequences</code> several input
are accepted. <code>annotation</code> can be a <code>GRangesList</code>,
a <code>GFF3File</code> or a <code>TxDb</code> object. Internally, a
<code>GFF3File</code> is converted to a <code>TxDb</code> object and a
<code>GRangesList</code> is retrieved using the <code>exonsBy</code>
function.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EndSequenceData-class.html">End5SequenceData</a></span><span class="op">(</span><span class="va">files</span>, annotation <span class="op">=</span> <span class="va">annotation</span>, </span>
<span>                            sequences <span class="op">=</span> <span class="va">sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Import genomic features from the file as a GRanges object ... OK</span></span>
<span><span class="co">## Prepare the 'metadata' data frame ... OK</span></span>
<span><span class="co">## Make the TxDb object ... OK</span></span>
<span><span class="co">## Loading 5'-end position data from BAM files ... OK</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqdata</span></span></code></pre></div>
<pre><code><span><span class="co">## End5SequenceData with 60 elements containing 3 data columns and 3 metadata columns</span></span>
<span><span class="co">## - Data columns:</span></span>
<span><span class="co">##  end5.treated.1 end5.treated.2 end5.treated.3</span></span>
<span><span class="co">##       &lt;integer&gt;      &lt;integer&gt;      &lt;integer&gt;</span></span>
<span><span class="co">## -  Seqinfo object with 84 sequences from an unspecified genome; no seqlengths:</span></span></code></pre>
<p><code>SequenceData</code> extends from a
<code>CompressedSplitDataFrameList</code> and contains the data per
transcript alongside the annotation information and the sequence. The
additional data stored within the <code>SequenceData</code> can be
accessed by several functions.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">seqdata</span><span class="op">)</span> <span class="co"># matches the transcript names as returned by a TxDb object</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">seqdata</span><span class="op">)</span> <span class="co"># returns a CharacterList of all column names</span></span>
<span><span class="fu"><a href="../reference/Modifier-functions.html">bamfiles</a></span><span class="op">(</span><span class="va">seqdata</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/range-squeezers.html" class="external-link">ranges</a></span><span class="op">(</span><span class="va">seqdata</span><span class="op">)</span> <span class="co"># generate from a TxDb object</span></span>
<span><span class="fu"><a href="../reference/Modifier-functions.html">sequences</a></span><span class="op">(</span><span class="va">seqdata</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqinfo</a></span><span class="op">(</span><span class="va">seqdata</span><span class="op">)</span></span></code></pre></div>
<p>Currently the following <code>SequenceData</code> classes are
implemented:</p>
<ul>
<li><code>End5SequenceData</code></li>
<li><code>End3SequenceData</code></li>
<li><code>EndSequenceData</code></li>
<li><code>ProtectedEndSequenceData</code></li>
<li><code>CoverageSequenceData</code></li>
<li><code>PileupSequenceData</code></li>
<li><code>NormEnd5SequenceData</code></li>
<li><code>NormEnd3SequenceData</code></li>
</ul>
<p>The data types and names of the columns are different for most of the
<code>SequenceData</code> classes. As a naming convenction a descriptor
is combined with the condition as defined in the files input and the
replicate number. For more details please have a look at the man pages,
e.g. <code><a href="../reference/EndSequenceData-class.html">?End5SequenceData</a></code>.</p>
<p><code>SequenceData</code> objects can be subset like a
<code>CompressedSplitDataFrameList</code>. Elements are returned as a
<code>SequenceDataFrame</code> dependent of the type of
<code>SequenceData</code> used. For each <code>SequenceData</code> class
a matching <code>SequenceDataFrame</code> is implemented.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqdata</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## End5SequenceData with 1 elements containing 3 data columns and 3 metadata columns</span></span>
<span><span class="co">## - Data columns:</span></span>
<span><span class="co">##  end5.treated.1 end5.treated.2 end5.treated.3</span></span>
<span><span class="co">##       &lt;integer&gt;      &lt;integer&gt;      &lt;integer&gt;</span></span>
<span><span class="co">## -  Seqinfo object with 84 sequences from an unspecified genome; no seqlengths:</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sdf</span> <span class="op">&lt;-</span> <span class="va">seqdata</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">sdf</span></span></code></pre></div>
<pre><code><span><span class="co">## End5SequenceDataFrame with 1649 rows and 3 columns</span></span>
<span><span class="co">##      end5.treated.1 end5.treated.2 end5.treated.3</span></span>
<span><span class="co">##           &lt;integer&gt;      &lt;integer&gt;      &lt;integer&gt;</span></span>
<span><span class="co">## 1                 1              4              0</span></span>
<span><span class="co">## 2                 0              2              0</span></span>
<span><span class="co">## 3                 0              0              0</span></span>
<span><span class="co">## 4                 0              0              0</span></span>
<span><span class="co">## 5                 0              0              0</span></span>
<span><span class="co">## ...             ...            ...            ...</span></span>
<span><span class="co">## 1645              0              0              0</span></span>
<span><span class="co">## 1646              0              0              0</span></span>
<span><span class="co">## 1647              0              0              0</span></span>
<span><span class="co">## 1648              0              0              0</span></span>
<span><span class="co">## 1649              0              0              0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## containing a GRanges object with 1 range and 3 metadata columns:</span></span>
<span><span class="co">##             seqnames    ranges strand |   exon_id   exon_name exon_rank</span></span>
<span><span class="co">##                &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   [1] Q0020_15S_RRNA    1-1649      + |         1       Q0020         1</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 60 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## and a 1649-letter RNAString object</span></span>
<span><span class="co">## seq: <span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #FFAFAF;">AAAAAA</span><span style="color: #000000; background-color: #FFFFAF;">UUU</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFFFAF;">UU</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFFFAF;">UU</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFFFAF;">U</span>...<span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #AFFFFF;">GGG</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFFFAF;">UU</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFFFAF;">UU</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFFFAF;">UU</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFFFAF;">UU</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span></span></span></code></pre>
<p>The <code>SequenceDataFrame</code> objects retains some accessor
functions from the <code>SequenceData</code> class.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sdf</span><span class="op">)</span> <span class="co"># this returns the columns names of the data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/range-squeezers.html" class="external-link">ranges</a></span><span class="op">(</span><span class="va">sdf</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Modifier-functions.html">sequences</a></span><span class="op">(</span><span class="va">sdf</span><span class="op">)</span></span></code></pre></div>
<p>Subsetting of a <code>SequenceDataFrame</code> returns a
<code>SequenceDataFrame</code> or <code>DataFrame</code>, depending on
whether it is subset by a column or row, respectively. The
<code>drop</code> argument is ignored for column subsetting.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sdf</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## End5SequenceDataFrame with 1649 rows and 2 columns</span></span>
<span><span class="co">##      end5.treated.1 end5.treated.2</span></span>
<span><span class="co">##           &lt;integer&gt;      &lt;integer&gt;</span></span>
<span><span class="co">## 1                 1              4</span></span>
<span><span class="co">## 2                 0              2</span></span>
<span><span class="co">## 3                 0              0</span></span>
<span><span class="co">## 4                 0              0</span></span>
<span><span class="co">## 5                 0              0</span></span>
<span><span class="co">## ...             ...            ...</span></span>
<span><span class="co">## 1645              0              0</span></span>
<span><span class="co">## 1646              0              0</span></span>
<span><span class="co">## 1647              0              0</span></span>
<span><span class="co">## 1648              0              0</span></span>
<span><span class="co">## 1649              0              0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## containing a GRanges object with 1 range and 3 metadata columns:</span></span>
<span><span class="co">##             seqnames    ranges strand |   exon_id   exon_name exon_rank</span></span>
<span><span class="co">##                &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   [1] Q0020_15S_RRNA    1-1649      + |         1       Q0020         1</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 60 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## and a 1649-letter RNAString object</span></span>
<span><span class="co">## seq: <span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #FFAFAF;">AAAAAA</span><span style="color: #000000; background-color: #FFFFAF;">UUU</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFFFAF;">UU</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFFFAF;">UU</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFFFAF;">U</span>...<span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #AFFFFF;">GGG</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFFFAF;">UU</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFFFAF;">UU</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #FFFFAF;">U</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFFFAF;">UU</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFFFAF;">UU</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span></span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sdf</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 3 rows and 3 columns</span></span>
<span><span class="co">##   end5.treated.1 end5.treated.2 end5.treated.3</span></span>
<span><span class="co">##        &lt;integer&gt;      &lt;integer&gt;      &lt;integer&gt;</span></span>
<span><span class="co">## 1              1              4              0</span></span>
<span><span class="co">## 2              0              2              0</span></span>
<span><span class="co">## 3              0              0              0</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="modifier">Modifier<a class="anchor" aria-label="anchor" href="#modifier"></a>
</h3>
<p>Whereas, the <code>SequenceData</code> classes are used to hold the
data, <code>Modifier</code> classes are used to detect certain features
within high throughput sequencing data to assign the presence of
specific modifications for an established pattern. The
<code>Modifier</code> class (and its nucleotide specific subclasses
<code>RNAModifier</code> and <code>DNAModifier</code>) is virtual and
can be addapted to individual methods. For example mapped reads can be
analyzed using the <code>ModInosine</code> class to reveal the presence
of I by detecting a A to G conversion in normal RNA-Seq data. Therefore,
<code>ModInosine</code> inherits from <code>RNAModifier</code>.</p>
<p>To fix the data processing and detection strategy, for each type of
sequencing method a <code>Modifier</code> class can be developed
alongside to detect modifications. For more information on how to
develop such a class and potentially a new corresponding
<code>SequenceData</code> class, please have a look at the <a href="RNAmodR.creation.html">vignette for creating a new
analysis</a>.</p>
<p>For now three <code>Modifier</code> classes are available:</p>
<ul>
<li><code>ModInosine</code></li>
<li>
<code>ModRiboMethSeq</code> from the
<code>RNAmodR.RiboMethSeq</code> package</li>
<li>
<code>ModAlkAnilineSeq</code> from the
<code>RNAmodR.AlkAnilineSeq</code> package</li>
</ul>
<p><code>Modifier</code> objects can use and wrap multiple
<code>SequenceData</code> objects as elements of a
<code>SequenceDataSet</code> class. The elements of this class are
different types of <code>SequenceData</code>, which are required by the
specific <code>Modifier</code> class. However, they are required to
contain data for the same annotation and sequence data.</p>
<p><code>Modifier</code> objects are created with the same arguments as
<code>SequenceData</code> objects and will start loading the necessary
<code>SequenceData</code> objects from these. In addition they will
automatically start to calculate any additional scores (aggregation) and
then start to search for modifications, if the optional argument
<code>find.mod</code> is not set to <code>FALSE</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModInosine.html">ModInosine</a></span><span class="op">(</span><span class="va">files</span>, annotation <span class="op">=</span> <span class="va">annotation</span>, sequences <span class="op">=</span> <span class="va">sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Import genomic features from the file as a GRanges object ... OK</span></span>
<span><span class="co">## Prepare the 'metadata' data frame ... OK</span></span>
<span><span class="co">## Make the TxDb object ... OK</span></span>
<span><span class="co">## Loading Pileup data from BAM files ... OK</span></span>
<span><span class="co">## Aggregating data and calculating scores ... Starting to search for 'Inosine' ... done.</span></span></code></pre>
<p>(Hint: If you use an artificial genome, name the chromosomes
chr1-chrN. It will make some things easier for subsequent visualization,
which relies on the <code>Gviz</code> package)</p>
<p>Since the <code>Modifier</code> class wraps a
<code>SequenceData</code> object the accessors to data contained within
work similarly to the <code>SequenceData</code> accessors described
above. What type of conditions the <code>Modifier</code> class
expects/supports is usually described in the man pages of the Modifier
class.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mi</span><span class="op">)</span> <span class="co"># matches the transcript names as returned by a TxDb object</span></span>
<span><span class="fu"><a href="../reference/Modifier-functions.html">bamfiles</a></span><span class="op">(</span><span class="va">mi</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/range-squeezers.html" class="external-link">ranges</a></span><span class="op">(</span><span class="va">mi</span><span class="op">)</span> <span class="co"># generated from a TxDb object</span></span>
<span><span class="fu"><a href="../reference/Modifier-functions.html">sequences</a></span><span class="op">(</span><span class="va">mi</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqinfo</a></span><span class="op">(</span><span class="va">mi</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Modifier-functions.html">sequenceData</a></span><span class="op">(</span><span class="va">mi</span><span class="op">)</span> <span class="co"># returns the SequenceData </span></span></code></pre></div>
<div class="section level4">
<h4 id="settings">Settings<a class="anchor" aria-label="anchor" href="#settings"></a>
</h4>
<p>The behavior of a <code>Modifier</code> class can be fine tuned using
settings. The <code><a href="../reference/settings.html">settings()</a></code> function is a getter/setter for
arguments used in the analysis and my differ between different
<code>Modifier</code> classes depending on the particular strategy and
whether they are implemented as flexible settings.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/settings.html">settings</a></span><span class="op">(</span><span class="va">mi</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $minCoverage</span></span>
<span><span class="co">## [1] 10</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $minReplicate</span></span>
<span><span class="co">## [1] 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $find.mod</span></span>
<span><span class="co">## [1] TRUE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $minScore</span></span>
<span><span class="co">## [1] 0.4</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/settings.html">settings</a></span><span class="op">(</span><span class="va">mi</span>,<span class="st">"minScore"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.4</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/settings.html">settings</a></span><span class="op">(</span><span class="va">mi</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>minScore <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/settings.html">settings</a></span><span class="op">(</span><span class="va">mi</span>,<span class="st">"minScore"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.5</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="modifierset">ModifierSet<a class="anchor" aria-label="anchor" href="#modifierset"></a>
</h3>
<p>Each <code>Modifier</code> object is able to represent one sample set
with multiple replicates of data. To easily compare multiple sample sets
the <code>ModifierSet</code> class is implemented.</p>
<p>The <code>ModifierSet</code> object is created from a named list of
named character vectors or <code>BamFileList</code> objects. Each
element in the list is a sample type with a corresponding name. Each
entry in the character vector/<code>BamFileList</code> is a replicate
(Alternatively a <code>ModifierSet</code> can also be created from a
<code>list</code> of <code>Modifier</code> objects, if they are of the
same type).</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html" class="external-link">RNAmodR.Data.example.AAS.fasta</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html" class="external-link">GFF3File</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html" class="external-link">RNAmodR.Data.example.AAS.gff3</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"SampleSet1"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>treated <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html" class="external-link">RNAmodR.Data.example.wt.1</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                               treated <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html" class="external-link">RNAmodR.Data.example.wt.2</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                               treated <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html" class="external-link">RNAmodR.Data.example.wt.3</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>              <span class="st">"SampleSet2"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>treated <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html" class="external-link">RNAmodR.Data.example.bud23.1</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                               treated <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html" class="external-link">RNAmodR.Data.example.bud23.2</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>              <span class="st">"SampleSet3"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>treated <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html" class="external-link">RNAmodR.Data.example.trm8.1</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                               treated <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/RNAmodR.Data/man/RNAmodR.Data.AAS.html" class="external-link">RNAmodR.Data.example.trm8.2</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">msi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModInosine.html">ModSetInosine</a></span><span class="op">(</span><span class="va">files</span>, annotation <span class="op">=</span> <span class="va">annotation</span>, sequences <span class="op">=</span> <span class="va">sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Import genomic features from the file as a GRanges object ... OK</span></span>
<span><span class="co">## Prepare the 'metadata' data frame ... OK</span></span>
<span><span class="co">## Make the TxDb object ... OK</span></span></code></pre>
<p>The creation of the <code>ModifierSet</code> will itself trigger the
creation of a <code>Modifier</code> object each containing data from one
sample set. This step is parallelized using the
<code>BiocParallel</code> package. If a <code>Modifier</code> class
itself uses parallel computing for its analysis, it is switched off
unless <code>internalBP = TRUE</code> is set. In this case each
<code>Modifier</code> object is created in sequence, allowing parallel
computing during the creation of each object.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">msi</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "SampleSet1" "SampleSet2" "SampleSet3"</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">msi</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## A ModInosine object containing PileupSequenceData with 11 elements.</span></span>
<span><span class="co">## | Input files:</span></span>
<span><span class="co">##    - treated: /github/home/.cache/R/ExperimentHub/232d576c0eff_2544</span></span>
<span><span class="co">##    - treated: /github/home/.cache/R/ExperimentHub/232d70523712_2546</span></span>
<span><span class="co">##    - treated: /github/home/.cache/R/ExperimentHub/232d3bd7c2ca_2548</span></span>
<span><span class="co">## | Nucleotide - Modification type(s):  RNA  -  I </span></span>
<span><span class="co">## | Modifications found: yes (6) </span></span>
<span><span class="co">## | Settings:</span></span>
<span><span class="co">##   minCoverage minReplicate  find.mod  minScore</span></span>
<span><span class="co">##     &lt;integer&gt;    &lt;integer&gt; &lt;logical&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##            10            1      TRUE       0.4</span></span></code></pre>
<p>Again accessors remain mostly the same as described above for the
<code>Modifier</code> class returning a list of results, one element for
each <code>Modifier</code> object.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Modifier-functions.html">bamfiles</a></span><span class="op">(</span><span class="va">msi</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/range-squeezers.html" class="external-link">ranges</a></span><span class="op">(</span><span class="va">msi</span><span class="op">)</span> <span class="co"># generate from a TxDb object</span></span>
<span><span class="fu"><a href="../reference/Modifier-functions.html">sequences</a></span><span class="op">(</span><span class="va">msi</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqinfo</a></span><span class="op">(</span><span class="va">msi</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="analysis-of-detected-modifications">Analysis of detected modifications<a class="anchor" aria-label="anchor" href="#analysis-of-detected-modifications"></a>
</h2>
<p>Found modifications can be retrieved from a <code>Modifier</code> or
<code>ModifierSet</code> object via the <code><a href="../reference/modify.html">modifications()</a></code>
function. The function returns a <code>GRanges</code> or
<code>GRangesList</code> object, respectively, which contains the
coordinates of the modifications with respect to the genome used. For
example if a transcript starts at position 100 and contains a modified
nucleotide at position 50 of the transcript, the returned coordinate
will 150.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/modify.html">modifications</a></span><span class="op">(</span><span class="va">msi</span><span class="op">)</span></span>
<span><span class="va">mod</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 5 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |         mod      source        type     score</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr2        34      + |           I     RNAmodR      RNAMOD  0.900932</span></span>
<span><span class="co">##   [2]     chr4        35      + |           I     RNAmodR      RNAMOD  0.899622</span></span>
<span><span class="co">##   [3]     chr6        34      + |           I     RNAmodR      RNAMOD  0.984035</span></span>
<span><span class="co">##   [4]     chr7        67      + |           I     RNAmodR      RNAMOD  0.934553</span></span>
<span><span class="co">##   [5]     chr9         7      + |           I     RNAmodR      RNAMOD  0.709758</span></span>
<span><span class="co">##   [6]    chr11        35      + |           I     RNAmodR      RNAMOD  0.874027</span></span>
<span><span class="co">##            Parent</span></span>
<span><span class="co">##       &lt;character&gt;</span></span>
<span><span class="co">##   [1]           2</span></span>
<span><span class="co">##   [2]           4</span></span>
<span><span class="co">##   [3]           6</span></span>
<span><span class="co">##   [4]           7</span></span>
<span><span class="co">##   [5]           9</span></span>
<span><span class="co">##   [6]          11</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 11 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>To retrieve the coordinates with respect to the transcript
boundaries, use the optional argument <code>perTranscript = TRUE</code>.
In the example provided here, this will yield the same coordinates,
since a custom genome was used for mapping of the example, which does
not contain transcripts on the negative strand and per transcript
chromosomes.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/modify.html">modifications</a></span><span class="op">(</span><span class="va">msi</span>, perTranscript <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">mod</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 5 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |         mod      source        type     score</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr2        34      * |           I     RNAmodR      RNAMOD  0.900932</span></span>
<span><span class="co">##   [2]     chr4        35      * |           I     RNAmodR      RNAMOD  0.899622</span></span>
<span><span class="co">##   [3]     chr6        34      * |           I     RNAmodR      RNAMOD  0.984035</span></span>
<span><span class="co">##   [4]     chr7        67      * |           I     RNAmodR      RNAMOD  0.934553</span></span>
<span><span class="co">##   [5]     chr9         7      * |           I     RNAmodR      RNAMOD  0.709758</span></span>
<span><span class="co">##   [6]    chr11        35      * |           I     RNAmodR      RNAMOD  0.874027</span></span>
<span><span class="co">##            Parent</span></span>
<span><span class="co">##       &lt;character&gt;</span></span>
<span><span class="co">##   [1]           2</span></span>
<span><span class="co">##   [2]           4</span></span>
<span><span class="co">##   [3]           6</span></span>
<span><span class="co">##   [4]           7</span></span>
<span><span class="co">##   [5]           9</span></span>
<span><span class="co">##   [6]          11</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 11 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="section level3">
<h3 id="compairing-results">Compairing results<a class="anchor" aria-label="anchor" href="#compairing-results"></a>
</h3>
<p>To compare results between samples, a <code>ModifierSet</code> as
well as a definition of positions to compare are required. To construct
a set of positions, we will use the intersection of all modifications
found as an example.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/modify.html">modifications</a></span><span class="op">(</span><span class="va">msi</span><span class="op">)</span></span>
<span><span class="va">coord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">mod</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">coord</span><span class="op">$</span><span class="va">score</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">coord</span><span class="op">$</span><span class="va">sd</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="fu"><a href="../reference/compareByCoord.html">compareByCoord</a></span><span class="op">(</span><span class="va">msi</span>,<span class="va">coord</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 6 columns</span></span>
<span><span class="co">##   SampleSet1 SampleSet2 SampleSet3    names positions         mod</span></span>
<span><span class="co">##    &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt; &lt;factor&gt;  &lt;factor&gt; &lt;character&gt;</span></span>
<span><span class="co">## 1   0.900932   0.998134   0.953651       2         34           I</span></span>
<span><span class="co">## 2   0.899622   0.856241   0.976928       4         35           I</span></span>
<span><span class="co">## 3   0.984035   0.992012   0.993128       6         34           I</span></span>
<span><span class="co">## 4   0.934553   0.942905   0.943773       7         67           I</span></span>
<span><span class="co">## 5   0.709758   0.766484   0.681451       9         7            I</span></span>
<span><span class="co">## 6   0.874027   0.971474   0.954782       11        35           I</span></span></code></pre>
<p>The result can also be plotted using <code>plotCompareByCoord</code>,
which accepts an optional argument <code>alias</code> to allow
transcript ids to be converted to other identifiers. For this step it is
probably helpful to construct a <code>TxDb</code> object right at the
beginning and use it for constructing the
<code>Modifier</code>/<code>ModifierSet</code> object as the
<code>annotation</code> argument.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html" class="external-link">makeTxDbFromGFF</a></span><span class="op">(</span><span class="va">annotation</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Import genomic features from the file as a GRanges object ... OK</span></span>
<span><span class="co">## Prepare the 'metadata' data frame ... OK</span></span>
<span><span class="co">## Make the TxDb object ... OK</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alias</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>tx_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/id2name.html" class="external-link">id2name</a></span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                    name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/id2name.html" class="external-link">id2name</a></span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compareByCoord.html">plotCompareByCoord</a></span><span class="op">(</span><span class="va">msi</span>, <span class="va">coord</span>, alias <span class="op">=</span> <span class="va">alias</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="RNAmodR_files/figure-html/plot1-1.png" alt="Heatmap for identified Inosine positions." width="700"><p class="caption">
Heatmap for identified Inosine positions.
</p>
</div>
<p>Additionally, the order of sample sets can be adjusted, normalized to
any of the sample sets and the numbering of positions shown per
transcript.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compareByCoord.html">plotCompareByCoord</a></span><span class="op">(</span><span class="va">msi</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span>, <span class="va">coord</span>, alias <span class="op">=</span> <span class="va">alias</span>, normalize <span class="op">=</span> <span class="st">"SampleSet3"</span>,</span>
<span>                   perTranscript <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="RNAmodR_files/figure-html/plot2-1.png" alt="Heatmap for identified Inosine positions with normalized scores." width="700"><p class="caption">
Heatmap for identified Inosine positions with normalized scores.
</p>
</div>
<p>The calculated scores and data can be visualized along the
transcripts or chunks of the transcript. With the optional argument
<code>showSequenceData</code> the plotting of the sequence data in
addition to the score data can be triggered by setting it to
<code>TRUE</code>.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotData.html">plotData</a></span><span class="op">(</span><span class="va">msi</span>, <span class="st">"2"</span>, from <span class="op">=</span> <span class="fl">10L</span>, to <span class="op">=</span> <span class="fl">45L</span>, alias <span class="op">=</span> <span class="va">alias</span><span class="op">)</span> <span class="co"># showSequenceData = FALSE</span></span></code></pre></div>
<div class="figure">
<img src="RNAmodR_files/figure-html/plot3-1.png" alt="Scores along a transcript containing a A to G conversion indicating the presence of Inosine." width="700"><p class="caption">
Scores along a transcript containing a A to G conversion indicating the
presence of Inosine.
</p>
</div>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotData.html">plotData</a></span><span class="op">(</span><span class="va">msi</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, <span class="st">"2"</span>, from <span class="op">=</span> <span class="fl">10L</span>, to <span class="op">=</span> <span class="fl">45L</span>, showSequenceData <span class="op">=</span> <span class="cn">TRUE</span>, alias <span class="op">=</span> <span class="va">alias</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="RNAmodR_files/figure-html/plot4-1.png" alt="Scores along a transcript containing a A to G conversion indicating the presence of Inosine. This figure includes the detailed pileup sequence data." width="700"><p class="caption">
Scores along a transcript containing a A to G conversion indicating the
presence of Inosine. This figure includes the detailed pileup sequence
data.
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="performance-measurements">Performance measurements<a class="anchor" aria-label="anchor" href="#performance-measurements"></a>
</h2>
<p>Since the detection of modifications from high throughput sequencing
data relies usually on thresholds for calling modifications, there is
considerable interest in analyzing the performance of the method based
on scores chosen and available samples. To analyse the performance, the
function <code><a href="../reference/plotROC.html">plotROC()</a></code> is implemented, which is a wrapper
around the functionality of the <code>ROCR</code> package <a href="#References"><span class="citation">Sing et al.
(2005)</span></a>.</p>
<p>For the example data used in this vignette, the information gained is
rather limited and the following figure should be regarded just as a
proof of concept. In addition, the use of found modifications sites as
an input for <code>plotROC</code> is strongly discouraged, since defeats
the purpose of the test. Therefore, please regard this aspect of the
next chunk as proof of concept as well.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotROC.html">plotROC</a></span><span class="op">(</span><span class="va">msi</span>, <span class="va">coord</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="RNAmodR_files/figure-html/plot5-1.png" alt="TPR vs. FPR plot." width="700"><p class="caption">
TPR vs. FPR plot.
</p>
</div>
<p>Please have a look at <code><a href="../reference/plotROC.html">?plotROC</a></code> for additional details.
Most of the functionality from the <code>ROCR</code> package is
available via additional arguments, thus the output of
<code>plotROC</code> can be heavily customized.</p>
</div>
<div class="section level2">
<h2 id="additional-informations">Additional informations<a class="anchor" aria-label="anchor" href="#additional-informations"></a>
</h2>
<p>To have a look at metadata of reads for an analysis with
<code>RNAmodR</code> the function <code><a href="../reference/stats.html">stats()</a></code> can be used. It
can be used with a bunch of object types: <code>SequenceData</code>,
<code>SequenceDataList</code>, <code>SequenceDataSet</code>,
<code>Modifier</code> or <code>ModifierSet</code>. For
<code>SequenceData*</code> objects, the <code>BamFile</code> to be
analyzed must be provided as well, which automatically done for
<code>Modifier*</code> objects. For more details have a look at
<code><a href="../reference/stats.html">?stats</a></code>.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/stats.html">stats</a></span><span class="op">(</span><span class="va">msi</span><span class="op">)</span></span>
<span><span class="va">stats</span></span></code></pre></div>
<pre><code><span><span class="co">## List of length 3</span></span>
<span><span class="co">## names(3): SampleSet1 SampleSet2 SampleSet3</span></span></code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stats</span><span class="op">[[</span><span class="st">"SampleSet1"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrameList of length 3</span></span>
<span><span class="co">## names(3): treated treated treated</span></span></code></pre>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stats</span><span class="op">[[</span><span class="st">"SampleSet1"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"treated"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 12 rows and 6 columns</span></span>
<span><span class="co">##     seqnames seqlength    mapped  unmapped          used     used_distro</span></span>
<span><span class="co">##     &lt;factor&gt; &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;IntegerList&gt;          &lt;List&gt;</span></span>
<span><span class="co">## 1       chr1      1800    197050         0        159782 83,1252,860,...</span></span>
<span><span class="co">## 2       chr2        85      5863         0          2459     2,16,16,...</span></span>
<span><span class="co">## 3       chr3        76     76905         0         63497 35,478,4106,...</span></span>
<span><span class="co">## 4       chr4        77      8299         0          6554     6,27,36,...</span></span>
<span><span class="co">## 5       chr5        74     11758         0          8818  520,105,93,...</span></span>
<span><span class="co">## ...      ...       ...       ...       ...           ...             ...</span></span>
<span><span class="co">## 8      chr8         75    144293         0        143068    14,44,48,...</span></span>
<span><span class="co">## 9      chr9         75     13790         0          9753     1,49,43,...</span></span>
<span><span class="co">## 10     chr10        85     19861         0         17729    35,21,10,...</span></span>
<span><span class="co">## 11     chr11        77     10532         0          9086   53,92,185,...</span></span>
<span><span class="co">## 12     *             0         0    961095            NA              NA</span></span></code></pre>
<p>The data returned by <code><a href="../reference/stats.html">stats()</a></code> is a
<code>DataFrame</code>, which can be wrapped as a
<code>DataFrameList</code> or a <code>SimpleList</code> depending on the
input type. Analysis of the data must be manually done and can be used
to produced output like the following plot for distribution of lengths
for reads analyzed.</p>
<div class="figure">
<img src="RNAmodR_files/figure-html/plot6-1.png" alt="Distribution of lengths for reads used in the analysis" width="700"><p class="caption">
Distribution of lengths for reads used in the analysis
</p>
</div>
</div>
<div class="section level2">
<h2 id="further-development">Further development<a class="anchor" aria-label="anchor" href="#further-development"></a>
</h2>
<p>The development of <code>RNAmodR</code> will continue. General
ascpects of the analysis workflow will be addressed in the
<code>RNAmodR</code> package, whereas additional classes for new
sequencing techniques targeted at detecting post-transcriptional will be
wrapped in individual packages. This will allow general improvements to
propagate upstream, but not hinder individual requirements of each
detection strategy.</p>
<p>For an example have a look at the <code>RNAmodR.RiboMethSeq</code>
and <code>RNAmodR.AlkAnilineSeq</code> packages.</p>
<p><img src="RNAmodR.png"></p>
<p>Features, which might be added in the future:</p>
<ul>
<li>interaction with our packages for data aggregation (for example meta
gene aggregation)</li>
<li>interaction with our packages for downstream analysis for
visualization</li>
</ul>
<p>We welcome contributions of any sort.</p>
</div>
<div class="section level2">
<h2 id="sessioninfo">Sessioninfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2024-03-24 r86185)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] RNAmodR_1.17.1           Modstrings_1.19.0        RNAmodR.Data_1.17.0     </span></span>
<span><span class="co">##  [4] ExperimentHubData_1.29.0 AnnotationHubData_1.33.1 futile.logger_1.4.3     </span></span>
<span><span class="co">##  [7] ExperimentHub_2.11.1     AnnotationHub_3.11.3     BiocFileCache_2.11.1    </span></span>
<span><span class="co">## [10] dbplyr_2.5.0             txdbmaker_0.99.7         GenomicFeatures_1.55.4  </span></span>
<span><span class="co">## [13] AnnotationDbi_1.65.2     Biobase_2.63.0           Rsamtools_2.19.4        </span></span>
<span><span class="co">## [16] Biostrings_2.71.5        XVector_0.43.1           rtracklayer_1.63.1      </span></span>
<span><span class="co">## [19] GenomicRanges_1.55.4     GenomeInfoDb_1.39.9      IRanges_2.37.1          </span></span>
<span><span class="co">## [22] S4Vectors_0.41.5         BiocGenerics_0.49.1      BiocStyle_2.31.0        </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] BiocIO_1.13.0               bitops_1.0-7               </span></span>
<span><span class="co">##   [3] filelock_1.0.3              tibble_3.2.1               </span></span>
<span><span class="co">##   [5] graph_1.81.0                XML_3.99-0.16.1            </span></span>
<span><span class="co">##   [7] rpart_4.1.23                lifecycle_1.0.4            </span></span>
<span><span class="co">##   [9] httr2_1.0.0                 lattice_0.22-6             </span></span>
<span><span class="co">##  [11] ensembldb_2.27.1            OrganismDbi_1.45.1         </span></span>
<span><span class="co">##  [13] backports_1.4.1             magrittr_2.0.3             </span></span>
<span><span class="co">##  [15] Hmisc_5.1-2                 sass_0.4.9                 </span></span>
<span><span class="co">##  [17] rmarkdown_2.26              jquerylib_0.1.4            </span></span>
<span><span class="co">##  [19] yaml_2.3.8                  RUnit_0.4.33               </span></span>
<span><span class="co">##  [21] Gviz_1.47.1                 DBI_1.2.2                  </span></span>
<span><span class="co">##  [23] RColorBrewer_1.1-3          abind_1.4-5                </span></span>
<span><span class="co">##  [25] zlibbioc_1.49.3             purrr_1.0.2                </span></span>
<span><span class="co">##  [27] AnnotationFilter_1.27.0     biovizBase_1.51.0          </span></span>
<span><span class="co">##  [29] RCurl_1.98-1.14             nnet_7.3-19                </span></span>
<span><span class="co">##  [31] VariantAnnotation_1.49.6    rappdirs_0.3.3             </span></span>
<span><span class="co">##  [33] GenomeInfoDbData_1.2.11     AnnotationForge_1.45.0     </span></span>
<span><span class="co">##  [35] pkgdown_2.0.7               codetools_0.2-19           </span></span>
<span><span class="co">##  [37] DelayedArray_0.29.9         xml2_1.3.6                 </span></span>
<span><span class="co">##  [39] tidyselect_1.2.1            farver_2.1.1               </span></span>
<span><span class="co">##  [41] matrixStats_1.2.0           base64enc_0.1-3            </span></span>
<span><span class="co">##  [43] GenomicAlignments_1.39.4    jsonlite_1.8.8             </span></span>
<span><span class="co">##  [45] Formula_1.2-5               systemfonts_1.0.6          </span></span>
<span><span class="co">##  [47] tools_4.4.0                 progress_1.2.3             </span></span>
<span><span class="co">##  [49] ragg_1.3.0                  stringdist_0.9.12          </span></span>
<span><span class="co">##  [51] Rcpp_1.0.12                 glue_1.7.0                 </span></span>
<span><span class="co">##  [53] gridExtra_2.3               SparseArray_1.3.4          </span></span>
<span><span class="co">##  [55] BiocBaseUtils_1.5.1         xfun_0.43                  </span></span>
<span><span class="co">##  [57] MatrixGenerics_1.15.0       dplyr_1.1.4                </span></span>
<span><span class="co">##  [59] withr_3.0.0                 formatR_1.14               </span></span>
<span><span class="co">##  [61] BiocManager_1.30.22         fastmap_1.1.1              </span></span>
<span><span class="co">##  [63] latticeExtra_0.6-30         fansi_1.0.6                </span></span>
<span><span class="co">##  [65] digest_0.6.35               R6_2.5.1                   </span></span>
<span><span class="co">##  [67] mime_0.12                   textshaping_0.3.7          </span></span>
<span><span class="co">##  [69] colorspace_2.1-0            jpeg_0.1-10                </span></span>
<span><span class="co">##  [71] dichromat_2.0-0.1           biomaRt_2.59.1             </span></span>
<span><span class="co">##  [73] RSQLite_2.3.5               utf8_1.2.4                 </span></span>
<span><span class="co">##  [75] generics_0.1.3              data.table_1.15.2          </span></span>
<span><span class="co">##  [77] prettyunits_1.2.0           httr_1.4.7                 </span></span>
<span><span class="co">##  [79] htmlwidgets_1.6.4           S4Arrays_1.3.6             </span></span>
<span><span class="co">##  [81] pkgconfig_2.0.3             gtable_0.3.4               </span></span>
<span><span class="co">##  [83] blob_1.2.4                  htmltools_0.5.8            </span></span>
<span><span class="co">##  [85] bookdown_0.38               RBGL_1.79.0                </span></span>
<span><span class="co">##  [87] ProtGenerics_1.35.4         scales_1.3.0               </span></span>
<span><span class="co">##  [89] png_0.1-8                   colorRamps_2.3.4           </span></span>
<span><span class="co">##  [91] knitr_1.45                  lambda.r_1.2.4             </span></span>
<span><span class="co">##  [93] rstudioapi_0.16.0           reshape2_1.4.4             </span></span>
<span><span class="co">##  [95] rjson_0.2.21                checkmate_2.3.1            </span></span>
<span><span class="co">##  [97] curl_5.2.1                  biocViews_1.71.1           </span></span>
<span><span class="co">##  [99] cachem_1.0.8                stringr_1.5.1              </span></span>
<span><span class="co">## [101] BiocVersion_3.19.1          parallel_4.4.0             </span></span>
<span><span class="co">## [103] foreign_0.8-86              restfulr_0.0.15            </span></span>
<span><span class="co">## [105] desc_1.4.3                  pillar_1.9.0               </span></span>
<span><span class="co">## [107] grid_4.4.0                  vctrs_0.6.5                </span></span>
<span><span class="co">## [109] cluster_2.1.6               htmlTable_2.4.2            </span></span>
<span><span class="co">## [111] evaluate_0.23               cli_3.6.2                  </span></span>
<span><span class="co">## [113] compiler_4.4.0              futile.options_1.0.1       </span></span>
<span><span class="co">## [115] rlang_1.1.3                 crayon_1.5.2               </span></span>
<span><span class="co">## [117] labeling_0.4.3              interp_1.1-6               </span></span>
<span><span class="co">## [119] plyr_1.8.9                  fs_1.6.3                   </span></span>
<span><span class="co">## [121] stringi_1.8.3               deldir_2.0-4               </span></span>
<span><span class="co">## [123] BiocParallel_1.37.1         BiocCheck_1.39.27          </span></span>
<span><span class="co">## [125] munsell_0.5.0               lazyeval_0.2.2             </span></span>
<span><span class="co">## [127] Matrix_1.7-0                BSgenome_1.71.2            </span></span>
<span><span class="co">## [129] hms_1.1.3                   bit64_4.0.5                </span></span>
<span><span class="co">## [131] ggplot2_3.5.0               KEGGREST_1.43.0            </span></span>
<span><span class="co">## [133] SummarizedExperiment_1.33.3 highr_0.10                 </span></span>
<span><span class="co">## [135] ROCR_1.0-11                 memoise_2.0.1              </span></span>
<span><span class="co">## [137] bslib_0.6.2                 bit_4.0.5</span></span></code></pre>
<p><a name="References"></a></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Birkedal.2015" class="csl-entry">
Birkedal, Ulf, Mikkel Christensen-Dalsgaard, Nicolai Krogh,
Radhakrishnan Sabarinathan, Jan Gorodkin, and Henrik Nielsen. 2015.
<span>“Profiling of Ribose Methylations in RNA by High-Throughput
Sequencing.”</span> <em>Angewandte Chemie (International Ed. In
English)</em> 54 (2): 451–55. <a href="https://doi.org/10.1002/anie.201408362" class="external-link">https://doi.org/10.1002/anie.201408362</a>.
</div>
<div id="ref-Carlile.2014" class="csl-entry">
Carlile, Thomas M., Maria F. Rojas-Duran, Boris Zinshteyn, Hakyung Shin,
Kristen M. Bartoli, and Wendy V. Gilbert. 2014. <span>“Pseudouridine
Profiling Reveals Regulated mRNA Pseudouridylation in Yeast and Human
Cells.”</span> <em>Nature</em> 515 (7525): 143–46.
</div>
<div id="ref-Marchand.2018" class="csl-entry">
Marchand, Virginie, Lilia Ayadi, Felix G. M. Ernst, Jasmin Hertler,
Valérie Bourguignon-Igel, Adeline Galvanin, Annika Kotter, Mark Helm,
Denis L. J. Lafontaine, and Yuri Motorin. 2018. <span>“AlkAniline-Seq:
Profiling of m7G and m3C RNA Modifications at Single Nucleotide
Resolution.”</span> <em>Angewandte Chemie International Edition</em> 57
(51): 16785–90. <a href="https://doi.org/10.1002/anie.201810946" class="external-link">https://doi.org/10.1002/anie.201810946</a>.
</div>
<div id="ref-Sing.2005" class="csl-entry">
Sing, Tobias, Oliver Sander, Niko Beerenwinkel, and Thomas Lengauer.
2005. <span>“ROCR: Visualizing Classifier Performance in r.”</span>
<em>Bioinformatics (Oxford, England)</em> 21 (20): 3940–41. <a href="https://doi.org/10.1093/bioinformatics/bti623" class="external-link">https://doi.org/10.1093/bioinformatics/bti623</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
