<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RNAmodR: creating classes for additional modification detection from high throughput sequencing. â€¢ RNAmodR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="RNAmodR: creating classes for additional modification detection from high throughput sequencing.">
<meta property="og:description" content="RNAmodR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/FelixErnst/RNAmodR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="RNAmodR.creation_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>RNAmodR: creating classes for additional modification detection from high throughput sequencing.</h1>
                        <h4 class="author">Felix G.M. Ernst</h4>
            
            <h4 class="date">2021-01-12</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/FelixErnst/RNAmodR/blob/master/vignettes/RNAmodR.creation.Rmd"><code>vignettes/RNAmodR.creation.Rmd</code></a></small>
      <div class="hidden name"><code>RNAmodR.creation.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>For users interested in the general aspect of any <code>RNAmodR</code> based package please have a look at the <a href="RNAmodR.html">main vignette</a> of the package.</p>
<p>This vignette is aimed at developers and researchers, who want to use the functionality of the <code>RNAmodR</code> package to develop a new modification strategy based on high throughput sequencing data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/FelixErnst/RNAmodR">RNAmodR</a></span><span class="op">)</span></code></pre></div>
<p>Two classes have to be considered to establish a new analysis pipeline using <code>RNAmodR</code>. These are the <code>SequenceData</code> and the <code>Modifier</code> class.</p>
</div>
<div id="a-new-sequencedata-class" class="section level1">
<h1 class="hasAnchor">
<a href="#a-new-sequencedata-class" class="anchor"></a>A new <code>SequenceData</code> class</h1>
<p>First, the <code>SequenceData</code> class has to be considered. Several classes are already implemented, which are:</p>
<ul>
<li><code>End5SequenceData</code></li>
<li><code>End3SequenceData</code></li>
<li><code>EndSequenceData</code></li>
<li><code>ProtectedEndSequenceData</code></li>
<li><code>CoverageSequenceData</code></li>
<li><code>PileupSequenceData</code></li>
<li><code>NormEnd5SequenceData</code></li>
<li><code>NormEnd3SequenceData</code></li>
</ul>
<p>If these cannot be reused, a new class can be implemented quite easily. First the DataFrame class, the Data class and a constructor has to defined. The only value, which has to be provided, is a default <code>minQuality</code> integer value and some basic information.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">setClass</span><span class="op">(</span>Class <span class="op">=</span> <span class="st">"ExampleSequenceDataFrame"</span>,
         contains <span class="op">=</span> <span class="st">"SequenceDFrame"</span><span class="op">)</span>
<span class="va">ExampleSequenceDataFrame</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">df</span>, <span class="va">ranges</span>, <span class="va">sequence</span>, <span class="va">replicate</span>,
                                      <span class="va">condition</span>, <span class="va">bamfiles</span>, <span class="va">seqinfo</span><span class="op">)</span><span class="op">{</span>
  <span class="fu">RNAmodR</span><span class="fu">:::</span><span class="fu">.SequenceDataFrame</span><span class="op">(</span><span class="st">"Example"</span>,<span class="va">df</span>, <span class="va">ranges</span>, <span class="va">sequence</span>, <span class="va">replicate</span>,
                               <span class="va">condition</span>, <span class="va">bamfiles</span>, <span class="va">seqinfo</span><span class="op">)</span>
<span class="op">}</span>
<span class="kw">setClass</span><span class="op">(</span>Class <span class="op">=</span> <span class="st">"ExampleSequenceData"</span>,
         contains <span class="op">=</span> <span class="st">"SequenceData"</span>,
         slots <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>unlistData <span class="op">=</span> <span class="st">"ExampleSequenceDataFrame"</span><span class="op">)</span>,
         prototype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>unlistData <span class="op">=</span> <span class="fu">ExampleSequenceDataFrame</span><span class="op">(</span><span class="op">)</span>,
                          unlistType <span class="op">=</span> <span class="st">"ExampleSequenceDataFrame"</span>,
                          minQuality <span class="op">=</span> <span class="fl">5L</span>,
                          dataDescription <span class="op">=</span> <span class="st">"Example data"</span><span class="op">)</span><span class="op">)</span>
<span class="va">ExampleSequenceData</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">bamfiles</span>, <span class="va">annotation</span>, <span class="va">sequences</span>, <span class="va">seqinfo</span>, <span class="va">...</span><span class="op">)</span><span class="op">{</span>
  <span class="fu">RNAmodR</span><span class="fu">:::</span><span class="fu"><a href="../reference/SequenceData-class.html">SequenceData</a></span><span class="op">(</span><span class="st">"Example"</span>, bamfiles <span class="op">=</span> <span class="va">bamfiles</span>, 
                         annotation <span class="op">=</span> <span class="va">annotation</span>, sequences <span class="op">=</span> <span class="va">sequences</span>,
                         seqinfo <span class="op">=</span> <span class="va">seqinfo</span>, <span class="va">...</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>Second, the <code>getData</code> function has to be implemented. This is used to load the data from a bam file and must return a named list <code>IntegerList</code>, <code>NumericList</code> or <code>CompressedSplitDataFrameList</code> per file.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">setMethod</span><span class="op">(</span><span class="st">"getData"</span>,
          signature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ExampleSequenceData"</span>,
                        bamfiles <span class="op">=</span> <span class="st">"BamFileList"</span>,
                        grl <span class="op">=</span> <span class="st">"GRangesList"</span>,
                        sequences <span class="op">=</span> <span class="st">"XStringSet"</span>,
                        param <span class="op">=</span> <span class="st">"ScanBamParam"</span><span class="op">)</span>,
          definition <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">bamfiles</span>, <span class="va">grl</span>, <span class="va">sequences</span>, <span class="va">param</span>, <span class="va">args</span><span class="op">)</span><span class="op">{</span>
            <span class="co">###</span>
          <span class="op">}</span>
<span class="op">)</span></code></pre></div>
<p>Third, the <code>aggregate</code> function has to be implemented. This function is used to aggregate data over replicates for all or one of the conditions. The resulting data is passed on to the <code>Modifier</code> class.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">setMethod</span><span class="op">(</span><span class="st">"aggregateData"</span>,
          signature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ExampleSequenceData"</span><span class="op">)</span>,
          <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">condition</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Both"</span>,<span class="st">"Treated"</span>,<span class="st">"Control"</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
            <span class="co">###</span>
          <span class="op">}</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="a-new-modifier-class" class="section level1">
<h1 class="hasAnchor">
<a href="#a-new-modifier-class" class="anchor"></a>A new <code>Modifier</code> class</h1>
<p>A new <code>Modifier</code> class is probably the main class, which needs to be implemented. Three variable have to be set. <code>mod</code> must be a single element from the <code><a href="https://rdrr.io/pkg/Modstrings/man/alphabet.html">Modstrings::shortName(Modstrings::ModRNAString())</a></code>. <code>score</code> is the default score, which is used for several function. A column with this name should be returned from the <code>aggregate</code> function. <code>dataType</code> defines the <code>SequenceData</code> class to be used. <code>dataType</code> can contain multiple names of a <code>SequenceData</code> class, which are then combined to form a <code>SequenceDataSet</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">setClass</span><span class="op">(</span><span class="st">"ModExample"</span>,
         contains <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"RNAModifier"</span><span class="op">)</span>,
         prototype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>mod <span class="op">=</span> <span class="st">"X"</span>,
                          score <span class="op">=</span> <span class="st">"score"</span>,
                          dataType <span class="op">=</span> <span class="st">"ExampleSequenceData"</span><span class="op">)</span><span class="op">)</span>
<span class="va">ModExample</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">annotation</span>, <span class="va">sequences</span>, <span class="va">seqinfo</span>, <span class="va">...</span><span class="op">)</span><span class="op">{</span>
  <span class="fu">RNAmodR</span><span class="fu">:::</span><span class="fu"><a href="../reference/Modifier-class.html">Modifier</a></span><span class="op">(</span><span class="st">"ModExample"</span>, x <span class="op">=</span> <span class="va">x</span>, annotation <span class="op">=</span> <span class="va">annotation</span>,
                     sequences <span class="op">=</span> <span class="va">sequences</span>, seqinfo <span class="op">=</span> <span class="va">seqinfo</span>, <span class="va">...</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><code>dataType</code> can also be a <code>list</code> of <code>character</code> vectors, which leads then to the creation of <code>SequenceDataList</code>. However, for now this is a hypothetical case and should only be used, if the detection of a modification requires bam files from two or more different methods to be used to detect one modification.</p>
<p>The <code>settings&lt;-</code> function can be amended to save specifc settings ( <code>.norm_example_args</code> must be defined seperatly to normalize input arguments in any way one sees fit).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">setReplaceMethod</span><span class="op">(</span>f <span class="op">=</span> <span class="st">"settings"</span>, 
                 signature <span class="op">=</span> <span class="fu">signature</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ModExample"</span><span class="op">)</span>,
                 definition <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">value</span><span class="op">)</span><span class="op">{</span>
                   <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">callNextMethod</span><span class="op">(</span><span class="op">)</span>
                   <span class="co"># validate special setting here</span>
                   <span class="va">x</span><span class="op">@</span><span class="va">settings</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unname.html">unname</a></span><span class="op">(</span><span class="fu">.norm_example_args</span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span>
                   <span class="va">x</span>
                 <span class="op">}</span><span class="op">)</span></code></pre></div>
<p>The <code>aggregateData</code> function is used to take the aggregated data from the <code>SequenceData</code> object and to calculate the specific scores, which are then stored in the <code>aggregate</code> slot.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">setMethod</span><span class="op">(</span>f <span class="op">=</span> <span class="st">"aggregateData"</span>, 
          signature <span class="op">=</span> <span class="fu">signature</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ModExample"</span><span class="op">)</span>,
          definition <span class="op">=</span> 
            <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">force</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">{</span>
              <span class="co"># Some data with element per transcript</span>
            <span class="op">}</span>
<span class="op">)</span></code></pre></div>
<p>The <code>findMod</code> function takes the aggregate data and searches for modifications, which are then returned as a GRanges object and stored in the <code>modifications</code> slot.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">setMethod</span><span class="op">(</span><span class="st">"findMod"</span>,
          signature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ModExample"</span><span class="op">)</span>,
          <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span>
            <span class="co"># an element per modification found.</span>
          <span class="op">}</span>
<span class="op">)</span></code></pre></div>
<div id="a-new-modifierset-class" class="section level2">
<h2 class="hasAnchor">
<a href="#a-new-modifierset-class" class="anchor"></a>A new <code>ModifierSet</code> class</h2>
<p>The <code>ModifierSet</code> class is implemented very easily by defining the class and the constructor. The functionality is defined by the <code>Modifier</code> class.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">setClass</span><span class="op">(</span><span class="st">"ModSetExample"</span>,
         contains <span class="op">=</span> <span class="st">"ModifierSet"</span>,
         prototype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>elementType <span class="op">=</span> <span class="st">"ModExample"</span><span class="op">)</span><span class="op">)</span>
<span class="va">ModSetExample</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">annotation</span>, <span class="va">sequences</span>, <span class="va">seqinfo</span>, <span class="va">...</span><span class="op">)</span><span class="op">{</span>
  <span class="fu">RNAmodR</span><span class="fu">:::</span><span class="fu"><a href="../reference/ModifierSet-class.html">ModifierSet</a></span><span class="op">(</span><span class="st">"ModExample"</span>, x <span class="op">=</span> <span class="va">x</span>, annotation <span class="op">=</span> <span class="va">annotation</span>,
                        sequences <span class="op">=</span> <span class="va">sequences</span>, seqinfo <span class="op">=</span> <span class="va">seqinfo</span>, <span class="va">...</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
</div>
</div>
<div id="visualization-functions" class="section level1">
<h1 class="hasAnchor">
<a href="#visualization-functions" class="anchor"></a>Visualization functions</h1>
<p>Additional functions, which need to be implemented, are <code>getDataTrack</code> for the new <code>SequenceData</code> and new <code>Modifier</code> classes and <code>plotData</code>/<code>plotDataByCoord</code> for the new <code>Modifier</code> and <code>ModifierSet</code> classes. <code>name</code> defines a transcript name found in <code><a href="https://rdrr.io/r/base/names.html">names(ranges(x))</a></code> and <code>type</code> is the data type typically found as a column in the <code>aggregate</code> slot.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">setMethod</span><span class="op">(</span>
  f <span class="op">=</span> <span class="st">"getDataTrack"</span>,
  signature <span class="op">=</span> <span class="fu">signature</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ExampleSequenceData"</span><span class="op">)</span>,
  definition <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span>
    <span class="co">###</span>
  <span class="op">}</span>
<span class="op">)</span>
<span class="kw">setMethod</span><span class="op">(</span>
  f <span class="op">=</span> <span class="st">"getDataTrack"</span>,
  signature <span class="op">=</span> <span class="fu">signature</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ModExample"</span><span class="op">)</span>,
  definition <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, <span class="va">type</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span>
  <span class="op">}</span>
<span class="op">)</span>
<span class="kw">setMethod</span><span class="op">(</span>
  f <span class="op">=</span> <span class="st">"plotDataByCoord"</span>,
  signature <span class="op">=</span> <span class="fu">signature</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ModExample"</span>, coord <span class="op">=</span> <span class="st">"GRanges"</span><span class="op">)</span>,
  definition <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">type</span> <span class="op">=</span> <span class="st">"score"</span>, <span class="va">window.size</span> <span class="op">=</span> <span class="fl">15L</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span>
  <span class="op">}</span>
<span class="op">)</span>
<span class="kw">setMethod</span><span class="op">(</span>
  f <span class="op">=</span> <span class="st">"plotData"</span>,
  signature <span class="op">=</span> <span class="fu">signature</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ModExample"</span><span class="op">)</span>,
  definition <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, <span class="va">from</span>, <span class="va">to</span>, <span class="va">type</span> <span class="op">=</span> <span class="st">"score"</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span>
  <span class="op">}</span>
<span class="op">)</span>
<span class="kw">setMethod</span><span class="op">(</span>
  f <span class="op">=</span> <span class="st">"plotDataByCoord"</span>,
  signature <span class="op">=</span> <span class="fu">signature</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ModSetExample"</span>, coord <span class="op">=</span> <span class="st">"GRanges"</span><span class="op">)</span>,
  definition <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">type</span> <span class="op">=</span> <span class="st">"score"</span>, <span class="va">window.size</span> <span class="op">=</span> <span class="fl">15L</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span>
  <span class="op">}</span>
<span class="op">)</span>
<span class="kw">setMethod</span><span class="op">(</span>
  f <span class="op">=</span> <span class="st">"plotData"</span>,
  signature <span class="op">=</span> <span class="fu">signature</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ModSetExample"</span><span class="op">)</span>,
  definition <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, <span class="va">from</span>, <span class="va">to</span>, <span class="va">type</span> <span class="op">=</span> <span class="st">"score"</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span>
  <span class="op">}</span>
<span class="op">)</span></code></pre></div>
<p>If unsure, how to modify these functions, have a look a the code in the <code>Modifier-Inosine-viz.R</code> file of this package.</p>
</div>
<div id="summary" class="section level1">
<h1 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary</h1>
<p>As suggested directly above, for a more detailed example have a look at the <code>ModInosine</code> class source code found in the <code>Modifier-Inosine-class.R</code> and <code>Modifier-Inosine-viz.R</code> files of this package.</p>
</div>
<div id="sessioninfo" class="section level1">
<h1 class="hasAnchor">
<a href="#sessioninfo" class="anchor"></a>Sessioninfo</h1>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R Under development (unstable) (2021-01-07 r79806)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] RNAmodR_1.5.3        Modstrings_1.7.1     Biostrings_2.59.2   
##  [4] XVector_0.31.1       GenomicRanges_1.43.1 GenomeInfoDb_1.27.3 
##  [7] IRanges_2.25.6       S4Vectors_0.29.6     BiocGenerics_0.37.0 
## [10] BiocStyle_2.19.1    
## 
## loaded via a namespace (and not attached):
##   [1] colorspace_2.0-0            rjson_0.2.20               
##   [3] colorRamps_2.3              ellipsis_0.3.1             
##   [5] rprojroot_2.0.2             biovizBase_1.39.0          
##   [7] htmlTable_2.1.0             base64enc_0.1-3            
##   [9] fs_1.5.0                    dichromat_2.0-0            
##  [11] rstudioapi_0.13             bit64_4.0.5                
##  [13] AnnotationDbi_1.53.0        xml2_1.3.2                 
##  [15] splines_4.1.0               knitr_1.30                 
##  [17] Formula_1.2-4               Rsamtools_2.7.0            
##  [19] cluster_2.1.0               dbplyr_2.0.0               
##  [21] png_0.1-7                   BiocManager_1.30.10        
##  [23] compiler_4.1.0              httr_1.4.2                 
##  [25] backports_1.2.1             lazyeval_0.2.2             
##  [27] assertthat_0.2.1            Matrix_1.3-2               
##  [29] htmltools_0.5.1             prettyunits_1.1.1          
##  [31] tools_4.1.0                 gtable_0.3.0               
##  [33] glue_1.4.2                  GenomeInfoDbData_1.2.4     
##  [35] reshape2_1.4.4              dplyr_1.0.2                
##  [37] rappdirs_0.3.1              Rcpp_1.0.5                 
##  [39] Biobase_2.51.0              pkgdown_1.6.1              
##  [41] vctrs_0.3.6                 rtracklayer_1.51.3         
##  [43] xfun_0.20                   stringr_1.4.0              
##  [45] lifecycle_0.2.0             ensembldb_2.15.1           
##  [47] restfulr_0.0.13             XML_3.99-0.5               
##  [49] zlibbioc_1.37.0             scales_1.1.1               
##  [51] BSgenome_1.59.2             VariantAnnotation_1.37.1   
##  [53] ProtGenerics_1.23.6         ragg_0.4.1                 
##  [55] hms_0.5.3                   MatrixGenerics_1.3.0       
##  [57] SummarizedExperiment_1.21.1 AnnotationFilter_1.15.0    
##  [59] RColorBrewer_1.1-2          yaml_2.2.1                 
##  [61] curl_4.3                    memoise_1.1.0              
##  [63] gridExtra_2.3               ggplot2_3.3.3              
##  [65] biomaRt_2.47.1              rpart_4.1-15               
##  [67] latticeExtra_0.6-29         stringi_1.5.3              
##  [69] RSQLite_2.2.2               BiocIO_1.1.2               
##  [71] desc_1.2.0                  checkmate_2.0.0            
##  [73] GenomicFeatures_1.43.3      filelock_1.0.2             
##  [75] BiocParallel_1.25.2         rlang_0.4.10               
##  [77] pkgconfig_2.0.3             systemfonts_0.3.2          
##  [79] matrixStats_0.57.0          bitops_1.0-6               
##  [81] evaluate_0.14               lattice_0.20-41            
##  [83] ROCR_1.0-11                 purrr_0.3.4                
##  [85] GenomicAlignments_1.27.2    htmlwidgets_1.5.3          
##  [87] bit_4.0.4                   tidyselect_1.1.0           
##  [89] plyr_1.8.6                  magrittr_2.0.1             
##  [91] bookdown_0.21               R6_2.5.0                   
##  [93] generics_0.1.0              Hmisc_4.4-2                
##  [95] DelayedArray_0.17.7         DBI_1.1.0                  
##  [97] pillar_1.4.7                foreign_0.8-81             
##  [99] survival_3.2-7              RCurl_1.98-1.2             
## [101] nnet_7.3-14                 tibble_3.0.4               
## [103] crayon_1.3.4                BiocFileCache_1.15.1       
## [105] rmarkdown_2.6               jpeg_0.1-8.1               
## [107] progress_1.2.2              grid_4.1.0                 
## [109] data.table_1.13.6           blob_1.2.1                 
## [111] digest_0.6.27               textshaping_0.2.1          
## [113] openssl_1.4.3               munsell_0.5.0              
## [115] Gviz_1.35.3                 askpass_1.1</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
