<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RNAmodR: creating classes for additional modification detection from high throughput sequencing. â€¢ RNAmodR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="RNAmodR: creating classes for additional modification detection from high throughput sequencing.">
<meta property="og:description" content="RNAmodR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.17.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/RNAmodR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/RNAmodR.creation.html">RNAmodR: creating classes for additional modification detection from high throughput sequencing.</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/FelixErnst/RNAmodR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>RNAmodR: creating classes for additional
modification detection from high throughput sequencing.</h1>
                        <h4 data-toc-skip class="author">Felix G.M.
Ernst</h4>
            
            <h4 data-toc-skip class="date">2024-03-18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/FelixErnst/RNAmodR/blob/HEAD/vignettes/RNAmodR.creation.Rmd" class="external-link"><code>vignettes/RNAmodR.creation.Rmd</code></a></small>
      <div class="hidden name"><code>RNAmodR.creation.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>For users interested in the general aspect of any
<code>RNAmodR</code> based package please have a look at the <a href="RNAmodR.html">main vignette</a> of the package.</p>
<p>This vignette is aimed at developers and researchers, who want to use
the functionality of the <code>RNAmodR</code> package to develop a new
modification strategy based on high throughput sequencing data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/FelixErnst/RNAmodR" class="external-link">RNAmodR</a></span><span class="op">)</span></span></code></pre></div>
<p>Two classes have to be considered to establish a new analysis
pipeline using <code>RNAmodR</code>. These are the
<code>SequenceData</code> and the <code>Modifier</code> class.</p>
</div>
<div class="section level2">
<h2 id="a-new-sequencedata-class">A new <code>SequenceData</code> class<a class="anchor" aria-label="anchor" href="#a-new-sequencedata-class"></a>
</h2>
<p>First, the <code>SequenceData</code> class has to be considered.
Several classes are already implemented, which are:</p>
<ul>
<li><code>End5SequenceData</code></li>
<li><code>End3SequenceData</code></li>
<li><code>EndSequenceData</code></li>
<li><code>ProtectedEndSequenceData</code></li>
<li><code>CoverageSequenceData</code></li>
<li><code>PileupSequenceData</code></li>
<li><code>NormEnd5SequenceData</code></li>
<li><code>NormEnd3SequenceData</code></li>
</ul>
<p>If these cannot be reused, a new class can be implemented quite
easily. First the DataFrame class, the Data class and a constructor has
to defined. The only value, which has to be provided, is a default
<code>minQuality</code> integer value and some basic information.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/methods/setClass.html" class="external-link">setClass</a></span><span class="op">(</span>Class <span class="op">=</span> <span class="st">"ExampleSequenceDataFrame"</span>,</span>
<span>         contains <span class="op">=</span> <span class="st">"SequenceDFrame"</span><span class="op">)</span></span>
<span><span class="va">ExampleSequenceDataFrame</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">df</span>, <span class="va">ranges</span>, <span class="va">sequence</span>, <span class="va">replicate</span>,</span>
<span>                                      <span class="va">condition</span>, <span class="va">bamfiles</span>, <span class="va">seqinfo</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">RNAmodR</span><span class="fu">:::</span><span class="fu">.SequenceDataFrame</span><span class="op">(</span><span class="st">"Example"</span>,<span class="va">df</span>, <span class="va">ranges</span>, <span class="va">sequence</span>, <span class="va">replicate</span>,</span>
<span>                               <span class="va">condition</span>, <span class="va">bamfiles</span>, <span class="va">seqinfo</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/methods/setClass.html" class="external-link">setClass</a></span><span class="op">(</span>Class <span class="op">=</span> <span class="st">"ExampleSequenceData"</span>,</span>
<span>         contains <span class="op">=</span> <span class="st">"SequenceData"</span>,</span>
<span>         slots <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>unlistData <span class="op">=</span> <span class="st">"ExampleSequenceDataFrame"</span><span class="op">)</span>,</span>
<span>         prototype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>unlistData <span class="op">=</span> <span class="fu">ExampleSequenceDataFrame</span><span class="op">(</span><span class="op">)</span>,</span>
<span>                          unlistType <span class="op">=</span> <span class="st">"ExampleSequenceDataFrame"</span>,</span>
<span>                          minQuality <span class="op">=</span> <span class="fl">5L</span>,</span>
<span>                          dataDescription <span class="op">=</span> <span class="st">"Example data"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ExampleSequenceData</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">bamfiles</span>, <span class="va">annotation</span>, <span class="va">sequences</span>, <span class="va">seqinfo</span>, <span class="va">...</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">RNAmodR</span><span class="fu">:::</span><span class="fu"><a href="../reference/SequenceData-class.html">SequenceData</a></span><span class="op">(</span><span class="st">"Example"</span>, bamfiles <span class="op">=</span> <span class="va">bamfiles</span>, </span>
<span>                         annotation <span class="op">=</span> <span class="va">annotation</span>, sequences <span class="op">=</span> <span class="va">sequences</span>,</span>
<span>                         seqinfo <span class="op">=</span> <span class="va">seqinfo</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Second, the <code>getData</code> function has to be implemented. This
is used to load the data from a bam file and must return a named list
<code>IntegerList</code>, <code>NumericList</code> or
<code>CompressedSplitDataFrameList</code> per file.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/methods/setMethod.html" class="external-link">setMethod</a></span><span class="op">(</span><span class="st">"getData"</span>,</span>
<span>          signature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ExampleSequenceData"</span>,</span>
<span>                        bamfiles <span class="op">=</span> <span class="st">"BamFileList"</span>,</span>
<span>                        grl <span class="op">=</span> <span class="st">"GRangesList"</span>,</span>
<span>                        sequences <span class="op">=</span> <span class="st">"XStringSet"</span>,</span>
<span>                        param <span class="op">=</span> <span class="st">"ScanBamParam"</span><span class="op">)</span>,</span>
<span>          definition <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">bamfiles</span>, <span class="va">grl</span>, <span class="va">sequences</span>, <span class="va">param</span>, <span class="va">args</span><span class="op">)</span><span class="op">{</span></span>
<span>            <span class="co">###</span></span>
<span>          <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Third, the <code>aggregate</code> function has to be implemented.
This function is used to aggregate data over replicates for all or one
of the conditions. The resulting data is passed on to the
<code>Modifier</code> class.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/methods/setMethod.html" class="external-link">setMethod</a></span><span class="op">(</span><span class="st">"aggregateData"</span>,</span>
<span>          signature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ExampleSequenceData"</span><span class="op">)</span>,</span>
<span>          <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">condition</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Both"</span>,<span class="st">"Treated"</span>,<span class="st">"Control"</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>            <span class="co">###</span></span>
<span>          <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="a-new-modifier-class">A new <code>Modifier</code> class<a class="anchor" aria-label="anchor" href="#a-new-modifier-class"></a>
</h2>
<p>A new <code>Modifier</code> class is probably the main class, which
needs to be implemented. Three variable have to be set. <code>mod</code>
must be a single element from the
<code>Modstrings::shortName(Modstrings::ModRNAString())</code>.
<code>score</code> is the default score, which is used for several
function. A column with this name should be returned from the
<code>aggregate</code> function. <code>dataType</code> defines the
<code>SequenceData</code> class to be used. <code>dataType</code> can
contain multiple names of a <code>SequenceData</code> class, which are
then combined to form a <code>SequenceDataSet</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/methods/setClass.html" class="external-link">setClass</a></span><span class="op">(</span><span class="st">"ModExample"</span>,</span>
<span>         contains <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNAModifier"</span><span class="op">)</span>,</span>
<span>         prototype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mod <span class="op">=</span> <span class="st">"X"</span>,</span>
<span>                          score <span class="op">=</span> <span class="st">"score"</span>,</span>
<span>                          dataType <span class="op">=</span> <span class="st">"ExampleSequenceData"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ModExample</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">annotation</span>, <span class="va">sequences</span>, <span class="va">seqinfo</span>, <span class="va">...</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">RNAmodR</span><span class="fu">:::</span><span class="fu"><a href="../reference/Modifier-class.html">Modifier</a></span><span class="op">(</span><span class="st">"ModExample"</span>, x <span class="op">=</span> <span class="va">x</span>, annotation <span class="op">=</span> <span class="va">annotation</span>,</span>
<span>                     sequences <span class="op">=</span> <span class="va">sequences</span>, seqinfo <span class="op">=</span> <span class="va">seqinfo</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><code>dataType</code> can also be a <code>list</code> of
<code>character</code> vectors, which leads then to the creation of
<code>SequenceDataList</code>. However, for now this is a hypothetical
case and should only be used, if the detection of a modification
requires bam files from two or more different methods to be used to
detect one modification.</p>
<p>The <code>settings&lt;-</code> function can be amended to save
specifc settings ( <code>.norm_example_args</code> must be defined
seperatly to normalize input arguments in any way one sees fit).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/GenericFunctions.html" class="external-link">setReplaceMethod</a></span><span class="op">(</span>f <span class="op">=</span> <span class="st">"settings"</span>, </span>
<span>                 signature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/GenericFunctions.html" class="external-link">signature</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ModExample"</span><span class="op">)</span>,</span>
<span>                 definition <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">value</span><span class="op">)</span><span class="op">{</span></span>
<span>                   <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/NextMethod.html" class="external-link">callNextMethod</a></span><span class="op">(</span><span class="op">)</span></span>
<span>                   <span class="co"># validate special setting here</span></span>
<span>                   <span class="va">x</span><span class="op">@</span><span class="va">settings</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="fu">.norm_example_args</span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span></span>
<span>                   <span class="va">x</span></span>
<span>                 <span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>The <code>aggregateData</code> function is used to take the
aggregated data from the <code>SequenceData</code> object and to
calculate the specific scores, which are then stored in the
<code>aggregate</code> slot.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/methods/setMethod.html" class="external-link">setMethod</a></span><span class="op">(</span>f <span class="op">=</span> <span class="st">"aggregateData"</span>, </span>
<span>          signature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/GenericFunctions.html" class="external-link">signature</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ModExample"</span><span class="op">)</span>,</span>
<span>          definition <span class="op">=</span> </span>
<span>            <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">force</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">{</span></span>
<span>              <span class="co"># Some data with element per transcript</span></span>
<span>            <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The <code>findMod</code> function takes the aggregate data and
searches for modifications, which are then returned as a GRanges object
and stored in the <code>modifications</code> slot.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/methods/setMethod.html" class="external-link">setMethod</a></span><span class="op">(</span><span class="st">"findMod"</span>,</span>
<span>          signature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ModExample"</span><span class="op">)</span>,</span>
<span>          <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>            <span class="co"># an element per modification found.</span></span>
<span>          <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="a-new-modifierset-class">A new <code>ModifierSet</code> class<a class="anchor" aria-label="anchor" href="#a-new-modifierset-class"></a>
</h3>
<p>The <code>ModifierSet</code> class is implemented very easily by
defining the class and the constructor. The functionality is defined by
the <code>Modifier</code> class.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/methods/setClass.html" class="external-link">setClass</a></span><span class="op">(</span><span class="st">"ModSetExample"</span>,</span>
<span>         contains <span class="op">=</span> <span class="st">"ModifierSet"</span>,</span>
<span>         prototype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>elementType <span class="op">=</span> <span class="st">"ModExample"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ModSetExample</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">annotation</span>, <span class="va">sequences</span>, <span class="va">seqinfo</span>, <span class="va">...</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">RNAmodR</span><span class="fu">:::</span><span class="fu"><a href="../reference/ModifierSet-class.html">ModifierSet</a></span><span class="op">(</span><span class="st">"ModExample"</span>, x <span class="op">=</span> <span class="va">x</span>, annotation <span class="op">=</span> <span class="va">annotation</span>,</span>
<span>                        sequences <span class="op">=</span> <span class="va">sequences</span>, seqinfo <span class="op">=</span> <span class="va">seqinfo</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="visualization-functions">Visualization functions<a class="anchor" aria-label="anchor" href="#visualization-functions"></a>
</h2>
<p>Additional functions, which need to be implemented, are
<code>getDataTrack</code> for the new <code>SequenceData</code> and new
<code>Modifier</code> classes and
<code>plotData</code>/<code>plotDataByCoord</code> for the new
<code>Modifier</code> and <code>ModifierSet</code> classes.
<code>name</code> defines a transcript name found in
<code>names(ranges(x))</code> and <code>type</code> is the data type
typically found as a column in the <code>aggregate</code> slot.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/methods/setMethod.html" class="external-link">setMethod</a></span><span class="op">(</span></span>
<span>  f <span class="op">=</span> <span class="st">"getDataTrack"</span>,</span>
<span>  signature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/GenericFunctions.html" class="external-link">signature</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ExampleSequenceData"</span><span class="op">)</span>,</span>
<span>  definition <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co">###</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/methods/setMethod.html" class="external-link">setMethod</a></span><span class="op">(</span></span>
<span>  f <span class="op">=</span> <span class="st">"getDataTrack"</span>,</span>
<span>  signature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/GenericFunctions.html" class="external-link">signature</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ModExample"</span><span class="op">)</span>,</span>
<span>  definition <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, <span class="va">type</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/methods/setMethod.html" class="external-link">setMethod</a></span><span class="op">(</span></span>
<span>  f <span class="op">=</span> <span class="st">"plotDataByCoord"</span>,</span>
<span>  signature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/GenericFunctions.html" class="external-link">signature</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ModExample"</span>, coord <span class="op">=</span> <span class="st">"GRanges"</span><span class="op">)</span>,</span>
<span>  definition <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">type</span> <span class="op">=</span> <span class="st">"score"</span>, <span class="va">window.size</span> <span class="op">=</span> <span class="fl">15L</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/methods/setMethod.html" class="external-link">setMethod</a></span><span class="op">(</span></span>
<span>  f <span class="op">=</span> <span class="st">"plotData"</span>,</span>
<span>  signature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/GenericFunctions.html" class="external-link">signature</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ModExample"</span><span class="op">)</span>,</span>
<span>  definition <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, <span class="va">from</span>, <span class="va">to</span>, <span class="va">type</span> <span class="op">=</span> <span class="st">"score"</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/methods/setMethod.html" class="external-link">setMethod</a></span><span class="op">(</span></span>
<span>  f <span class="op">=</span> <span class="st">"plotDataByCoord"</span>,</span>
<span>  signature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/GenericFunctions.html" class="external-link">signature</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ModSetExample"</span>, coord <span class="op">=</span> <span class="st">"GRanges"</span><span class="op">)</span>,</span>
<span>  definition <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">coord</span>, <span class="va">type</span> <span class="op">=</span> <span class="st">"score"</span>, <span class="va">window.size</span> <span class="op">=</span> <span class="fl">15L</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/methods/setMethod.html" class="external-link">setMethod</a></span><span class="op">(</span></span>
<span>  f <span class="op">=</span> <span class="st">"plotData"</span>,</span>
<span>  signature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/GenericFunctions.html" class="external-link">signature</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ModSetExample"</span><span class="op">)</span>,</span>
<span>  definition <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">name</span>, <span class="va">from</span>, <span class="va">to</span>, <span class="va">type</span> <span class="op">=</span> <span class="st">"score"</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>If unsure, how to modify these functions, have a look a the code in
the <code>Modifier-Inosine-viz.R</code> file of this package.</p>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>As suggested directly above, for a more detailed example have a look
at the <code>ModInosine</code> class source code found in the
<code>Modifier-Inosine-class.R</code> and
<code>Modifier-Inosine-viz.R</code> files of this package.</p>
</div>
<div class="section level2">
<h2 id="sessioninfo">Sessioninfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2024-03-13 r86113)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] RNAmodR_1.17.1       Modstrings_1.19.0    Biostrings_2.71.4   </span></span>
<span><span class="co">##  [4] XVector_0.43.1       GenomicRanges_1.55.3 GenomeInfoDb_1.39.9 </span></span>
<span><span class="co">##  [7] IRanges_2.37.1       S4Vectors_0.41.4     BiocGenerics_0.49.1 </span></span>
<span><span class="co">## [10] BiocStyle_2.31.0    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3          rstudioapi_0.15.0          </span></span>
<span><span class="co">##   [3] jsonlite_1.8.8              magrittr_2.0.3             </span></span>
<span><span class="co">##   [5] GenomicFeatures_1.55.4      rmarkdown_2.26             </span></span>
<span><span class="co">##   [7] fs_1.6.3                    BiocIO_1.13.0              </span></span>
<span><span class="co">##   [9] zlibbioc_1.49.3             ragg_1.3.0                 </span></span>
<span><span class="co">##  [11] vctrs_0.6.5                 ROCR_1.0-11                </span></span>
<span><span class="co">##  [13] memoise_2.0.1               Rsamtools_2.19.3           </span></span>
<span><span class="co">##  [15] RCurl_1.98-1.14             base64enc_0.1-3            </span></span>
<span><span class="co">##  [17] htmltools_0.5.7             S4Arrays_1.3.6             </span></span>
<span><span class="co">##  [19] progress_1.2.3              curl_5.2.1                 </span></span>
<span><span class="co">##  [21] SparseArray_1.3.4           Formula_1.2-5              </span></span>
<span><span class="co">##  [23] sass_0.4.9                  bslib_0.6.1                </span></span>
<span><span class="co">##  [25] htmlwidgets_1.6.4           desc_1.4.3                 </span></span>
<span><span class="co">##  [27] plyr_1.8.9                  Gviz_1.47.1                </span></span>
<span><span class="co">##  [29] httr2_1.0.0                 cachem_1.0.8               </span></span>
<span><span class="co">##  [31] GenomicAlignments_1.39.4    lifecycle_1.0.4            </span></span>
<span><span class="co">##  [33] pkgconfig_2.0.3             Matrix_1.6-5               </span></span>
<span><span class="co">##  [35] R6_2.5.1                    fastmap_1.1.1              </span></span>
<span><span class="co">##  [37] GenomeInfoDbData_1.2.11     MatrixGenerics_1.15.0      </span></span>
<span><span class="co">##  [39] digest_0.6.35               colorspace_2.1-0           </span></span>
<span><span class="co">##  [41] AnnotationDbi_1.65.2        textshaping_0.3.7          </span></span>
<span><span class="co">##  [43] Hmisc_5.1-2                 RSQLite_2.3.5              </span></span>
<span><span class="co">##  [45] filelock_1.0.3              colorRamps_2.3.4           </span></span>
<span><span class="co">##  [47] fansi_1.0.6                 httr_1.4.7                 </span></span>
<span><span class="co">##  [49] abind_1.4-5                 compiler_4.4.0             </span></span>
<span><span class="co">##  [51] bit64_4.0.5                 htmlTable_2.4.2            </span></span>
<span><span class="co">##  [53] backports_1.4.1             BiocParallel_1.37.1        </span></span>
<span><span class="co">##  [55] DBI_1.2.2                   biomaRt_2.59.1             </span></span>
<span><span class="co">##  [57] rappdirs_0.3.3              DelayedArray_0.29.9        </span></span>
<span><span class="co">##  [59] rjson_0.2.21                tools_4.4.0                </span></span>
<span><span class="co">##  [61] foreign_0.8-86              nnet_7.3-19                </span></span>
<span><span class="co">##  [63] glue_1.7.0                  restfulr_0.0.15            </span></span>
<span><span class="co">##  [65] grid_4.4.0                  checkmate_2.3.1            </span></span>
<span><span class="co">##  [67] reshape2_1.4.4              cluster_2.1.6              </span></span>
<span><span class="co">##  [69] generics_0.1.3              gtable_0.3.4               </span></span>
<span><span class="co">##  [71] BSgenome_1.71.2             ensembldb_2.27.1           </span></span>
<span><span class="co">##  [73] data.table_1.15.2           hms_1.1.3                  </span></span>
<span><span class="co">##  [75] xml2_1.3.6                  utf8_1.2.4                 </span></span>
<span><span class="co">##  [77] pillar_1.9.0                stringr_1.5.1              </span></span>
<span><span class="co">##  [79] dplyr_1.1.4                 BiocFileCache_2.11.1       </span></span>
<span><span class="co">##  [81] lattice_0.22-5              deldir_2.0-4               </span></span>
<span><span class="co">##  [83] rtracklayer_1.63.1          bit_4.0.5                  </span></span>
<span><span class="co">##  [85] biovizBase_1.51.0           tidyselect_1.2.1           </span></span>
<span><span class="co">##  [87] knitr_1.45                  gridExtra_2.3              </span></span>
<span><span class="co">##  [89] bookdown_0.38               ProtGenerics_1.35.4        </span></span>
<span><span class="co">##  [91] SummarizedExperiment_1.33.3 xfun_0.42                  </span></span>
<span><span class="co">##  [93] Biobase_2.63.0              matrixStats_1.2.0          </span></span>
<span><span class="co">##  [95] stringi_1.8.3               lazyeval_0.2.2             </span></span>
<span><span class="co">##  [97] yaml_2.3.8                  evaluate_0.23              </span></span>
<span><span class="co">##  [99] codetools_0.2-19            interp_1.1-6               </span></span>
<span><span class="co">## [101] tibble_3.2.1                BiocManager_1.30.22        </span></span>
<span><span class="co">## [103] cli_3.6.2                   rpart_4.1.23               </span></span>
<span><span class="co">## [105] systemfonts_1.0.6           munsell_0.5.0              </span></span>
<span><span class="co">## [107] jquerylib_0.1.4             Rcpp_1.0.12                </span></span>
<span><span class="co">## [109] dichromat_2.0-0.1           dbplyr_2.4.0               </span></span>
<span><span class="co">## [111] png_0.1-8                   XML_3.99-0.16.1            </span></span>
<span><span class="co">## [113] parallel_4.4.0              pkgdown_2.0.7              </span></span>
<span><span class="co">## [115] ggplot2_3.5.0               blob_1.2.4                 </span></span>
<span><span class="co">## [117] prettyunits_1.2.0           jpeg_0.1-10                </span></span>
<span><span class="co">## [119] latticeExtra_0.6-30         AnnotationFilter_1.27.0    </span></span>
<span><span class="co">## [121] bitops_1.0-7                txdbmaker_0.99.6           </span></span>
<span><span class="co">## [123] VariantAnnotation_1.49.6    scales_1.3.0               </span></span>
<span><span class="co">## [125] purrr_1.0.2                 crayon_1.5.2               </span></span>
<span><span class="co">## [127] rlang_1.1.3                 KEGGREST_1.43.0</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
